---
title: "RADICL-seq Notebook"
author: "Federico Agostini <federico.agostini@crick.ac.uk>"
date: "19 July 2017"
output:
   html_notebook:
      fig_align: center
      code_folding: hide
---

```{r setup, include=FALSE, warning=FALSE}
require("knitr")
knitr::opts_chunk$set(dev='CairoPNG', dpi=300, fig_height=480, fig_width=800)
```

## Session information: {-}

```{r setwd, echo=FALSE}
working.folder = "/Users/agostif/Desktop/Bonetti/"

while( !dir.exists(working.folder) )
   working.folder = readline(prompt="Working folder not found! Please enter a valid path: ")
if( substr(working.folder, nchar(working.folder), nchar(working.folder)) != "/")
   working.folder = paste0(working.folder, "/")
setwd(working.folder)

knitr::opts_chunk$set(root.dir=working.folder)

cat(sessionInfo()$R.version$version.string, fill=TRUE)
cat(paste("Platform", sessionInfo()$platform), fill=TRUE)
cat(paste("Running under", sessionInfo()$running), fill=TRUE)
cat(paste("Last knitted on", date()), fill=TRUE)
```

```{r packages, message=FALSE, results="hide"}
require("BSgenome.Mmusculus.UCSC.mm10")
require("GenomicFeatures")
require("data.table")
require("Rmisc")
require("ggpubr")
require("RColorBrewer")
require("gridExtra")
require("DT")
require("UpSetR")
require("VennDiagram")
require("eulerr")
require("ComplexHeatmap")
require("circlize")
require("Cairo")
require("png")
require("ggthemes")
require("cowplot")
require("fedefun")
require("scales")
require("ggsci")
require("viridis")
require("ggforce")
require("GGally")
require("CAGEr")
```

## Load the packages: {-}

  * BSgenome.Mmusculus.UCSC.mm10
  * GenomicFeatures
  * data.table
  * RColorBrewer
  * ggpubr
  * gridExtra
  * DT
  * VennDiagram
  * eulerr
  * UpSetR
  * ComplexHeatmap
  * circlize
  * Cairo

Custom function: _read.paired.bed()_

```{r readPairedBed, echo=TRUE}
read.paired.bed <- function(file, min.mapq=37){
   columns = paste(c("chr", "start", "stop", "name", "score", "strand"), rep(c("R", "D"), each=6), sep=".")
   types = rep(c("character", "numeric", "numeric", "character", "numeric", "character"), 2)
   if( grepl("*.gz", file) ){
      bed = fread(paste("gunzip -c", file),
                  col.names=c(columns, "annotation.R", "annotation.D", "distance.R.D"),
                  colClasses=c(types, rep("character", 3)))
   }else{
      bed = fread(file,
                  col.names=c(columns, "annotation.R", "annotation.D", "distance.R.D"),
                  colClasses=c(types, rep("character", 3)))
   }
   
   bed = bed[score.R>=min.mapq & score.D>=min.mapq]
   bed = bed[(stop.R-start.R+1)<30 & (stop.D-start.D+1)<30]
   
   bed = GRangesList(resize(with(bed,
                          GRanges(chr.R, IRanges(start.R, stop.R), strand.R, name=name.R,
                                  score=score.R, annotation=annotation.R, distance=distance.R.D, index=1:nrow(bed))),
                          width=1, fix="center"),
                     resize(with(bed,
                          GRanges(chr.D, IRanges(start.D, stop.D), strand.D, name=name.D,
                                  score=score.D, annotation=annotation.D, distance=distance.R.D, index=1:nrow(bed))),
                          width=1, fix="center"))
   
   return(bed)
}
```

The function reads a non-standard BED file and convert it into a GRangesList object containing the RNA (first element in the list) and DNA (second element in the list) information.
It stores also the additional columns, though these will be removed as they are not useful for the analysis.

__Note:__ There is no on-the-fly filtering on the BWA score.

## Reference annotation

<!-- ### Gencode annotation -->

<!-- The _Gencode mouse reference release (version M14)_ was obtained from [here](https://www.gencodegenes.org/mouse_releases/14.html). -->
<!-- The _comprehensive gene annotation_ GFF3 file is read and gene coordinates and information are store in a data.table objects. -->
<!-- Genic regions have been sub-divided into _5'-UTR_, _exon_, _intron_ and _3'-UTR_. -->
<!-- The analysis is limited to the standard chromosome, to level 1 (_validated_), 2 (_manual annotation_) and 3 (_automated annotation_) genes and to the following genes biotypes: -->

<!--   * protein_coding -->
<!--   * lincRNA -->
<!--   * antisense -->
<!--   * processed_pseudogene -->
<!--   * unprocessed_pseudogene -->
<!--   * miRNA -->
<!--   * snRNA -->
<!--   * snoRNA -->
<!--   * processed_transcript -->
<!--   * misc_RNA -->

<!-- ```{r createAnnotationM14, eval=FALSE} -->
<!-- all_gff = fread("gunzip -c gencode.vM14.annotation.gff3.nohashtags.gz", -->
<!--                 sel=c(1, 3, 4, 5, 7, 9), col.names=c("chr", "feature", "start", "end", "strand", "attr")) -->
<!-- all_gff = all_gff[!feature%in%c("transcript", "start_codon", "stop_codon", "stop_codon_redefined_as_selenocysteine")] -->
<!-- all_gff[, attr := gsub("\"|;", " ", attr)] -->
<!-- all_gff[, gene_id := sub(".*gene_id=(\\S+).*", "\\1", attr)] -->
<!-- all_gff[, gene_name := sub(".*gene_name=(\\S+).*", "\\1", attr)] -->
<!-- all_gff[, gene_type := sub(".*gene_type=(\\S+).*", "\\1", attr)] -->
<!-- all_gff[, level := sub(".*level=(\\S+).*", "\\1", attr)] -->
<!-- all_gff[, attr := NULL] -->

<!-- biotype_names = unique(all_gff[, gene_type]) -->
<!-- gff = lapply(unique(all_gff[, gene_type]), -->
<!--                  function(x) -->
<!--                     all_gff[gene_type%in%as.character(x)]) -->
<!-- names(gff) = biotype_names -->

<!-- gff = lapply(gff, -->
<!--              function(x){ -->
<!--                 genes = with(x[feature%in%"gene",], GRanges(chr, IRanges(start, end), strand, gene_id)) -->
<!--                 genes = split(genes, genes$gene_id) -->
<!--                 exons = with(x[feature%in%"exon",], GRanges(chr, IRanges(start, end), strand, gene_id)) -->
<!--                 exons = reduce(split(exons, exons$gene_id)) -->
<!--                 introns = setdiff(genes, exons) -->
<!--                 utrs5 = with(x[feature%in%"five_prime_UTR",], GRanges(chr, IRanges(start, end), strand, gene_id)) -->
<!--                 utrs5 = reduce(split(utrs5, utrs5$gene_id)) -->
<!--                 utrs3 = with(x[feature%in%"three_prime_UTR",], GRanges(chr, IRanges(start, end), strand, gene_id)) -->
<!--                 utrs3 = reduce(split(utrs3, utrs3$gene_id)) -->

<!--                 exons5 = exons[names(exons)%in%names(utrs5)] -->
<!--                 exons = c(setdiff(exons5, utrs5), exons[!names(exons)%in%names(utrs5)]) -->
<!--                 exons3 = exons[names(exons)%in%names(utrs3)] -->
<!--                 exons = c(setdiff(exons3, utrs3), exons[!names(exons)%in%names(utrs3)]) -->

<!--                 myList = list(utrs5, exons, introns, utrs3) -->
<!--                 myList = lapply(myList, as.data.table) -->
<!--                 names(myList) = c("five_prime_UTR", "exon", "intron", "three_prime_UTR") -->
<!--                 myList = rbindlist(myList, idcol="gene_feature") -->
<!--                 setnames(myList, "group_name", "gene_id") -->
<!--                 myList[, list(seqnames, start, end, strand, gene_id, gene_feature)] -->
<!--              }) -->

<!-- gff = rbindlist(gff, idcol="gene_type") -->
<!-- setkey(gff, gene_id) -->
<!-- tab = all_gff[feature%in%"gene"][, list(gene_id, gene_name, level)] -->
<!-- setkey(tab, gene_id) -->
<!-- gff = gff[tab, nomatch=0] -->
<!-- setcolorder(gff, c(2:7, 1, 8:9)) -->

<!-- saveRDS(gff, file="gencode.annotation.m14_level_1_2_3.rds") -->
<!-- ``` -->

<!-- __Note:__ The `r gencode_version="m14"; toupper(gencode_version)` is the annotation version used for all downstream analyses. -->

<!-- ```{r processAnnotation, echo=FALSE} -->
<!-- if(!exists("gencode_version")) -->
<!--    gencode_version = "m14" -->

<!-- gff = readRDS(paste0("gencode.annotation.", gencode_version, "_level_1_2_3.rds")) -->

<!-- coding.sel = unique(gff[gene_type%in%"protein_coding", gene_id]) -->
<!-- noncoding.names = c("processed_pseudogene", "lincRNA", "antisense", "unprocessed_pseudogene", "miRNA", "snoRNA", "snRNA", "processed_transcript", "misc_RNA") -->
<!-- noncoding.sel = lapply(noncoding.names, -->
<!--                        function(x) -->
<!--                           gff[gene_type%in%as.character(x), gene_id]) -->
<!-- names(noncoding.sel) = noncoding.names -->
<!-- ``` -->

<!-- ### Promoter regions -->

<!-- The promoter of a gene is usually comprised within the 1000 nucleotide upstream of its transcriptional start site. -->
<!-- Therefore, these regions upstream of protein coding or lincRNA genes are added to the analysis. -->
<!-- However, if the promoter encompasses another annotated gene on either strand, the overlapping portion is removed. -->
<!-- Finally, promoters enclosed in a larger feature or reduced to less than 200nt were removed from the analysis. -->

<!-- ```{r generatePromoters} -->
<!-- genes = with(gff, GRanges(seqnames, IRanges(start, end), strand, gene_id, gene_feature, gene_type, gene_name, level)) -->
<!-- genes_prom = genes[genes$gene_type%in%c("protein_coding", "lincRNA")] -->
<!-- genes_prom = unlist(reduce(split(genes_prom, genes_prom$gene_id))) -->
<!-- genes_prom$gene_id = names(genes_prom) -->
<!-- names(genes_prom) = NULL -->
<!-- tab = unique(gff[, list(gene_id, gene_type, gene_name, level)]) -->
<!-- mcols(genes_prom) = gff[match(genes_prom$gene_id, gff[, gene_id]), list(gene_id, gene_type, gene_name, level)] -->

<!-- prom.gr = promoters(genes_prom, upstream=1e3, downstream=0) -->
<!-- overlaps = as.data.table(findOverlaps(prom.gr, c(prom.gr, genes_prom), ignore.strand=TRUE, minoverlap=1)) -->
<!-- overlaps = overlaps[queryHits!=subjectHits] -->
<!-- overlaps = cbind(overlaps, -->
<!--                  data.table(as.data.frame(prom.gr[overlaps[, queryHits],])), -->
<!--                  data.table(as.data.frame(c(prom.gr, genes_prom)[overlaps[, subjectHits],]))[, list(st=start, en=end)]) -->
<!-- overlaps[, width := as.numeric(width)] -->
<!-- overlaps[start>=st & end<=en, width := as.numeric(NA)] -->
<!-- overlaps[start>=st & end>en,  width := as.numeric(ifelse(strand=="+", end-en, as.numeric(NA)))] -->
<!-- overlaps[start<st & end<=en,  width := as.numeric(ifelse(strand=="+", as.numeric(NA), st-start))] -->
<!-- overlaps[start<st & end>en,   width := as.numeric(ifelse(strand=="+", end-en, st-start))] -->

<!-- prom.gr = prom.gr[!prom.gr$gene_id%in%overlaps[, gene_id],] -->
<!-- temp.gr = with(overlaps[!is.na(width) & width>=200,], -->
<!--                GRanges(seqnames, IRanges(start=ifelse(strand=="+", end-width+1, start), -->
<!--                                          end=ifelse(strand=="+", end, start+width-1)),  -->
<!--                        strand, gene_id, gene_type, gene_name, level)) -->
<!-- prom.gr = sort.GenomicRanges(c(prom.gr, temp.gr), ignore.strand=TRUE) -->
<!-- prom.gr$gene_feature = "promoter" -->

<!-- prom.gr = setcolorder(as.data.table(prom.gr), c(1:6, 10, 7:9)) -->
<!-- prom.gr = with(prom.gr, GRanges(seqnames, IRanges(start, end), strand, gene_id, gene_feature, gene_type, gene_name, level)) -->

<!-- genes = sort.GenomicRanges(c(genes, prom.gr), ignore.strand=TRUE) -->

<!-- genes.dt = data.table(as.data.frame(genes), index=1:length(genes), key="index") -->
<!-- setkey(genes.dt, index) -->

<!-- genes.ext = genes -->
<!-- genes.ext.dt = genes.dt -->

<!-- genes = unlist(reduce(split(genes, genes$gene_id))) -->
<!-- mcols(genes) = tab[match(names(genes), gene_id),] -->
<!-- names(genes) = NULL -->
<!-- genes = sort.GenomicRanges(genes, ignore.strand=TRUE) -->
<!-- genes.dt = data.table(as.data.frame(genes), index=seq_along(genes), key="index") -->
<!-- ``` -->

## RADICL-seq analyses (RNA and DNA genes)

```{r readTable}
sampleTable = fread("RNA-DNA_Hiseq_runs.txt",
                    header=FALSE, col.names=c("Experiment", "Condition", "Barcode", "Replicate"))
sampleTable[, Label := paste(Experiment, Barcode, sep="_")]
setkey(sampleTable, Label)

tab0 = copy(sampleTable)
tab0[, c("Condition", "BiolReplicate") := tstrsplit(Condition, "\\.")][, BiolReplicate := as.numeric(BiolReplicate)]
setnames(tab0, "Replicate", "TechReplicate")
setcolorder(tab0, c(1,2,6,4,3,5))
```

```{r loadNewAnnotation}
if(!exists("gencode_version"))
   gencode_version = "m14"

load(paste0("GencodeReference/mm10_Gencode", toupper(gencode_version), "_annotations.RData"))

genes = sort.GenomicRanges(with(gene.metadata,
                                GRanges(seqnames, IRanges(start, end), strand, gene_id, gene_type, gene_name, level)),
                           ignore.strand=TRUE)
genes.dt = data.table(as.data.frame(genes), index=seq_along(genes), key="index")

tx.regions = readRDS(paste0("GencodeReference/mm10_Gencode", toupper(gencode_version), "_annotations.all.genes.regions.rds"))
tx.regions.dt = data.table(as.data.frame(tx.regions), index=1:length(tx.regions), key="index")
```

### Read files

The _read.paired.bed()_ function is here employed to read the files containing the paired RNA-DNA reads. For each BED file, the corresponding gene names are assigned to the RNA and DNA fragments.
Pairs for which it was not possible to assign 1) either of the framents to a unique position on the genome (i.e., BWA score != 37) or 2) both fragments to annotated genes are discarded.
Cases were only one of the fragments was mapped to an annotated features were marked _intergenic_, while fragments mapping to more than one gene, which could lead to double counting or misassignment, were marked _multigenic_.

### Change annotation (M4 --> M14)

```{r changeAnnotation, eval=FALSE}
files = list.files("./RNA-DNA", recursive=TRUE, full.names=TRUE, pattern="*.gz$")
identifiers = paste(tstrsplit(files, "\\.|_|/")[[4]], tstrsplit(files, "\\.|_|/")[[5]], tstrsplit(files, "\\.|_|/")[[7]], sep="_")

interactions = list()
for( i in seq_along(files) ){
   file = files[i]
   bed = read.paired.bed(file, min.mapq=37)
   
   # RNA
   overlaps.rna = as.data.table(findOverlaps(bed[[1]], tx.regions, ignore.strand=FALSE))
   setkey(overlaps.rna, subjectHits)
   overlaps.rna = overlaps.rna[tx.regions.dt[, list(index, gene_id.R=gene_id, annotation.R=annotation, gene_type.R=gene_type, region.R=type)], nomatch=0][, subjectHits := NULL]
   setkey(overlaps.rna, queryHits)
   # Multigenic
   overlaps.rna[queryHits%in%overlaps.rna[, .N, by="queryHits"][N>1, queryHits], `:=`(gene_id.R="Multigenic", annotation.R="Multigenic", gene_type.R="Multigenic", region.R="Multigenic")]
   overlaps.rna = unique(overlaps.rna)
   # Intergenic
   missing.rna = data.table(queryHits=setdiff(1:length(bed[[1]]), overlaps.rna[, queryHits]))
   sel = overlapsAny(bed[[1]][missing.rna[, queryHits],], tx.regions)
   missing.rna[sel,  `:=`(gene_id.R="Genic_unassigned", annotation.R="Genic_unassigned", gene_type.R="Genic_unassigned", region.R="Genic_unassigned")]
   missing.rna[!sel, `:=`(gene_id.R="Intergenic", annotation.R="Intergenic", gene_type.R="Intergenic", region.R="Intergenic")]

   overlaps.rna = rbindlist(list(overlaps.rna, missing.rna))
   setkey(overlaps.rna, queryHits)
   
   # DNA
   overlaps.dna = as.data.table(findOverlaps(bed[[2]], tx.regions, ignore.strand=TRUE))
   setkey(overlaps.dna, subjectHits)
   overlaps.dna = overlaps.dna[tx.regions.dt[, list(index, gene_id.D=gene_id, annotation.D=annotation, gene_type.D=gene_type, region.D=type)], nomatch=0][, subjectHits := NULL]
   setkey(overlaps.dna, queryHits)
   # Multigenic
   overlaps.dna[queryHits%in%overlaps.dna[, .N, by="queryHits"][N>1, queryHits], `:=`(gene_id.D="Multigenic", annotation.D="Multigenic", gene_type.D="Multigenic", region.D="Multigenic")]
   overlaps.dna = unique(overlaps.dna)
   # Intergenic
   missing.dna = data.table(queryHits=setdiff(1:length(bed[[2]]), overlaps.dna[, queryHits]))
   sel = overlapsAny(bed[[2]][missing.dna[, queryHits],], tx.regions)
   missing.dna[sel,  `:=`(gene_id.D="Genic_unassigned", annotation.D="Genic_unassigned", gene_type.D="Genic_unassigned", region.D="Genic_unassigned")]
   missing.dna[!sel, `:=`(gene_id.D="Intergenic", annotation.D="Intergenic", gene_type.D="Intergenic", region.D="Intergenic")]
   
   overlaps.dna = rbindlist(list(overlaps.dna, missing.dna))
   setkey(overlaps.dna, queryHits)
   
   bed = lapply(bed, function(x) data.table(as.data.frame(x), key="index"))
   
   overlaps.rna = overlaps.rna[bed[[1]][, list(index, chr.R=seqnames, start.R=start, end.R=end, strand.R=strand)], nomatch=0]
   setcolorder(overlaps.rna, c(1, 6:9,  2:5))
   overlaps.dna = overlaps.dna[bed[[2]][, list(index, chr.D=seqnames, start.D=start, end.D=end)], nomatch=0]
   setcolorder(overlaps.dna, c(1, 6:8,  2:5))
   
   rm(bed); gc()

   overlaps = overlaps.rna[overlaps.dna, nomatch=0][, queryHits := NULL]
   overlaps[, `:=`(region.R = droplevels(region.R), region.D = droplevels(region.D))]
   
   rm(overlaps.rna, overlaps.dna); gc()
   
   # writeGzip(overlaps,
   #           file=paste0("biotype/RNA-DNA_", identifiers[i], "_biotypes_", gencode_version, "_level_1_2_3_AllOverlapsUnique.tsv"),
   #           col.names=TRUE, showProgress=TRUE, wait=FALSE, level=9)

   interactions[[i]] = overlaps

   rm(overlaps)
   gc()
   
}

names(interactions) = identifiers

# Concatenate the list
interactions = rbindlist(interactions, idcol="name")

setkey(interactions, name)
interactions = interactions[tab0[, list(Label, Condition, Replicate=BiolReplicate)], nomatch=0]
interactions[, name := NULL]

for( cond in unique(interactions[, Condition]) )
   for( repl in unique(interactions[Condition%in%cond, Replicate]) )
      writeGzip(interactions[Condition%in%cond & Replicate%in%repl,],
                file=paste0("RNA-DNA_", cond, "_", repl, "_biotypes_", gencode_version, "_level_1_2_3_AllOverlapsUnique.tsv"),
                col.names=TRUE, showProgress=TRUE, wait=FALSE, level=9)

writeGzip(interactions,
          file=paste0("RNA-DNA_biotypes_", gencode_version, "_level_1_2_3_AllOverlapsUnique.tsv"),
          col.names=TRUE, showProgress=TRUE, wait=FALSE, level=9)

tab.R = interactions[, .N, keyby=c("Condition", "annotation.R", "region.R")]
tab.D = interactions[, .N, keyby=c("Condition", "annotation.D", "region.D")]

save(tab.R, tab.D, file=paste0("RData/RNA-DNA_biotypes_", gencode_version, "_level_1_2_3_AllOverlapsUnique_stats.RData"))

writeGzip(interactions[!(gene_id.R%in%c("Multigenic", "Intergenic") | gene_id.D%in%c("Multigenic", "Intergenic"))],
          file=paste0("RNA-DNA_biotypes_", gencode_version, "_level_1_2_3_AllOverlapsUnique_noIntergenic.tsv"),
          col.names=TRUE, showProgress=TRUE, wait=FALSE, level=9)
```

```{r piechart, eval=FALSE}
load(paste0("RData/RNA-DNA_biotypes_", gencode_version, "_level_1_2_3_AllOverlapsUnique_stats.RData"))

for( cond in unique(tab.R[, Condition]) ){
   
   tab.1 = tab.R[Condition%in%cond,]
   tab.1[, Genomic := factor(ifelse(annotation.R%in%"Intergenic", "Intergenic", "Genic"), levels=c("Genic", "Intergenic"))]
   tab.1[, Genic := region.R]
   tab.1[annotation.R%in%"other", Genic := "other"]
   tab.1[annotation.R%in%"long_ncRNA", Genic := paste(annotation.R, region.R, sep="_")]
   tab.1[annotation.R%in%"ncRNA", Genic := "ncRNA"]
   tab.1[annotation.R%in%"protein_coding", Genic := paste("coding", region.R, sep="_")]
   tab.1[Genic%in%"coding_CDS", Genic := "CDS"][Genic%in%"coding_five_prime_UTR", Genic := "UTR5"][Genic%in%"coding_three_prime_UTR", Genic := "UTR3"]
   tab.1[, Genic := factor(Genic, levels=c("UTR5", "CDS", "coding_exon", "coding_intron", "UTR3", "long_ncRNA_exon", "long_ncRNA_intron", "ncRNA", "other", "Intergenic"))]
   
   tab = rbindlist(list(`1`=tab.1[, list(N=sum(N)), by="Genomic"], `2`=tab.1[, list(N=sum(N)), by="Genic"]), idcol="Group")
   tab[, Genomic := factor(Genomic, levels=c("Genic", "UTR5", "CDS", "coding_exon", "coding_intron", "UTR3", "long_ncRNA_exon", "long_ncRNA_intron", "ncRNA", "other", "Intergenic"))]
   tab = tab[order(Genomic)][, list(Genomic, N, value=N/sum(N), percentage=round(N/sum(N)*100, 1), position=cumsum(N/sum(N)) - 0.5*(N/sum(N))), by="Group"]
   tab[, Group := as.numeric(Group)]
   
   gg = ggplot(tab, aes(x=Group, y=value, fill=Genomic)) +
      geom_bar(position="stack", stat="identity", col="black", width=0.7) +
      scale_fill_ptol("") +
      coord_polar(theta="y", direction=-1) +
      theme_minimal() +
      geom_text(aes(x = Group+0.5, y = 1-position, label = paste0(percentage, "%")), col="black", size = 4) +
      theme(axis.text = element_blank(), axis.ticks = element_blank(), panel.grid  = element_blank(),
            axis.text.y = element_blank(), axis.text.x = element_blank(), axis.title=element_blank(),
            panel.background = element_blank(), strip.background = element_blank(), strip.text.x = element_text(size = 14),
            strip.text.y = element_text(size = 14), legend.title=element_text(size = 14), legend.text=element_text(size = 14))
   ggsave(gg, filename=paste0("plots/Figure2A_", cond, ".pdf"), width=12, height=8)
   
   tab.2 = tab.D[Condition%in%cond,]
   tab.2[, Genomic := annotation.D]
   tab.2[!Genomic%in%c("Intergenic", "Multigenic"), Genomic := "Genic"]
   tab.2[Genomic%in%"Multigenic", Genomic := "Genic_unassigned"]
   tab.2[, Genic := region.D]
   tab.2[annotation.D%in%"Multigenic", Genic := "Genic_unassigned"]
   tab.2[annotation.D%in%"other", Genic := "other"]
   tab.2[annotation.D%in%"long_ncRNA", Genic := paste(annotation.D, region.D, sep="_")]
   tab.2[annotation.D%in%"ncRNA", Genic := "ncRNA"]
   tab.2[annotation.D%in%"protein_coding", Genic := paste("coding", region.D, sep="_")]
   tab.2[Genic%in%"coding_CDS", Genic := "CDS"][Genic%in%"coding_five_prime_UTR", Genic := "UTR5"][Genic%in%"coding_three_prime_UTR", Genic := "UTR3"]
   
   tab = rbindlist(list(`1`=tab.2[, list(N=sum(N)), by="Genomic"], `2`=tab.2[, list(N=sum(N)), by="Genic"]), idcol="Group")
   tab[, Genomic := factor(Genomic, levels=c("Genic", "UTR5", "CDS", "coding_exon", "coding_intron", "UTR3", "long_ncRNA_exon", "long_ncRNA_intron", "ncRNA", "other", "Genic_unassigned", "Intergenic"))]
   tab = tab[order(Genomic)][, list(Genomic, N, value=N/sum(N), percentage=round(N/sum(N)*100, 1), position=cumsum(N/sum(N)) - 0.5*(N/sum(N))), by="Group"]
   tab[, Group := as.numeric(Group)]
   
   gg = ggplot(tab, aes(x=Group, y=value, fill=Genomic)) +
      geom_bar(position="stack", stat="identity", col="black", width=0.7) +
      scale_fill_ptol("") +
      coord_polar(theta="y", direction=-1) +
      theme_minimal() +
      geom_text(aes(x = Group+0.5, y = 1-position, label = paste0(percentage, "%")), col="black", size = 4) +
      theme(axis.text = element_blank(), axis.ticks = element_blank(), panel.grid  = element_blank(),
            axis.text.y = element_blank(), axis.text.x = element_blank(), axis.title=element_blank(),
            panel.background = element_blank(), strip.background = element_blank(), strip.text.x = element_text(size = 14),
            strip.text.y = element_text(size = 14), legend.title=element_text(size = 14), legend.text=element_text(size = 14))
   ggsave(gg, filename=paste0("plots/Figure2B_", cond, ".pdf"), width=12, height=8)
}
```

```{r changeTrimmed, eval=FALSE}
files = list.files("./Trimmed", recursive=TRUE, full.names=TRUE, pattern="*.gz$")
identifiers = paste(tstrsplit(files, "\\.|_|/")[[4]], tstrsplit(files, "\\.|_|/")[[5]], tstrsplit(files, "\\.|_|/")[[7]], sep="_")

interactions = list()
for( i in seq_along(files) ){
   file = files[i]
   bed = read.paired.bed(file, min.mapq=37)
   
   # RNA
   overlaps.rna = as.data.table(findOverlaps(bed[[1]], tx.regions, ignore.strand=FALSE))
   setkey(overlaps.rna, subjectHits)
   overlaps.rna = overlaps.rna[tx.regions.dt[, list(index, gene_id.R=gene_id, annotation.R=annotation, gene_type.R=gene_type, region.R=type)], nomatch=0][, subjectHits := NULL]
   setkey(overlaps.rna, queryHits)
   # Multigenic
   overlaps.rna[queryHits%in%overlaps.rna[, .N, by="queryHits"][N>1, queryHits], `:=`(gene_id.R="Multigenic", annotation.R="Multigenic", gene_type.R="Multigenic", region.R="Multigenic")]
   overlaps.rna = unique(overlaps.rna)
   # Intergenic
   missing.rna = data.table(queryHits=setdiff(1:length(bed[[1]]), overlaps.rna[, queryHits]))
   sel = overlapsAny(bed[[1]][missing.rna[, queryHits],], tx.regions)
   missing.rna[sel,  `:=`(gene_id.R="Genic_unassigned", annotation.R="Genic_unassigned", gene_type.R="Genic_unassigned", region.R="Genic_unassigned")]
   missing.rna[!sel, `:=`(gene_id.R="Intergenic", annotation.R="Intergenic", gene_type.R="Intergenic", region.R="Intergenic")]

   overlaps.rna = rbindlist(list(overlaps.rna, missing.rna))
   setkey(overlaps.rna, queryHits)
   
   # DNA
   overlaps.dna = as.data.table(findOverlaps(bed[[2]], tx.regions, ignore.strand=TRUE))
   setkey(overlaps.dna, subjectHits)
   overlaps.dna = overlaps.dna[tx.regions.dt[, list(index, gene_id.D=gene_id, annotation.D=annotation, gene_type.D=gene_type, region.D=type)], nomatch=0][, subjectHits := NULL]
   setkey(overlaps.dna, queryHits)
   # Multigenic
   overlaps.dna[queryHits%in%overlaps.dna[, .N, by="queryHits"][N>1, queryHits], `:=`(gene_id.D="Multigenic", annotation.D="Multigenic", gene_type.D="Multigenic", region.D="Multigenic")]
   overlaps.dna = unique(overlaps.dna)
   # Intergenic
   missing.dna = data.table(queryHits=setdiff(1:length(bed[[2]]), overlaps.dna[, queryHits]))
   sel = overlapsAny(bed[[2]][missing.dna[, queryHits],], tx.regions)
   missing.dna[sel,  `:=`(gene_id.D="Genic_unassigned", annotation.D="Genic_unassigned", gene_type.D="Genic_unassigned", region.D="Genic_unassigned")]
   missing.dna[!sel, `:=`(gene_id.D="Intergenic", annotation.D="Intergenic", gene_type.D="Intergenic", region.D="Intergenic")]
   
   overlaps.dna = rbindlist(list(overlaps.dna, missing.dna))
   setkey(overlaps.dna, queryHits)
   
   bed = lapply(bed, function(x) data.table(as.data.frame(x), key="index"))
   
   overlaps.rna = overlaps.rna[bed[[1]][, list(index, chr.R=seqnames, start.R=start, end.R=end, strand.R=strand)], nomatch=0]
   setcolorder(overlaps.rna, c(1, 6:9,  2:5))
   overlaps.dna = overlaps.dna[bed[[2]][, list(index, chr.D=seqnames, start.D=start, end.D=end)], nomatch=0]
   setcolorder(overlaps.dna, c(1, 6:8,  2:5))
   
   rm(bed); gc()

   overlaps = overlaps.rna[overlaps.dna, nomatch=0][, queryHits := NULL]
   overlaps[, `:=`(region.R = droplevels(region.R), region.D = droplevels(region.D))]
   
   rm(overlaps.rna, overlaps.dna); gc()
   
   # writeGzip(overlaps,
   #           file=paste0("biotype/RNA-DNA_", identifiers[i], "_biotypes_", gencode_version, "_level_1_2_3_AllOverlapsUnique.tsv"),
   #           col.names=TRUE, showProgress=TRUE, wait=FALSE, level=9)

   interactions[[i]] = overlaps

   rm(overlaps)
   gc()
   
}

names(interactions) = identifiers

# Concatenate the list
interactions = rbindlist(interactions, idcol="name")

interactions[, c("Condition", "Replicate"):=tstrsplit(tstrsplit(name, "_")[[1]], "n")]
interactions[, name := NULL]

writeGzip(interactions,
          file=paste0("RNA-DNA_trimmed_", gencode_version, "_level_1_2_3_AllOverlapsUnique.tsv"),
          col.names=TRUE, showProgress=TRUE, wait=FALSE, level=9)

tab.R = interactions[, .N, keyby=c("Condition", "annotation.R", "region.R")]
tab.D = interactions[, .N, keyby=c("Condition", "annotation.D", "region.D")]

save(tab.R, tab.D, file=paste0("RData/RNA-DNA_trimmed_", gencode_version, "_level_1_2_3_AllOverlapsUnique_stats.RData"))
```

```{r piechartTrimmed, eval=FALSE}
load(paste0("RData/RNA-DNA_trimmed_", gencode_version, "_level_1_2_3_AllOverlapsUnique_stats.RData"))

for( cond in unique(tab.R[, Condition]) ){
   
   tab.1 = tab.R[Condition%in%cond,]
   tab.1[, Genomic := factor(ifelse(annotation.R%in%"Intergenic", "Intergenic", "Genic"), levels=c("Genic", "Intergenic"))]
   tab.1[, Genic := region.R]
   tab.1[annotation.R%in%"other", Genic := "other"]
   tab.1[annotation.R%in%"long_ncRNA", Genic := paste(annotation.R, region.R, sep="_")]
   tab.1[annotation.R%in%"ncRNA", Genic := "ncRNA"]
   tab.1[annotation.R%in%"protein_coding", Genic := paste("coding", region.R, sep="_")]
   tab.1[Genic%in%"coding_CDS", Genic := "CDS"][Genic%in%"coding_five_prime_UTR", Genic := "UTR5"][Genic%in%"coding_three_prime_UTR", Genic := "UTR3"]
   tab.1[, Genic := factor(Genic, levels=c("UTR5", "CDS", "coding_exon", "coding_intron", "UTR3", "long_ncRNA_exon", "long_ncRNA_intron", "ncRNA", "other", "Intergenic"))]
   
   tab = rbindlist(list(`1`=tab.1[, list(N=sum(N)), by="Genomic"], `2`=tab.1[, list(N=sum(N)), by="Genic"]), idcol="Group")
   tab[, Genomic := factor(Genomic, levels=c("Genic", "UTR5", "CDS", "coding_exon", "coding_intron", "UTR3", "long_ncRNA_exon", "long_ncRNA_intron", "ncRNA", "other", "Intergenic"))]
   tab = tab[order(Genomic)][, list(Genomic, N, value=N/sum(N), percentage=round(N/sum(N)*100, 1), position=cumsum(N/sum(N)) - 0.5*(N/sum(N))), by="Group"]
   tab[, Group := as.numeric(Group)]
   
   gg = ggplot(tab, aes(x=Group, y=value, fill=Genomic)) +
      geom_bar(position="stack", stat="identity", col="black", width=0.7) +
      scale_fill_ptol("") +
      coord_polar(theta="y", direction=-1) +
      theme_minimal() +
      geom_text(aes(x = Group+0.5, y = 1-position, label = paste0(percentage, "%")), col="black", size = 4) +
      theme(axis.text = element_blank(), axis.ticks = element_blank(), panel.grid  = element_blank(),
            axis.text.y = element_blank(), axis.text.x = element_blank(), axis.title=element_blank(),
            panel.background = element_blank(), strip.background = element_blank(), strip.text.x = element_text(size = 14),
            strip.text.y = element_text(size = 14), legend.title=element_text(size = 14), legend.text=element_text(size = 14))
   ggsave(gg, filename=paste0("plots/Figure2A_trimmed", cond, ".pdf"), width=12, height=8)
   
   tab.2 = tab.D[Condition%in%cond,]
   tab.2[, Genomic := annotation.D]
   tab.2[!Genomic%in%c("Intergenic", "Multigenic"), Genomic := "Genic"]
   tab.2[Genomic%in%"Multigenic", Genomic := "Genic_unassigned"]
   tab.2[, Genic := region.D]
   tab.2[annotation.D%in%"Multigenic", Genic := "Genic_unassigned"]
   tab.2[annotation.D%in%"other", Genic := "other"]
   tab.2[annotation.D%in%"long_ncRNA", Genic := paste(annotation.D, region.D, sep="_")]
   tab.2[annotation.D%in%"ncRNA", Genic := "ncRNA"]
   tab.2[annotation.D%in%"protein_coding", Genic := paste("coding", region.D, sep="_")]
   tab.2[Genic%in%"coding_CDS", Genic := "CDS"][Genic%in%"coding_five_prime_UTR", Genic := "UTR5"][Genic%in%"coding_three_prime_UTR", Genic := "UTR3"]
   
   tab = rbindlist(list(`1`=tab.2[, list(N=sum(N)), by="Genomic"], `2`=tab.2[, list(N=sum(N)), by="Genic"]), idcol="Group")
   tab[, Genomic := factor(Genomic, levels=c("Genic", "UTR5", "CDS", "coding_exon", "coding_intron", "UTR3", "long_ncRNA_exon", "long_ncRNA_intron", "ncRNA", "other", "Genic_unassigned", "Intergenic"))]
   tab = tab[order(Genomic)][, list(Genomic, N, value=N/sum(N), percentage=round(N/sum(N)*100, 1), position=cumsum(N/sum(N)) - 0.5*(N/sum(N))), by="Group"]
   tab[, Group := as.numeric(Group)]
   
   gg = ggplot(tab, aes(x=Group, y=value, fill=Genomic)) +
      geom_bar(position="stack", stat="identity", col="black", width=0.7) +
      scale_fill_ptol("") +
      coord_polar(theta="y", direction=-1) +
      theme_minimal() +
      geom_text(aes(x = Group+0.5, y = 1-position, label = paste0(percentage, "%")), col="black", size = 4) +
      theme(axis.text = element_blank(), axis.ticks = element_blank(), panel.grid  = element_blank(),
            axis.text.y = element_blank(), axis.text.x = element_blank(), axis.title=element_blank(),
            panel.background = element_blank(), strip.background = element_blank(), strip.text.x = element_text(size = 14),
            strip.text.y = element_text(size = 14), legend.title=element_text(size = 14), legend.text=element_text(size = 14))
   ggsave(gg, filename=paste0("plots/Figure2B_trimmed", cond, ".pdf"), width=12, height=8)
}
```

```{r setFDR, echo=FALSE}
if(!exists("fdr.th")) fdr.th = 0.05
```

The FDR is set to `r fdr.th`.

```{r significance, eval=FALSE}
if(!exists("interactions"))
   interactions = fread(paste0("gunzip -c RNA-DNA_biotypes_", gencode_version, "_level_1_2_3_AllOverlapsUnique.tsv.gz"))

# Calculate the count per interaction pair
setkeyv(interactions, c("Condition", "gene_id.R", "gene_id.D"))
subset = interactions[, .N, by=key(interactions)]
setnames(subset, "N", "Count")

# subset = subset[!(gene_id.R%in%c("Multigenic", "Intergenic") | gene_id.D%in%c("Multigenic", "Intergenic"))]

# # Add the correction for the distance (ignore the multi- and inter-genic, they will be removed downstream anyway)
# pairs.unique = unique(subset[!(gene_id.R%in%c("Multigenic", "Intergenic") | gene_id.D%in%c("Multigenic", "Intergenic")), list(gene_id.R, gene_id.D)])
# # pairs.unique = unique(subset[, list(gene_id.R, gene_id.D)])
# pairs.unique[, distance := distance(genes[match(pairs.unique[, gene_id.R], genes$gene_id)],
#                                     genes[match(pairs.unique[, gene_id.D], genes$gene_id)],
#                                     ignore.strand=TRUE)]
# # Calculate the distance distribution and the weight factors
# h = hist(pairs.unique[distance>0 & distance<=1e6, distance], 100, plot=FALSE)
# h = data.table(breaks=h$breaks[-1], count=log10(h$counts))
# h[, index := seq_along(breaks)][, scaled := (count-min(count))/(2*(max(count)-min(count)))+0.5]
# setkey(h, index)
# 
# # Set a large number (it will not be used) for the interactions with no distance info
# pairs.unique[is.na(distance), distance := 7e6]
# # Assign the weight factors
# pairs.unique[, interval := findInterval(distance, c(0, h$breaks))]
# setkey(pairs.unique, interval)
# # Interactions farther than 1e6 bp (or interchromosomal) will have the lowest weight
# pairs.unique[, distance := as.double(distance)][distance>1e6, distance := min(h[,scaled])]
# pairs.unique = merge(pairs.unique, h, by.x="interval", by.y="index", all.x=TRUE)
# pairs.unique[!is.na(scaled), distance := scaled]
# setkeyv(pairs.unique, c("gene_id.R", "gene_id.D"))

frandom=1

# Calculate the frequency per RNA
setkeyv(subset, c("Condition", "gene_id.R"))
rj = subset[, list(coverage=sum(Count)), by=key(subset)]
rj[, relative := coverage/sum(coverage), by=c("Condition")]
setkeyv(rj, c("Condition", "gene_id.R"))
subset[rj, rj := relative]

# Calculate the frequency per DNA (or BIN)
setkeyv(subset, c("Condition", "gene_id.D"))
rh = subset[, list(coverage=sum(Count)), by=key(subset)]
rh[, relative := coverage/sum(coverage), by=c("Condition")]
setkeyv(rh, c("Condition", "gene_id.D"))
subset[rh, rh := relative]

# Calculate the total number of interactions (needed for the binomial test)
setkeyv(subset, c("Condition"))
subset[, N := sum(Count), by=key(subset)]

# Remove the Multi- and Inter-genic
subset = subset[!(gene_id.R%in%c("Multigenic", "Intergenic") | gene_id.D%in%c("Multigenic", "Intergenic"))]

# # Assign the distance weight to each interaction pair
# setkeyv(subset, c("gene_id.R", "gene_id.D"))
# subset[pairs.unique, Distance := distance]
# 
# # Free some memory
# rm(pairs.unique)
gc()

# Calculate the conditional (and weighted) probability
# subset[, `:=`(pjh = rj*rh*frandom*Distance)][, `:=`(Distance = NULL, rj = NULL, rh = NULL)]
subset[, `:=`(pjh = rj*rh*frandom)][, `:=`(rj = NULL, rh = NULL)]

# Perform the binomial test
setkeyv(subset, c("Condition", "gene_id.R", "gene_id.D"))
subset[, `:=`(pval=binom.test(Count, N, p=pjh, alternative="greater")$p.value), by=key(subset)]

# Perform the binomial test
setkeyv(subset, c("Condition", "gene_id.R", "gene_id.D"))
subset[, `:=`(pval=binom.test(Count, N, p=pjh, alternative="greater")$p.value), by=key(subset)]

# Perform the correction (Benjamini-Hochberg and Benjamini-Yekutieli) per transcript
# Calculate the total number of interacting bins (whether with genic or intergenic regions)
unique.bins = length(subset[, unique(gene_id.D)])
setkeyv(subset, c("Condition", "gene_id.R"))
subset[, fdr.tsc := p.adjust(pval, method="BH", n=unique.bins), by=key(subset)]
# Perform the correction (Benjamini-Hochberg and Benjamini-Yekutieli) per condition
# Calculate the total number of interacting bins (whether with genic or intergenic regions)
total.inter = as.numeric(length(subset[, unique(gene_id.R)]))*as.numeric(length(subset[, unique(gene_id.D)]))
setkey(subset, Condition)
subset[, fdr.cnd := p.adjust(pval, method="BH", n=total.inter), by=key(subset)]

# Calculate the observed vs expected ratio
subset[, `:=`(Rjh = log2(Count/(pjh*N)))]

# Remove some columns
subset[, `:=`(pjh = NULL, N = NULL)]

# save the entire object
saveRDS(subset,
        file=paste0("RData/RNA-DNA_counts_", gencode_version, "_level_1_2_3_AllOverlapsUnique_significant_noDistance.rds"))

setkeyv(subset, c("Condition", "gene_id.R", "gene_id.D"))
setkeyv(interactions, key(subset))

output = interactions[subset[, list(Condition, gene_id.R, gene_id.D, Pvalue=pval, FDR=fdr.tsc)], nomatch=0]

writeGzip(output[FDR<fdr.th,], file=paste0("RNA-DNA_counts_", gencode_version, "_level_1_2_3_AllOverlapsUnique_significant", fdr.th, "_extended_tsc_noDistance.tsv"),
          col.names=TRUE, showProgress=TRUE, wait=FALSE, level=9)

output = interactions[subset[, list(Condition, gene_id.R, gene_id.D, Pvalue=pval, FDR=fdr.cnd)], nomatch=0]

writeGzip(output[FDR<fdr.th,], file=paste0("RNA-DNA_counts_", gencode_version, "_level_1_2_3_AllOverlapsUnique_significant", fdr.th, "_extended_cnd_noDistance.tsv"),
          col.names=TRUE, showProgress=TRUE, wait=FALSE, level=9)

rm(interactions, output); gc()
```

```{r, statsDatasets, eval=FALSE}
##### Raw #####
interactions = fread(paste0("gunzip -c RNA-DNA_biotypes_", gencode_version, "_level_1_2_3_AllOverlapsUnique_noIntergenic.tsv.gz"))

setkeyv(interactions, c("Condition", "chr.R", "gene_id.R", "chr.D", "gene_id.D"))
subset = interactions[, list(Count=.N), by=key(interactions)]

stats1 = subset[, list(Unique_Interactions=.N, Cis=nrow(.SD[chr.R==chr.D,])/.N*100, Trans=nrow(.SD[chr.R!=chr.D,])/.N*100), by="Condition"]

fwrite(stats1, file=paste0("stats/RNA-DNA_CisTransUniqueInteractions_", gencode_version, "_level_1_2_3_AllOverlapsUnique.tsv"), quote=FALSE, sep='\t', row.names=FALSE)

setkeyv(interactions, c("Condition", "Replicate", "chr.R", "gene_id.R", "chr.D", "gene_id.D", "annotation.R", "gene_type.R"))
interactions = interactions[, list(Count=.N), by=key(interactions)]

setkeyv(interactions, c("Condition", "chr.R", "gene_id.R", "chr.D", "gene_id.D", "annotation.R", "gene_type.R"))
subset = interactions[, list(Count=sum(Count)), by=key(interactions)]

##### Filtered #####
# filtered = fread(paste0("gunzip -c RNA-BIN_counts_", gencode_version, "_level_1_2_3_AllOverlapsUnique_", bin_width, "_filtered_extended.tsv.gz"))
# 
# setkeyv(filtered, c("Condition", "chr.R", "gene_id.R", "chr.D", "gene_id.D"))
# filtered = filtered[, list(Count=.N), by=key(filtered)]
# 
# stats4 = filtered[, list(Unique_Interactions=.N, Cis=nrow(.SD[chr.R==chr.D,])/.N*100, Trans=nrow(.SD[chr.R!=chr.D,])/.N*100), by="Condition"]
# 
# fwrite(stats4, file=paste0("stats/RNA-BIN_CisTransUniqueInteractionsFiltered_", gencode_version, "_level_1_2_3_AllOverlapsUnique_", bin_width, ".tsv"), quote=FALSE, sep='\t', row.names=FALSE)
# 
# filtered = fread(paste0("gunzip -c RNA-BIN_counts_", gencode_version, "_level_1_2_3_AllOverlapsUnique_", bin_width, "_filtered.tsv.gz"))
# 
# setkeyv(filtered, c("Condition", "gene_id.R", "gene_id.D"))
# filtered = filtered[, list(Count=sum(N)), by=key(filtered)]
# 
# subset[filtered, Set := "Filtered"]
# 
# stats1 = subset[, list(filtered=length(unique(.SD[Set%in%"Filtered", gene_id.R])), total=length(unique(gene_id.R))), by=c("Condition", "biotype.R")] 
# stats1 = stats1[, Percentage := round(filtered/total*100, 2)][order(Condition, -Percentage)]
# 
# fwrite(stats1, file=paste0("stats/RNA-BIN_TranscriptBiotypesFiltered_", gencode_version, "_level_1_2_3_AllOverlapsUnique_", bin_width, ".tsv"), quote=FALSE, sep='\t', row.names=FALSE)
# 
# subset[, Bin := findInterval(log2(Count), seq(min(log2(Count)), round(max(log2(Count))), 0.5), left.open=TRUE)]
# 
# gg = ggplot(subset[Condition%in%"1FA", list(Count=sum(Count)), by=c("Condition", "Set", "Bin")], aes(x=Bin, y=Count, fill=Set)) +
#    ggtitle("Interactions distribution for RADICL-seq (1%FA)", subtitle="Raw (blue) and filtered (red) interactions") +
#    geom_bar(stat="identity", position="stack", col="black", alpha=0.75) +
#    scale_x_continuous("Interactions count per RNA-DNA pair (log2)") +
#    scale_fill_brewer("Dataset", palette="Set1") +
#    theme_bw() + facet_zoom(y = Set%in%"Filtered") +
#    theme(legend.position=c(0.9, 0.9), legend.box.background = element_rect(),
#                       legend.title=element_text(size=12)) 
# ggsave(gg, filename="plots/Figure4A.pdf", width=12, height=6)

##### Significant #####
significant = fread(paste0("gunzip -c RNA-DNA_counts_", gencode_version, "_level_1_2_3_AllOverlapsUnique_significant", fdr.th, "_extended_tsc_noDistance.tsv.gz"), showProgress=FALSE)

setkeyv(significant, key(subset))
subset[, Set := factor("Raw", levels=c("Raw", "Significant"))][significant, Set := "Significant"]

stats2 = subset[, list(significant=length(unique(.SD[Set%in%"Significant", gene_id.R])), total=length(unique(gene_id.R))), by=c("Condition", "annotation.R")] 
stats2 = stats2[, Percentage := round(significant/total*100, 2)][order(Condition, -Percentage)]

fwrite(stats2, file=paste0("stats/RNA-DNA_TranscriptBiotypesSignificant", fdr.th, "_", gencode_version, "_level_1_2_3_AllOverlapsUnique.tsv"), quote=FALSE, sep='\t', row.names=FALSE)

subset[, Bin := findInterval(log2(Count), seq(min(log2(Count)), round(max(log2(Count))), 0.5), left.open=TRUE)]

gg = ggplot(subset[Condition%in%"1FA", list(Count=sum(Count)), by=c("Condition", "Set", "Bin")], aes(x=Bin, y=Count, fill=Set)) +
   ggtitle("Interactions distribution for RADICL-seq (1%FA)", subtitle="Raw (blue) and significant (red) interactions") +
   geom_bar(stat="identity", position="stack", col="black", alpha=0.75) +
   scale_x_continuous("Interactions count per RNA-DNA pair (log2)") +
   scale_fill_brewer("Dataset", palette="Set1") +
   theme_bw() + facet_zoom(y = Set%in%"Significant") +
   theme(legend.position=c(0.9, 0.9), legend.box.background = element_rect(),
                      legend.title=element_text(size=12)) 
ggsave(gg, filename="plots/Figure3A_RNA-DNA.pdf", width=12, height=6)

tab = interactions[, list(Intra=sum(.SD[chr.R==chr.D, Count]), Total=sum(Count)), by=c("Condition", "Replicate")][, Perc := Intra/Total*100]
fwrite(tab, file=paste0("stats/RNA-DNA_CisTransInteractions_", gencode_version, "_level_1_2_3_AllOverlapsUnique.tsv"), quote=FALSE, sep='\t', row.names=FALSE)
tab = interactions[, list(Intra=sum(.SD[chr.R==chr.D, Count]), Total=sum(Count)), by=c("Condition")][, Perc := Intra/Total*100]
fwrite(tab, file=paste0("stats/RNA-DNA_CisTransInteractions_", gencode_version, "_level_1_2_3_AllOverlapsUnique_Condition.tsv"), quote=FALSE, sep='\t', row.names=FALSE)

setkeyv(significant, c("Condition", "Replicate", "chr.R", "gene_id.R", "chr.D", "gene_id.D", "annotation.R", "gene_type.R"))
significant = significant[, .N, by=key(significant)]
setnames(significant, "N", "Count")

tab2 = significant[, list(Intra=sum(.SD[chr.R==chr.D, Count]), Total=sum(Count)), by=c("Condition", "Replicate")][, Perc := Intra/Total*100]
fwrite(tab2, file=paste0("stats/RNA-DNA_CisTransInteractions_", gencode_version, "_level_1_2_3_AllOverlapsUnique_significant.tsv"), quote=FALSE, sep='\t', row.names=FALSE)
tab2 = significant[, list(Intra=sum(.SD[chr.R==chr.D, Count]), Total=sum(Count)), by=c("Condition")][, Perc := Intra/Total*100]
fwrite(tab2, file=paste0("stats/RNA-DNA_CisTransInteractions_", gencode_version, "_level_1_2_3_AllOverlapsUnique_Condition_significant.tsv"), quote=FALSE, sep='\t', row.names=FALSE)

setkeyv(significant, c("Condition", "gene_id.R", "annotation.R", "gene_type.R"))

tab3 = significant[, list(binNum=length(unique(gene_id.D)), binCount=sum(Count)), by=key(significant)][order(-binNum)]
fwrite(tab3, file=paste0("stats/RNA-DNA_TranscriptInteractionsBinNumber_", gencode_version, "_level_1_2_3_AllOverlapsUnique_significant.tsv"), quote=FALSE, sep='\t', row.names=FALSE)

tab4 = significant[, list(binNum=length(unique(gene_id.D)), binCount=sum(Count)), by=key(significant)][order(-binCount)]
fwrite(tab4, file=paste0("stats/RNA-DNA_TranscriptInteractionsBinCount_", gencode_version, "_level_1_2_3_AllOverlapsUnique_significant.tsv"), quote=FALSE, sep='\t', row.names=FALSE)
```

```{r changeAnnotationGRID, eval=FALSE}
files = grep("unique", list.files("./GRIDseq", recursive=TRUE, full.names=TRUE, pattern="*.gz$"), value=TRUE)
identifiers = paste(tstrsplit(files, "\\.|_|/")[[3]], gsub("rep", "", tstrsplit(files, "\\.|_|/")[[5]]), sep="_")

interactions = list()
for( i in seq_along(files) ){
   file = files[i]
   bed = read.paired.bed(file, min.mapq=37)
   
   # RNA
   overlaps.rna = as.data.table(findOverlaps(bed[[1]], tx.regions, ignore.strand=FALSE))
   setkey(overlaps.rna, subjectHits)
   overlaps.rna = overlaps.rna[tx.regions.dt[, list(index, gene_id.R=gene_id, annotation.R=annotation, gene_type.R=gene_type, region.R=type)], nomatch=0][, subjectHits := NULL]
   setkey(overlaps.rna, queryHits)
   # Multigenic
   overlaps.rna[queryHits%in%overlaps.rna[, .N, by="queryHits"][N>1, queryHits], `:=`(gene_id.R="Multigenic", annotation.R="Multigenic", gene_type.R="Multigenic", region.R="Multigenic")]
   overlaps.rna = unique(overlaps.rna)
   # Intergenic
   missing.rna = data.table(queryHits=setdiff(1:length(bed[[1]]), overlaps.rna[, queryHits]))
   sel = overlapsAny(bed[[1]][missing.rna[, queryHits],], tx.regions)
   missing.rna[sel,  `:=`(gene_id.R="Genic_unassigned", annotation.R="Genic_unassigned", gene_type.R="Genic_unassigned", region.R="Genic_unassigned")]
   missing.rna[!sel, `:=`(gene_id.R="Intergenic", annotation.R="Intergenic", gene_type.R="Intergenic", region.R="Intergenic")]

   overlaps.rna = rbindlist(list(overlaps.rna, missing.rna))
   setkey(overlaps.rna, queryHits)
   
   # DNA
   overlaps.dna = as.data.table(findOverlaps(bed[[2]], tx.regions, ignore.strand=TRUE))
   setkey(overlaps.dna, subjectHits)
   overlaps.dna = overlaps.dna[tx.regions.dt[, list(index, gene_id.D=gene_id, annotation.D=annotation, gene_type.D=gene_type, region.D=type)], nomatch=0][, subjectHits := NULL]
   setkey(overlaps.dna, queryHits)
   # Multigenic
   overlaps.dna[queryHits%in%overlaps.dna[, .N, by="queryHits"][N>1, queryHits], `:=`(gene_id.D="Multigenic", annotation.D="Multigenic", gene_type.D="Multigenic", region.D="Multigenic")]
   overlaps.dna = unique(overlaps.dna)
   # Intergenic
   missing.dna = data.table(queryHits=setdiff(1:length(bed[[2]]), overlaps.dna[, queryHits]))
   sel = overlapsAny(bed[[2]][missing.dna[, queryHits],], tx.regions)
   missing.dna[sel,  `:=`(gene_id.D="Genic_unassigned", annotation.D="Genic_unassigned", gene_type.D="Genic_unassigned", region.D="Genic_unassigned")]
   missing.dna[!sel, `:=`(gene_id.D="Intergenic", annotation.D="Intergenic", gene_type.D="Intergenic", region.D="Intergenic")]
   
   overlaps.dna = rbindlist(list(overlaps.dna, missing.dna))
   setkey(overlaps.dna, queryHits)
   
   bed = lapply(bed, function(x) data.table(as.data.frame(x), key="index"))
   
   overlaps.rna = overlaps.rna[bed[[1]][, list(index, chr.R=seqnames, start.R=start, end.R=end, strand.R=strand)], nomatch=0]
   setcolorder(overlaps.rna, c(1, 6:9,  2:5))
   overlaps.dna = overlaps.dna[bed[[2]][, list(index, chr.D=seqnames, start.D=start, end.D=end)], nomatch=0]
   setcolorder(overlaps.dna, c(1, 6:8,  2:5))
   
   rm(bed); gc()

   overlaps = overlaps.rna[overlaps.dna, nomatch=0][, queryHits := NULL]
   overlaps[, `:=`(region.R = droplevels(region.R), region.D = droplevels(region.D))]
   
   rm(overlaps.rna, overlaps.dna); gc()
   
   # writeGzip(overlaps,
   #           file=paste0("biotype/RNA-DNA_", identifiers[i], "_biotypes_", gencode_version, "_level_1_2_3_AllOverlapsUnique.tsv"),
   #           col.names=TRUE, showProgress=TRUE, wait=FALSE, level=9)

   interactions[[i]] = overlaps

   rm(overlaps)
   gc()
   
}

names(interactions) = identifiers

# Concatenate the list
interactions = rbindlist(interactions, idcol="name")
interactions[, c("Condition", "Replicate") := tstrsplit(name, "_")]
interactions[, name := NULL]

# Write all interactions
writeGzip(interactions,
          file=paste0("GRIDseq-DNA_biotypes_", gencode_version, "_level_1_2_3_AllOverlapsUnique_.tsv"),
          col.names=TRUE, showProgress=TRUE, wait=FALSE, level=9)

tab.R = interactions[, .N, keyby=c("annotation.R", "region.R")]
tab.D = interactions[, .N, keyby=c("annotation.D", "region.D")]

save(tab.R, tab.D, file=paste0("RData/GRIDseq-DNA_biotypes_", gencode_version, "_level_1_2_3_AllOverlapsUnique_stats.RData"))

# Write genic interactions
writeGzip(interactions[!(gene_id.R%in%c("Multigenic", "Intergenic") | gene_id.D%in%c("Multigenic", "Intergenic"))],
          file=paste0("GRIDseq-DNA_biotypes_", gencode_version, "_level_1_2_3_AllOverlapsUnique_noIntergenic.tsv"),
          col.names=TRUE, showProgress=TRUE, wait=FALSE, level=9)
```

```{r piechartGRID, eval=FALSE}
load(paste0("RData/GRIDseq-DNA_biotypes_", gencode_version, "_level_1_2_3_AllOverlapsUnique_stats.RData"))

tab.R[, Genomic := factor(ifelse(annotation.R%in%"Intergenic", "Intergenic", "Genic"), levels=c("Genic", "Intergenic"))]
tab.R[, Genic := region.R]
tab.R[annotation.R%in%"other", Genic := "other"]
tab.R[annotation.R%in%"long_ncRNA", Genic := paste(annotation.R, region.R, sep="_")]
tab.R[annotation.R%in%"ncRNA", Genic := "ncRNA"]
tab.R[annotation.R%in%"protein_coding", Genic := paste("coding", region.R, sep="_")]
tab.R[Genic%in%"coding_CDS", Genic := "CDS"][Genic%in%"coding_five_prime_UTR", Genic := "UTR5"][Genic%in%"coding_three_prime_UTR", Genic := "UTR3"]
tab.R[Genic%in%"intron", Genic := "other"]
tab.R[, Genic := factor(Genic, levels=c("UTR5", "CDS", "coding_exon", "coding_intron", "UTR3", "long_ncRNA_exon", "long_ncRNA_intron", "ncRNA", "other", "Intergenic"))]

tab = rbindlist(list(`1`=tab.R[, list(N=sum(N)), by="Genomic"], `2`=tab.R[, list(N=sum(N)), by="Genic"]), idcol="Group")
tab[, Genomic := factor(Genomic, levels=c("Genic", "UTR5", "CDS", "coding_exon", "coding_intron", "UTR3", "long_ncRNA_exon", "long_ncRNA_intron", "ncRNA", "other", "Intergenic"))]
tab = tab[order(Genomic)][, list(Genomic, N, value=N/sum(N), percentage=round(N/sum(N)*100, 1), position=cumsum(N/sum(N)) - 0.5*(N/sum(N))), by="Group"]
tab[, Group := as.numeric(Group)]

gg = ggplot(tab, aes(x=Group, y=value, fill=Genomic)) +
   geom_bar(position="stack", stat="identity", col="black", width=0.7) +
   scale_fill_ptol("") +
   coord_polar(theta="y", direction=-1) +
   theme_minimal() +
   geom_text(aes(x = Group+0.5, y = 1-position, label = paste0(percentage, "%")), col="black", size = 4) +
   theme(axis.text = element_blank(), axis.ticks = element_blank(), panel.grid  = element_blank(),
         axis.text.y = element_blank(), axis.text.x = element_blank(), axis.title=element_blank(),
         panel.background = element_blank(), strip.background = element_blank(), strip.text.x = element_text(size = 14),
         strip.text.y = element_text(size = 14), legend.title=element_text(size = 14), legend.text=element_text(size = 14))
ggsave(gg, filename="plots/Figure6C.pdf", width=12, height=8)

tab.D[, Genomic := annotation.D]
tab.D[!Genomic%in%c("Intergenic", "Multigenic"), Genomic := "Genic"]
tab.D[Genomic%in%"Multigenic", Genomic := "Genic_unassigned"]
tab.D[, Genic := region.D]
tab.D[annotation.D%in%"Multigenic", Genic := "Genic_unassigned"]
tab.D[annotation.D%in%"other", Genic := "other"]
tab.D[annotation.D%in%"long_ncRNA", Genic := paste(annotation.D, region.D, sep="_")]
tab.D[annotation.D%in%"ncRNA", Genic := "ncRNA"]
tab.D[annotation.D%in%"protein_coding", Genic := paste("coding", region.D, sep="_")]
tab.D[Genic%in%"coding_CDS", Genic := "CDS"][Genic%in%"coding_five_prime_UTR", Genic := "UTR5"][Genic%in%"coding_three_prime_UTR", Genic := "UTR3"]

tab = rbindlist(list(`1`=tab.D[, list(N=sum(N)), by="Genomic"], `2`=tab.D[, list(N=sum(N)), by="Genic"]), idcol="Group")
tab[, Genomic := factor(Genomic, levels=c("Genic", "UTR5", "CDS", "coding_exon", "coding_intron", "UTR3", "long_ncRNA_exon", "long_ncRNA_intron", "ncRNA", "other", "Genic_unassigned", "Intergenic"))]
tab = tab[order(Genomic)][, list(Genomic, N, value=N/sum(N), percentage=round(N/sum(N)*100, 1), position=cumsum(N/sum(N)) - 0.5*(N/sum(N))), by="Group"]
tab[, Group := as.numeric(Group)]

gg = ggplot(tab, aes(x=Group, y=value, fill=Genomic)) +
   geom_bar(position="stack", stat="identity", col="black", width=0.7) +
   scale_fill_ptol("") +
   coord_polar(theta="y", direction=-1) +
   theme_minimal() +
   geom_text(aes(x = Group+0.5, y = 1-position, label = paste0(percentage, "%")), col="black", size = 4) +
   theme(axis.text = element_blank(), axis.ticks = element_blank(), panel.grid  = element_blank(),
         axis.text.y = element_blank(), axis.text.x = element_blank(), axis.title=element_blank(),
         panel.background = element_blank(), strip.background = element_blank(), strip.text.x = element_text(size = 14),
         strip.text.y = element_text(size = 14), legend.title=element_text(size = 14), legend.text=element_text(size = 14))
ggsave(gg, filename="plots/Figure6D.pdf", width=12, height=8)
```

```{r barplotBOTH, eval=FALSE}
load(paste0("RData/RNA-DNA_biotypes_", gencode_version, "_level_1_2_3_AllOverlapsUnique_stats.RData"))

tab.BR = tab.R[Condition%in%"1FA",][, Condition := NULL]
tab.BD = tab.D[Condition%in%"1FA",][, Condition := NULL]

load(paste0("RData/GRIDseq-DNA_biotypes_", gencode_version, "_level_1_2_3_AllOverlapsUnique_stats.RData"))

tab.BR = rbindlist(list(RADICL = tab.BR, GRID = tab.R), idcol="Technique")
tab.BD = rbindlist(list(RADICL = tab.BD, GRID = tab.D), idcol="Technique")

tab.1 = tab.BR
tab.1[, Genomic := factor(ifelse(annotation.R%in%"Intergenic", "Intergenic", "Genic"), levels=c("Genic", "Intergenic"))]
tab.1[, Genic := region.R]
tab.1[annotation.R%in%"other", Genic := "other"]
tab.1[annotation.R%in%"long_ncRNA", Genic := paste(annotation.R, region.R, sep="_")]
tab.1[annotation.R%in%"ncRNA", Genic := "ncRNA"]
tab.1[annotation.R%in%"protein_coding", Genic := paste("coding", region.R, sep="_")]
tab.1[Genic%in%"coding_CDS", Genic := "CDS"][Genic%in%"coding_five_prime_UTR", Genic := "UTR5"][Genic%in%"coding_three_prime_UTR", Genic := "UTR3"]
tab.1[Genic%in%"intron", Genic := "other"]

tab.R = tab.1[, list(N=sum(N)), by=c("Technique", "Genic")]
tab.R = tab.R[, list(Genic, N, value=N/sum(N), percentage=round(N/sum(N)*100, 1), position=cumsum(N/sum(N)) - 0.5*(N/sum(N))), by="Technique"]

tab.2 = tab.BD
tab.2[, Genomic := annotation.D]
tab.2[!Genomic%in%c("Intergenic", "Multigenic"), Genomic := "Genic"]
tab.2[Genomic%in%"Multigenic", Genomic := "Genic_unassigned"]
tab.2[, Genic := region.D][!is.na(Genic)]
tab.2[annotation.D%in%"Multigenic", Genic := "Genic_unassigned"]
tab.2[annotation.D%in%"other", Genic := "other"]
tab.2[annotation.D%in%"long_ncRNA", Genic := paste(annotation.D, region.D, sep="_")]
tab.2[annotation.D%in%"ncRNA", Genic := "ncRNA"]
tab.2[annotation.D%in%"protein_coding", Genic := paste("coding", region.D, sep="_")]
tab.2[Genic%in%"coding_CDS", Genic := "CDS"][Genic%in%"coding_five_prime_UTR", Genic := "UTR5"][Genic%in%"coding_three_prime_UTR", Genic := "UTR3"]
tab.2[Genic%in%"intron", Genic := "other"]

tab.D = tab.2[, list(N=sum(N)), by=c("Technique", "Genic")]
tab.D = tab.D[, list(Genic, N, value=N/sum(N), percentage=round(N/sum(N)*100, 1), position=cumsum(N/sum(N)) - 0.5*(N/sum(N))), by="Technique"]

tab = rbindlist(list(RNA=tab.R, DNA=tab.D), idcol="Reads")[, Reads := factor(Reads, levels=c("RNA", "DNA"))]
tab[, Genic := factor(Genic, levels=c("UTR5", "CDS", "coding_exon", "coding_intron", "UTR3", "long_ncRNA_exon", "long_ncRNA_intron", "ncRNA", "other", "Genic_unassigned", "Intergenic"))]
tab[, Technique := factor(Technique, levels=c("RADICL", "GRID"))]
colour_fill = setNames(ptol_pal()(12)[seq_along(levels(tab[, Genic]))+1], levels(tab[, Genic]))

# gg = ggplot(tab, aes(x=Technique, y=value, fill=Genic)) +
#    geom_bar(position="stack", stat="identity", col="black", width=0.7, lwd=0.5) +
#    scale_fill_manual("", values=colour_fill) +
gg = ggplot(tab, aes(x=Genic, y=value, fill=Technique)) +
   geom_bar(position="dodge", stat="identity", col="black", width=0.7, lwd=0.5) +
   scale_fill_jco() +
   scale_y_continuous("Dataset percentage", labels=scales::percent) + 
   scale_x_discrete("") +
   facet_grid(~Reads) +
   theme_bw() +
   theme(axis.text.x = element_text(angle = 30, hjust=1, size=18), legend.position=c(0.9, 0.9),
      axis.text = element_text(size = 16), axis.title=element_text(size = 16),
         strip.background = element_blank(), strip.text.x = element_text(size = 16),
         strip.text.y = element_text(size = 14), legend.title=element_text(size = 14), legend.text=element_text(size = 14))
ggsave(gg, filename=paste0("plots/RADICL_vs_GRID.pdf"), width=16, height=10)

tab = tab[!Genic%like%"intron",]
gg = ggplot(tab, aes(x=Genic, y=value, fill=Technique)) +
   geom_bar(position="dodge", stat="identity", col="black", width=0.7, lwd=0.5) +
   scale_fill_jco() +
   scale_y_continuous("Dataset percentage", labels=scales::percent) + 
   scale_x_discrete("") +
   facet_grid(~Reads) +
   theme_bw() +
   theme(axis.text.x = element_text(angle = 30, hjust=1, size=18), legend.position=c(0.9, 0.9),
      axis.text = element_text(size = 16), axis.title=element_text(size = 16),
         strip.background = element_blank(), strip.text.x = element_text(size = 16),
         strip.text.y = element_text(size = 14), legend.title=element_text(size = 14), legend.text=element_text(size = 14))
ggsave(gg, filename=paste0("plots/RADICL_vs_GRID_noIntrons.pdf"), width=16, height=10)
```

<!-- ```{r processFiles, eval=FALSE} -->
<!-- files = list.files("./RNA-DNA", recursive=TRUE, full.names=TRUE, pattern="*.gz$") -->
<!-- identifiers = paste(tstrsplit(files, "\\.|_|/")[[4]], tstrsplit(files, "\\.|_|/")[[5]], tstrsplit(files, "\\.|_|/")[[7]], sep="_") -->

<!-- statsLib = data.frame(FragmentsPairs=as.numeric(), UniqueFragmentsPairs=as.numeric(), -->
<!--                       UniqueRNAFragments=as.numeric(), UniqueDNAFragments=as.numeric()) -->
<!-- statsGene = data.frame(GenicFragmentsPairs=as.numeric(), -->
<!--                        GenicRNAFragments=as.numeric(), GenicDNAFragments=as.numeric(), -->
<!--                        GenicUniqueRNAFragments=as.numeric(), GenicDNAUniqueFragments=as.numeric()) -->
<!-- statsMulti = data.frame(UniqueGenicFragmentsPairs=as.numeric(), -->
<!--                         UniqueGenicRNAFragments=as.numeric(), UniqueGenicDNAFragments=as.numeric(), -->
<!--                         UniqueGenicUniqueRNAFragments=as.numeric(), UniqueGenicDNAUniqueFragments=as.numeric()) -->
<!-- statsUnique = data.frame(FinalUniqueFragmentsPairs=as.numeric(), FinalUniqueGeneBindingFragmentsPairs=as.numeric(), -->
<!--                          FinalUniqueGeneBindingRNAFragments=as.numeric(), -->
<!--                          FinalUniqueRNAIDs=as.numeric(), FinalUniqueGeneBindingRNAIDs=as.numeric(), -->
<!--                          FinalUniqueGeneBindingDNAFragments=as.numeric(), -->
<!--                          FinalUniqueDNAIDs=as.numeric(), FinalUniqueGeneBindingDNAIDs=as.numeric(), -->
<!--                          FinalUniqueIntergenicFragmentPairs=as.numeric()) -->

<!-- interactions = list() -->
<!-- for( i in seq_along(files) ){ -->
<!--                      file = files[i] -->
<!--                      bed = read.paired.bed(file) -->

<!--                      statsLib[i,]$FragmentsPairs = length(bed[[1]]) -->
<!--                      statsLib[i,]$UniqueFragmentsPairs = sum(bed[[1]]$score==37 & bed[[2]]$score==37) -->
<!--                      statsLib[i,]$UniqueRNAFragments = sum(bed[[1]]$score==37) -->
<!--                      statsLib[i,]$UniqueDNAFragments = sum(bed[[2]]$score==37) -->

<!--                      overlaps.rna = as.data.table(findOverlaps(bed[[1]], genes, ignore.strand=FALSE, minoverlap=16)) -->
<!--                      setkey(overlaps.rna, subjectHits) -->
<!--                      overlaps.rna = overlaps.rna[genes.dt[, list(index, gene_id.R=gene_id)], nomatch=0][, subjectHits := NULL] -->
<!--                      setkey(overlaps.rna, queryHits) -->

<!--                      overlaps.dna = as.data.table(findOverlaps(bed[[2]], genes, ignore.strand=TRUE, minoverlap=16)) -->
<!--                      setkey(overlaps.dna, subjectHits) -->
<!--                      overlaps.dna = overlaps.dna[genes.dt[, list(index, gene_id.D=gene_id)], nomatch=0][, subjectHits := NULL] -->
<!--                      setkey(overlaps.dna, queryHits) -->

<!--                      bed = lapply(bed, function(x) data.table(as.data.frame(x), key="index")) -->

<!--                      statsGene[i,]$GenicFragmentsPairs = nrow(overlaps.rna[overlaps.dna, nomatch=0]) -->
<!--                      tmp = overlaps.rna[bed[[1]], nomatch=0] -->
<!--                      statsGene[i,]$GenicRNAFragments = length(unique(tmp[, gene_id.R])) -->
<!--                      statsGene[i,]$GenicUniqueRNAFragments = length(unique(tmp[score==37, gene_id.R])) -->
<!--                      tmp = overlaps.dna[bed[[2]], nomatch=0] -->
<!--                      statsGene[i,]$GenicDNAFragments = length(unique(tmp[, gene_id.D])) -->
<!--                      statsGene[i,]$GenicDNAUniqueFragments = length(unique(tmp[score==37, gene_id.D])) -->

<!--                      sel.rna = overlaps.rna[, .N, by=key(overlaps.rna)][N>1, queryHits] -->
<!--                      overlaps.rna = overlaps.rna[!queryHits%in%sel.rna,] -->

<!--                      sel.dna = overlaps.dna[, .N, by=key(overlaps.dna)][N>1, queryHits] -->
<!--                      overlaps.dna = overlaps.dna[!queryHits%in%sel.dna,] -->

<!--                      statsMulti[i,]$UniqueGenicFragmentsPairs = nrow(overlaps.rna[overlaps.dna, nomatch=0]) -->
<!--                      tmp = overlaps.rna[bed[[1]], nomatch=0] -->
<!--                      statsMulti[i,]$UniqueGenicRNAFragments = length(unique(tmp[, gene_id.R])) -->
<!--                      statsMulti[i,]$UniqueGenicUniqueRNAFragments = length(unique(tmp[score==37, gene_id.R])) -->
<!--                      tmp = overlaps.dna[bed[[2]], nomatch=0] -->
<!--                      statsMulti[i,]$UniqueGenicDNAFragments = length(unique(tmp[, gene_id.D])) -->
<!--                      statsMulti[i,]$UniqueGenicDNAUniqueFragments = length(unique(tmp[score==37, gene_id.D])) -->

<!--                      overlaps.rna = overlaps.rna[bed[[1]][, list(index, chr.R=seqnames, start.R=start, end.R=end, score.R=score, strand.R=strand)], nomatch=0] -->
<!--                      overlaps.dna = overlaps.dna[bed[[2]][, list(index, chr.D=seqnames, start.D=start, end.D=end, score.D=score)], nomatch=0] -->

<!--                      overlaps = merge(overlaps.rna, overlaps.dna, all=TRUE) -->

<!--                      overlaps = overlaps[bed[[1]][, list(index, seqnames, start, end, score, strand)], nomatch=0][score==37,] -->
<!--                      overlaps[is.na(gene_id.R), `:=`(chr.R=seqnames, start.R=start, end.R=end, score.R=score, strand.R=strand)] -->
<!--                      overlaps[, `:=`(seqnames=NULL, start=NULL, end=NULL, score=NULL, strand=NULL)] -->
<!--                      overlaps[queryHits%in%sel.rna, gene_id.R := "Multigenic"] -->
<!--                      overlaps[is.na(gene_id.R), gene_id.R := "Intergenic"] -->

<!--                      overlaps = overlaps[bed[[2]][, list(index, seqnames, start, end, score)], nomatch=0][score==37,] -->
<!--                      overlaps[is.na(gene_id.D), `:=`(chr.D=seqnames, start.D=start, end.D=end, score.D=score)] -->
<!--                      overlaps[, `:=`(seqnames=NULL, start=NULL, end=NULL, score=NULL)] -->
<!--                      overlaps[queryHits%in%sel.dna, gene_id.D := "Multigenic"] -->
<!--                      overlaps[is.na(gene_id.D), gene_id.D := "Intergenic"] -->

<!--                      orphan = sum(bed[[1]][!overlaps[, queryHits], score]==37 & bed[[2]][!overlaps[, queryHits], score]==37) -->
<!--                      overlaps[, queryHits := NULL] -->

<!--                      rm(bed, tmp, sel.rna, sel.dna) -->

<!--                      interactions[[i]] = overlaps[score.R==37 & score.D==37,] -->

<!--                      statsUnique[i,]$FinalUniqueFragmentsPairs = nrow(interactions[[i]]) -->
<!--                      statsUnique[i,]$FinalUniqueGeneBindingFragmentsPairs = nrow(interactions[[i]][!(gene_id.R%in%c("Multigenic", "Intergenic") | gene_id.D%in%c("Multigenic", "Intergenic"))]) -->
<!--                      statsUnique[i,]$FinalUniqueGeneBindingRNAFragments = nrow(interactions[[i]][!gene_id.D%in%c("Multigenic", "Intergenic"),]) -->
<!--                      statsUnique[i,]$FinalUniqueRNAIDs = length(unique(interactions[[i]][!gene_id.R%in%"Multigenic", gene_id.R])) -->
<!--                      statsUnique[i,]$FinalUniqueGeneBindingRNAIDs = length(unique(interactions[[i]][!gene_id.R%in%c("Multigenic", "Intergenic"), gene_id.R])) -->
<!--                      statsUnique[i,]$FinalUniqueGeneBindingDNAFragments = nrow(interactions[[i]][!gene_id.R%in%c("Multigenic", "Intergenic"),]) -->
<!--                      statsUnique[i,]$FinalUniqueDNAIDs = length(unique(interactions[[i]][!gene_id.D%in%"Multigenic", gene_id.D])) -->
<!--                      statsUnique[i,]$FinalUniqueGeneBindingDNAIDs = length(unique(interactions[[i]][!gene_id.D%in%c("Multigenic", "Intergenic"), gene_id.D])) -->
<!--                      statsUnique[i,]$FinalUniqueIntergenicFragmentPairs = orphan -->

<!--                      rm(overlaps, overlaps.rna, overlaps.dna) -->
<!--                      gc() -->
<!-- } -->

<!-- names(interactions) = identifiers -->
<!-- rownames(statsLib) = identifiers -->
<!-- rownames(statsGene) = identifiers -->
<!-- rownames(statsMulti) = identifiers -->
<!-- rownames(statsUnique) = identifiers -->

<!-- # Concatenate the list -->
<!-- interactions = rbindlist(interactions, idcol="name") -->

<!-- # Remove the scores (they are all 37...) -->
<!-- interactions[, `:=`(score.R = NULL, score.D = NULL)] -->

<!-- save(statsLib, statsGene, statsMulti, statsUnique, -->
<!--      file=paste0("RData/RNA-DNA_overlaps_", gencode_version, "_level_1_2_3_AllOverlapsUniqueStats.RData")) -->
<!-- saveRDS(interactions, -->
<!--         paste0("RData/RNA-DNA_overlaps_", gencode_version, "_level_1_2_3_AllOverlapsUnique.rds")) -->
<!-- ``` -->

<!-- ### Post-processing and filtering -->

<!-- ```{r filterJustHere, eval=FALSE} -->
<!-- if(!exists("interactions")) -->
<!--    interactions = readRDS(paste0("RData/RNA-DNA_overlaps_", gencode_version, "_level_1_2_3_AllOverlapsUnique.rds")) -->

<!-- interactions = interactions[gene_id.R%in%c(coding.sel, unlist(noncoding.sel), c("Multigenic", "Intergenic")) & gene_id.D%in%c(coding.sel, unlist(noncoding.sel), c("Multigenic", "Intergenic")),] -->
<!-- # interactions = interactions[!name%in%paste0("Hiseq190717_lane1_", c("ACAGTG", "ATCACG", "CAGATC", "GCCAAT")),] -->

<!-- setkey(interactions, name) -->
<!-- interactions = interactions[tab0[, list(Label, Condition, BiolReplicate)], nomatch=0] -->
<!-- interactions[, name := NULL] -->

<!-- # interactions[Condition%in%"Control" & BiolReplicate%in%c(1,2), Condition := "Control1"] -->
<!-- # interactions[Condition%in%"Control" & BiolReplicate%in%c(3,4), Condition := "Control2"] -->
<!-- # interactions[Condition%in%"Control2", BiolReplicate := ifelse(BiolReplicate==3, 1, 2)] -->

<!-- writeGzip(interactions, file=paste0("RNA-DNA_interactions_", gencode_version, "_level_1_2_3_AllOverlapsUnique.tsv"), col.names=TRUE, showProgress=TRUE, wait=FALSE, level=9) -->

<!-- # # FILTERED -->
<!-- # interactions = interactions[!(gene_id.R%in%c("Multigenic", "Intergenic") | gene_id.D%in%c("Multigenic", "Intergenic"))] -->
<!-- # -->
<!-- # setkeyv(interactions, c("Condition", "BiolReplicate", "gene_id.R", "gene_id.D")) -->
<!-- #  -->
<!-- # counts = interactions[, .N, by=key(interactions)] -->
<!-- #  -->
<!-- # counts = dcast.data.table(counts, Condition+gene_id.R+gene_id.D~BiolReplicate, value.var="N") -->
<!-- #  -->
<!-- # counts = counts[rowSums(counts[, 4:ncol(counts), with=FALSE]>1, na.rm=TRUE)>=2,] -->
<!-- #  -->
<!-- # counts = melt.data.table(counts, id.vars=c("Condition", "gene_id.R", "gene_id.D"), measure.vars=c("1", "2", "3"), variable.name="Replicate", value.name="N") -->
<!-- #  -->
<!-- # counts = counts[!is.na(N),] -->
<!-- #  -->
<!-- # fwrite(counts, file=paste0("RNA-DNA_counts_", gencode_version, "_level_1_2_3_AllOverlapsUnique_filtered.tsv"), quote=FALSE, sep='\t', row.names=FALSE) -->
<!-- #  -->
<!-- # # FILTERED (extended) -->
<!-- # setkeyv(counts, c("Condition", "gene_id.R", "gene_id.D", "Replicate")) -->
<!-- # setnames(interactions, "BiolReplicate", "Replicate") -->
<!-- # interactions[, Replicate := factor(Replicate)] -->
<!-- # setkeyv(interactions, key(counts)) -->
<!-- # subset = interactions[counts, nomatch=0][, N := NULL] -->
<!-- # fwrite(subset, file=paste0("RNA-DNA_counts_", gencode_version, "_level_1_2_3_AllOverlapsUnique_filtered_extended.tsv"), quote=FALSE, sep='\t', row.names=FALSE) -->
<!-- ``` -->

### General metrics

<!-- ```{r, stats, eval=FALSE} -->
<!-- interactions = fread(paste0("gunzip -c RNA-DNA_interactions_", gencode_version, "_level_1_2_3_AllOverlapsUnique.tsv.gz")) -->

<!-- setnames(interactions, "BiolReplicate", "Replicate") -->
<!-- setkeyv(interactions, c("Condition", "Replicate", "chr.R", "gene_id.R", "chr.D", "gene_id.D")) -->
<!-- interactions = interactions[, .N, by=key(interactions)] -->
<!-- setnames(interactions, "N", "Count") -->

<!-- tab = interactions[, list(Intra=sum(.SD[chr.R==chr.D, Count]), Total=sum(Count)), by=c("Condition", "Replicate")][, Perc := Intra/Total*100] -->

<!-- interactions = fread(paste0("gunzip -c RNA-DNA_counts_", gencode_version, "_level_1_2_3_AllOverlapsUnique_filtered_extended.tsv.gz")) -->

<!-- setkeyv(interactions, c("Condition", "Replicate", "chr.R", "gene_id.R", "chr.D", "gene_id.D")) -->
<!-- interactions = interactions[, .N, by=key(interactions)] -->
<!-- setnames(interactions, "N", "Count") -->

<!-- tab2 = interactions[, list(Intra=sum(.SD[chr.R==chr.D, Count]), Total=sum(Count)), by=c("Condition", "Replicate")][, Perc := Intra/Total*100] -->
<!-- ``` -->

### Enzymatic treatment

```{r processEnzymatic, eval=FALSE}
files = list.files("./Enzymatic", full.names=TRUE, pattern="*.bed.gz$")
identifiers = tstrsplit(files, "\\.|_|/")[[4]]

interactions = list()
for( i in seq_along(files) ){
   file = files[i]
   bed = read.paired.bed(file, min.mapq=37)
   
   # RNA
   overlaps.rna = as.data.table(findOverlaps(bed[[1]], tx.regions, ignore.strand=FALSE))
   setkey(overlaps.rna, subjectHits)
   overlaps.rna = overlaps.rna[tx.regions.dt[, list(index, gene_id.R=gene_id, annotation.R=annotation, gene_type.R=gene_type, region.R=type)], nomatch=0][, subjectHits := NULL]
   setkey(overlaps.rna, queryHits)
   # Multigenic
   overlaps.rna[queryHits%in%overlaps.rna[, .N, by="queryHits"][N>1, queryHits], `:=`(gene_id.R="Multigenic", annotation.R="Multigenic", gene_type.R="Multigenic", region.R="Multigenic")]
   overlaps.rna = unique(overlaps.rna)
   # Intergenic
   missing.rna = data.table(queryHits=setdiff(1:length(bed[[1]]), overlaps.rna[, queryHits]))
   sel = overlapsAny(bed[[1]][missing.rna[, queryHits],], tx.regions)
   missing.rna[sel,  `:=`(gene_id.R="Genic_unassigned", annotation.R="Genic_unassigned", gene_type.R="Genic_unassigned", region.R="Genic_unassigned")]
   missing.rna[!sel, `:=`(gene_id.R="Intergenic", annotation.R="Intergenic", gene_type.R="Intergenic", region.R="Intergenic")]

   overlaps.rna = rbindlist(list(overlaps.rna, missing.rna))
   setkey(overlaps.rna, queryHits)
   
   # DNA
   overlaps.dna = as.data.table(findOverlaps(bed[[2]], tx.regions, ignore.strand=TRUE))
   setkey(overlaps.dna, subjectHits)
   overlaps.dna = overlaps.dna[tx.regions.dt[, list(index, gene_id.D=gene_id, annotation.D=annotation, gene_type.D=gene_type, region.D=type)], nomatch=0][, subjectHits := NULL]
   setkey(overlaps.dna, queryHits)
   # Multigenic
   overlaps.dna[queryHits%in%overlaps.dna[, .N, by="queryHits"][N>1, queryHits], `:=`(gene_id.D="Multigenic", annotation.D="Multigenic", gene_type.D="Multigenic", region.D="Multigenic")]
   overlaps.dna = unique(overlaps.dna)
   # Intergenic
   missing.dna = data.table(queryHits=setdiff(1:length(bed[[2]]), overlaps.dna[, queryHits]))
   sel = overlapsAny(bed[[2]][missing.dna[, queryHits],], tx.regions)
   missing.dna[sel,  `:=`(gene_id.D="Genic_unassigned", annotation.D="Genic_unassigned", gene_type.D="Genic_unassigned", region.D="Genic_unassigned")]
   missing.dna[!sel, `:=`(gene_id.D="Intergenic", annotation.D="Intergenic", gene_type.D="Intergenic", region.D="Intergenic")]
   
   overlaps.dna = rbindlist(list(overlaps.dna, missing.dna))
   setkey(overlaps.dna, queryHits)
   
   bed = lapply(bed, function(x) data.table(as.data.frame(x), key="index"))
   
   overlaps.rna = overlaps.rna[bed[[1]][, list(index, chr.R=seqnames, start.R=start, end.R=end, strand.R=strand)], nomatch=0]
   setcolorder(overlaps.rna, c(1, 6:9,  2:5))
   overlaps.dna = overlaps.dna[bed[[2]][, list(index, chr.D=seqnames, start.D=start, end.D=end)], nomatch=0]
   setcolorder(overlaps.dna, c(1, 6:8,  2:5))
   
   rm(bed); gc()

   overlaps = overlaps.rna[overlaps.dna, nomatch=0][, queryHits := NULL]
   overlaps[, `:=`(region.R = droplevels(region.R), region.D = droplevels(region.D))]
   
   rm(overlaps.rna, overlaps.dna); gc()
   
   # writeGzip(overlaps,
   #           file=paste0("biotype/RNA-DNA_", identifiers[i], "_biotypes_", gencode_version, "_level_1_2_3_AllOverlapsUnique.tsv"),
   #           col.names=TRUE, showProgress=TRUE, wait=FALSE, level=9)

   interactions[[i]] = overlaps

   rm(overlaps)
   gc()
   
}

names(interactions) = identifiers

# Concatenate the list
interactions = rbindlist(interactions, idcol="name")

saveRDS(interactions,
        paste0("RData/Enzymatic_overlaps_", gencode_version, "_level_1_2_3_AllOverlapsUnique.rds"))
```

```{r statsEnzymatic, eval=FALSE}
interactions = readRDS(paste0("RData/Enzymatic_overlaps_", gencode_version, "_level_1_2_3_AllOverlapsUnique.rds"))

interactions[, distance := abs(start.R-end.D)]
binDistance = as.character(cut(interactions[, distance], breaks=c(-1, 1e3, 1e4, 1e6, Inf), labels=c("<1Kb", "1-10Kb", "10Kb-1Mb", ">1Mb")))
binDistance[interactions[, chr.R!=chr.D]] = "Interchromosomal" 
interactions[, distance := factor(binDistance, levels=c(c("<1Kb", "1-10Kb", "10Kb-1Mb", ">1Mb", "Interchromosomal")))]

dt = interactions[, .N, by=c("name", "distance")][, list(distance, N, P=N/sum(N)), by="name"][order(distance)]
dt[name%in%"NoEnzyme", name := "None"][name%in%"RNaseH", name := "H"][, name := factor(name, levels=c("None", "S1", "V1", "H"))]

gg = ggplot(dt, aes(x=distance, y=P, fill=name)) +
   geom_bar(stat="identity", position="dodge", col="black") +
   scale_fill_brewer("RNAse treatment", palette="Set1") +
   scale_x_discrete("Distance between RNA and DNA tags") +
   scale_y_continuous("Dataset fraction", limits=c(0,1), labels=scales::percent) +
   theme_bw() + theme(legend.position=c(0.1, 0.85), legend.box.background = element_rect())

ggsave(gg, filename="plots/Enzymatic_distance_between_tags.pdf", width=10, height=6)
```


```{r tagDistanceGRID, eval=FALSE}
interactions = fread(paste0("gunzip -c GRIDseq-DNA_biotypes_", gencode_version, "_level_1_2_3_AllOverlapsUnique_.tsv.gz"))

interactions[, distance := abs(start.R-end.D)]
binDistance = as.character(cut(interactions[, distance], breaks=c(-1, 1e3, 1e4, 1e5, 1e6, Inf), labels=c("<1Kb", "1-10Kb", "10-100Kb", "100Kb-1Mb", ">1Mb")))
binDistance[interactions[, chr.R!=chr.D]] = "Interchromosomal" 
interactions[, distance := factor(binDistance, levels=c(c("<1Kb", "1-10Kb", "10-100Kb", "100Kb-1Mb", ">1Mb", "Interchromosomal")))]

dt1 = interactions[, .N, by="distance"][, list(Condition="GRID", distance, N, P=N/sum(N))][order(distance)]

interactions = fread(paste0("gunzip -c RNA-DNA_biotypes_", gencode_version, "_level_1_2_3_AllOverlapsUnique.tsv.gz"))
interactions = interactions[Condition%in%"1FA",]

interactions[, distance := abs(start.R-end.D)]
binDistance = as.character(cut(interactions[, distance], breaks=c(-1, 1e3, 1e4, 1e5, 1e6, Inf), labels=c("<1Kb", "1-10Kb", "10-100Kb", "100Kb-1Mb", ">1Mb")))
binDistance[interactions[, chr.R!=chr.D]] = "Interchromosomal" 
interactions[, distance := factor(binDistance, levels=c(c("<1Kb", "1-10Kb", "10-100Kb", "100Kb-1Mb", ">1Mb", "Interchromosomal")))]

dt2 = interactions[, .N, by="distance"][, list(Condition="RADICL", distance, N, P=N/sum(N))][order(distance)]

dt = rbindlist(list(dt1, dt2))[, Condition := factor(Condition, levels=c("RADICL", "GRID"))]

gg = ggplot(dt, aes(x=distance, y=P, fill=Condition)) +
   geom_bar(stat="identity", position="dodge", col="black") +
   geom_text(aes(label=scales::percent(P)), position=position_dodge(width=0.9), vjust=-0.5) +
   scale_fill_brewer(palette="Set1") +
   scale_x_discrete("Distance between RNA and DNA tags") +
   scale_y_continuous("Dataset fraction", limits=c(0,1), labels=scales::percent) +
   theme_bw() + theme(legend.position=c(0.1, 0.85), legend.box.background = element_rect(colour="white"))
ggsave(gg, filename="plots/Comaprison_distance_between_tags.pdf", width=7, height=5)

# interactions = fread(paste0("gunzip -c GRIDseq-DNA_biotypes_", gencode_version, "_level_1_2_3_AllOverlapsUnique_.tsv.gz"))
# 
# interactions[, distance := abs(start.R-end.D)]
# binDistance = as.character(cut(interactions[, distance], breaks=c(-1, 1e3, 1e4, 1e6, Inf), labels=c("<1Kb", "1-10Kb", "10Kb-1Mb", ">1Mb")))
# binDistance[interactions[, chr.R!=chr.D]] = "Interchromosomal"
# binDistance[interactions[, gene_id.R==gene_id.D]] = "Gene body" 
# interactions[, distance := factor(binDistance, levels=c(c("Gene body", "<1Kb", "1-10Kb", "10Kb-1Mb", ">1Mb", "Interchromosomal")))]
# 
# dt1 = interactions[, .N, by="distance"][, list(Condition="GRID", distance, N, P=N/sum(N))][order(distance)]
# 
# interactions = fread(paste0("gunzip -c RNA-DNA_biotypes_", gencode_version, "_level_1_2_3_AllOverlapsUnique.tsv.gz"))
# interactions = interactions[Condition%in%"1FA",]
# 
# interactions[, distance := abs(start.R-end.D)]
# binDistance = as.character(cut(interactions[, distance], breaks=c(-1, 1e3, 1e4, 1e6, Inf), labels=c("<1Kb", "1-10Kb", "10Kb-1Mb", ">1Mb")))
# binDistance[interactions[, chr.R!=chr.D]] = "Interchromosomal" 
# binDistance[interactions[, gene_id.R==gene_id.D]] = "Gene body" 
# interactions[, distance := factor(binDistance, levels=c(c("Gene body", "<1Kb", "1-10Kb", "10Kb-1Mb", ">1Mb", "Interchromosomal")))]
# 
# dt2 = interactions[, .N, by="distance"][, list(Condition="RADICL", distance, N, P=N/sum(N))][order(distance)]
# 
# dt = rbindlist(list(dt1, dt2))[, Condition := factor(Condition, levels=c("RADICL", "GRID"))]
# 
# gg = ggplot(dt, aes(x=distance, y=P, fill=Condition)) +
#    geom_bar(stat="identity", position="dodge", col="black") +
#    geom_text(aes(label=scales::percent(P)), position=position_dodge(width=0.9), vjust=-0.5) +
#    scale_fill_brewer(palette="Set1") +
#    scale_x_discrete("Distance between RNA and DNA tags") +
#    scale_y_continuous("Dataset fraction", limits=c(0,1), labels=scales::percent) +
#    theme_bw() + theme(legend.position=c(0.1, 0.85), legend.box.background = element_rect(colour="white"))
# ggsave(gg, filename="plots/Comaprison_distance_between_tags_2.pdf", width=7, height=5)

```

### RNA-seq

```{r correlationsTestFunction}
cor.diff.test = function(x1, x2, y1, y2, method="pearson") {
  cor1 = cor.test(x1, x2, method=method)
  cor2 = cor.test(y1, y2, method=method)

  r1 = cor1$estimate
  r2 = cor2$estimate
  n1 = sum(complete.cases(x1, x2))
  n2 = sum(complete.cases(y1, y2))
  fisher = ((0.5*log((1+r1)/(1-r1)))-(0.5*log((1+r2)/(1-r2))))/((1/(n1-3))+(1/(n2-3)))^0.5

  p.value = (2*(pnorm(abs(fisher), lower.tail=F)))

  result= list(
    "cor1" = list(
      "estimate" = as.numeric(cor1$estimate),
      "p.value" = cor1$p.value,
      "n" = n1
    ),
    "cor2" = list(
      "estimate" = as.numeric(cor2$estimate),
      "p.value" = cor2$p.value,
      "n" = n2
    ),
    "p.value.twosided" = as.numeric(p.value),
    "p.value.onesided" = as.numeric(p.value) / 2
  )
  cat(paste(sep="",
          "cor1: r=", format(result$cor1$estimate, digits=3), ", p=", format(result$cor1$p.value, digits=3), ", n=", result$cor1$n, "\n",
          "cor2: r=", format(result$cor2$estimate, digits=3), ", p=", format(result$cor2$p.value, digits=3), ", n=", result$cor2$n, "\n",
          "diffence: p(one-sided)=", format(result$p.value.onesided, digits=3), ", p(two-sided)=", format(result$p.value.twosided, digits=3), "\n"
  ))
  return(result);
}
```

```{r rnaseq, eval=FALSE}
bed.nuc = fread("RNAseq/mES.R08.nuc.count.bed",  col.names=c("chr", "start", "end", "name", "score", "strand", "count"))
bed.cyt = fread("RNAseq/mES.R08.cyto.count.bed", col.names=c("chr", "start", "end", "name", "score", "strand", "count"))

ordered = genes[match(bed.nuc[, name], genes$gene_id),]

bed.nuc[, `:=`(RPKM = countToRPKM(count, ordered), TPM = countToTPM(count, ordered))]
setkey(bed.nuc, name)

ordered = genes[match(bed.cyt[, name], genes$gene_id),]

bed.cyt[, `:=`(RPKM = countToRPKM(count, ordered), TPM = countToTPM(count, ordered))]
setkey(bed.cyt, name)

interactions = fread(paste0("gunzip -c RNA-DNA_biotypes_", gencode_version, "_level_1_2_3_AllOverlapsUnique.tsv.gz"))

setkeyv(interactions, c("Condition", "gene_id.R", "annotation.R", "Replicate"))
interactions = interactions[, list(N = .N, genes = length(unique(gene_id.D))), by=key(interactions)]
setkeyv(interactions, c("Condition", "gene_id.R", "annotation.R"))
interactions = interactions[, list(Count = mean(N), N = mean(genes)), by=key(interactions)]

interactions = interactions[!gene_id.R%in%c("Multigenic", "Intergenic"),]
interactions = split(interactions, interactions[, Condition])

for( x in seq_along(interactions) ){
   ordered = genes[match(interactions[[x]][, gene_id.R], genes$gene_id),]
   interactions[[x]][, `:=`(RPK = Count/width(ordered)*1e3, RPKM = countToRPKM(Count, ordered), TPM = countToTPM(Count, ordered))]
   setkey(interactions[[x]], gene_id.R) 
   interactions[[x]][bed.nuc, Nuclear := i.TPM]
   interactions[[x]][bed.cyt, Cytosolic := i.TPM]
}

for( i in names(interactions) ){
   
   x1 = interactions[[i]][is.finite(log2(RPK)) & is.finite(log2(Nuclear)) & RPK>=1 & Nuclear>=1,][, log2(RPK)]
   x2 = interactions[[i]][is.finite(log2(RPK)) & is.finite(log2(Nuclear)) & RPK>=1 & Nuclear>=1, ][, log2(Nuclear)]
   y1 = interactions[[i]][is.finite(log2(RPK)) & is.finite(log2(Cytosolic)) & RPK>=1 & Cytosolic>=1,][, log2(RPK)]
   y2 = interactions[[i]][is.finite(log2(RPK)) & is.finite(log2(Cytosolic)) & RPK>=1 & Cytosolic>=1,][, log2(Cytosolic)]
   
   # summary(lm(V1 ~ V2, data = interactions[[1]][is.finite(log2(RPK)) & is.finite(log2(Nuclear))   & RPK>=1 & Nuclear>=1,  ][, list(log2(Nuclear),   log2(RPK))]))
   # summary(lm(V1 ~ V2, data = interactions[[1]][is.finite(log2(RPK)) & is.finite(log2(Cytosolic))   & RPK>=1 & Cytosolic>=1,  ][, list(log2(Cytosolic),   log2(RPK))]))
   # summary(lm(V1 ~ V2, data = interactions[[1]][is.finite(log2(RPK)) & is.finite(log2(N))   & RPK>=1 & N>=1,  ][, list(log2(N),   log2(RPK))]))
   
   res = cor.diff.test(x1, x2, y1, y2)
   
   subset = interactions[[i]][is.finite(log2(RPK)) & is.finite(log2(Nuclear)) & RPK>=1 & Nuclear>=1,]
   maxXY = max(subset[, list(log2(Nuclear),   log2(RPK))])
   lb = paste("R^2 == ", round(res$cor1$estimate^2, 2))
   gg1 = ggplot(subset, aes(x=log2(Nuclear), y=log2(RPK))) +
      geom_point(size=0.5) +
      geom_smooth(method="lm", colour="red", se=FALSE) +
      scale_x_continuous("Nuclear RNA-seq\nlog2(TPM)") +
      scale_y_continuous("RADICL-seq RNA reads\nlog2(RPK)") +
      annotate("text", x=15, y=1, label=lb, parse=TRUE, size=5) +
      coord_fixed(ratio=1, xlim=c(0, maxXY), ylim=c(0, maxXY)) +
      theme_few()
   
   subset = interactions[[i]][is.finite(log2(RPK)) & is.finite(log2(Cytosolic)) & RPK>=1 & Cytosolic>=1,]
   maxXY = max(subset[, list(log2(Cytosolic),   log2(RPK))])
   lb = paste("R^2 == ", round(res$cor2$estimate^2, 2))
   gg2 = ggplot(subset, aes(x=log2(Cytosolic), y=log2(RPK))) +
      geom_point(size=0.5) +
      geom_smooth(method="lm", colour="red", se=FALSE) +
      scale_x_continuous("Cytosolic RNA-seq\nlog2(TPM)") +
      scale_y_continuous("RADICL-seq RNA reads\nlog2(RPK)") +
      annotate("text", x=15, y=1, label=lb, parse=TRUE, size=5) +
      coord_fixed(ratio=1, xlim=c(0, maxXY), ylim=c(0, maxXY)) +
      theme_few()
   
   res = summary(lm(V1 ~ V2, data = interactions[[1]][is.finite(log2(RPK)) & is.finite(log2(N))   & RPK>=1 & N>=1,  ][, list(log2(N),   log2(RPK))]))
   
   subset = interactions[[i]][RPK>=1,]
   maxXY = max(subset[, list(log2(N),   log2(RPK))])
   lb = paste("R^2 == ", round(res$r.squared, 2))
   gg3 = ggplot(subset, aes(x=log2(RPK), y=log2(N))) +
      geom_point(size=0.5) +
      geom_smooth(method="lm", colour="red", se=FALSE) +
      scale_x_continuous("RADICL-seq RNA reads\nlog2(RPK)") +
      scale_y_continuous("RADICL-seq interactions\nlog2(Gene count)") +
      annotate("text", x=15, y=1, label=lb, parse=TRUE, size=5) +
      coord_fixed(ratio=1, xlim=c(0, maxXY), ylim=c(0, maxXY)) +
      theme_few()
   
   pdf(file=paste0("plots/RNA-DNA_RNAseq_correlations_", i, ".pdf"), width=12, height=4)
      print(ggarrange(gg1, gg2, gg3, nrow=1, ncol=3))
   dev.off()
   
   subset = interactions[[i]][RPK>=1,]
   # setkey(subset, "gene_id.R")
   # setkey(genes.dt, gene_id)
   subset[genes.dt, Biotype := annotation.R]
   
   # subset[Biotype%in%c("lincRNA", "processed_transcript", "antisense", "processed_pseudogene", "unprocessed_pseudogene"), Biotype := "lncRNA"]
   # subset[, Biotype := factor(Biotype, levels=c("protein_coding", "lncRNA", "miRNA", "misc_RNA", "snRNA", "snoRNA"))]
   # 
   maxXY = max(subset[, list(log2(N),   log2(RPK))])
   # lb = paste("R^2 == ", round(res$r.squared, 2))
   gg4 = ggplot(subset, aes(x=log2(RPK), y=log2(N))) +
      geom_point(size=0.5) +
      geom_smooth(method="lm", colour="red", se=FALSE) +
      scale_x_continuous("RADICL-seq RNA reads\nlog2(RPK)") +
      scale_y_continuous("RADICL-seq interactions\nlog2(Gene count)") +
      # annotate("text", x=15, y=1, label=lb, parse=TRUE, size=5) +
      coord_fixed(ratio=1, xlim=c(0, maxXY), ylim=c(0, maxXY)) +
      facet_grid(~Biotype) +
      theme_few()
   
   pdf(file=paste0("plots/RNA-DNA_RNAseq_correlations_", i, "_biotype.pdf"), width=16, height=4)
      print(gg4)
   dev.off()
}
```

### CAGE-seq

```{r cageseq, eval=FALSE}
bed.nuc = fread("CAGEseq/nuc.CAGE.TPM.txt",  col.names=c("name", "TPM"))
bed.cyt = fread("CAGEseq/cyto.CAGE.TPM.txt", col.names=c("name", "TPM"))

interactions = fread(paste0("gunzip -c RNA-DNA_biotypes_", gencode_version, "_level_1_2_3_AllOverlapsUnique.tsv.gz"))

setkeyv(interactions, c("Condition", "gene_id.R", "annotation.R", "Replicate"))
interactions = interactions[, list(N = .N, genes = length(unique(gene_id.D))), by=key(interactions)]
setkeyv(interactions, c("Condition", "gene_id.R", "annotation.R"))
interactions = interactions[, list(Count = mean(N), N = mean(genes)), by=key(interactions)]

interactions = interactions[!gene_id.R%in%c("Multigenic", "Intergenic"),]
interactions = split(interactions, interactions[, Condition])

for( x in seq_along(interactions) ){
   ordered = genes[match(interactions[[x]][, gene_id.R], genes$gene_id),]
   interactions[[x]][, `:=`(RPK = Count/width(ordered)*1e3, RPKM = countToRPKM(Count, ordered), TPM = countToTPM(Count, ordered))]
   setkey(interactions[[x]], gene_id.R) 
   interactions[[x]][bed.nuc, Nuclear := i.TPM]
   interactions[[x]][bed.cyt, Cytosolic := i.TPM]
}

x1 = interactions[[1]][is.finite(log2(RPK)) & is.finite(log2(Nuclear)) & RPK>=1 & Nuclear>=1,][, log2(RPK)]
x2 = interactions[[1]][is.finite(log2(RPK)) & is.finite(log2(Nuclear)) & RPK>=1 & Nuclear>=1, ][, log2(Nuclear)]
y1 = interactions[[1]][is.finite(log2(RPK)) & is.finite(log2(Cytosolic)) & RPK>=1 & Cytosolic>=1,][, log2(RPK)]
y2 = interactions[[1]][is.finite(log2(RPK)) & is.finite(log2(Cytosolic)) & RPK>=1 & Cytosolic>=1,][, log2(Cytosolic)]

summary(lm(V1 ~ V2, data = interactions[[1]][is.finite(log2(RPK)) & is.finite(log2(Nuclear))   & RPK>=1 & Nuclear>=1,  ][, list(log2(Nuclear),   log2(RPK))]))
summary(lm(V1 ~ V2, data = interactions[[1]][is.finite(log2(RPK)) & is.finite(log2(Cytosolic))   & RPK>=1 & Cytosolic>=1,  ][, list(log2(Cytosolic),   log2(RPK))]))

res = cor.diff.test(x1, x2, y1, y2)

subset = interactions[[1]][is.finite(log2(RPK)) & is.finite(log2(Nuclear)) & RPK>=1 & Nuclear>=1,]
maxXY = max(subset[, list(log2(Nuclear),   log2(RPK))])
lb = paste("R^2 == ", round(res$cor1$estimate^2, 2))
gg1 = ggplot(subset, aes(x=log2(Nuclear), y=log2(RPK))) +
   geom_point(size=0.5) +
   geom_smooth(method="lm", colour="red", se=FALSE) +
   scale_x_continuous("Nuclear CAGE-seq\nlog2(RPKM)") +
   scale_y_continuous("RADICL-seq RNA reads\nlog2(RPK)") +
   annotate("text", x=maxXY-2, y=maxXY-1, label=lb, parse=TRUE, size=5) +
   coord_fixed(ratio=1, xlim=c(0, maxXY), ylim=c(0, maxXY)) +
   theme_few()

subset = interactions[[1]][is.finite(log2(RPK)) & is.finite(log2(Cytosolic)) & RPK>=1 & Cytosolic>=1,]
maxXY = max(subset[, list(log2(Cytosolic),   log2(RPK))])
lb = paste("R^2 == ", round(res$cor2$estimate^2, 2))
gg2 = ggplot(subset, aes(x=log2(Cytosolic), y=log2(RPK))) +
   geom_point(size=0.5) +
   geom_smooth(method="lm", colour="red", se=FALSE) +
   scale_x_continuous("Cytosolic CAGE-seq\nlog2(RPKM)") +
   scale_y_continuous("RADICL-seq RNA reads\nlog2(RPK)") +
   annotate("text", x=maxXY-2, y=maxXY-1, label=lb, parse=TRUE, size=5) +
   coord_fixed(ratio=1, xlim=c(0, maxXY), ylim=c(0, maxXY)) +
   theme_few()

gg3 =  ggarrange(gg1, gg2, nrow=1, ncol=2)
ggsave(gg3, filename="plots/RNA-DNA_CAGEseq_correlations.pdf", width=10, height=5)
ggsave(gg3, filename="plots/RNA-DNA_CAGEseq_correlations.eps", width=10, height=5)
```

### Intronic Rate

```{r calculateExonIntron, eval=FALSE}
interactions = fread(paste0("gunzip -c RNA-DNA_biotypes_", gencode_version, "_level_1_2_3_AllOverlapsUnique.tsv.gz"))

interactions = interactions[annotation.R%in%c("protein_coding", "long_ncRNA") & region.R%in%c("exon", "intron", "CDS", "five_prime_UTR", "three_prime_UTR"),]
interactions[, region.R := ifelse(region.R%in%"intron", "intron", "exon")]
setkeyv(interactions, c("gene_id.R", "region.R"))

exon_intron_width = tx.regions.dt[annotation%in%c("protein_coding", "long_ncRNA") & type%in%c("exon", "intron", "CDS", "five_prime_UTR", "three_prime_UTR"),]
exon_intron_width = exon_intron_width[, type := ifelse(type%in%"intron", "intron", "exon")][, list(width = sum(width)), by=.(gene_id, type)]
# Select only genes with at least one intron
exon_intron_width = exon_intron_width[, list(type, width, N=.N), by=.(gene_id)][N==2]
setkeyv(exon_intron_width, c("gene_id", "type"))

interactions = interactions[exon_intron_width, width := width][!is.na(width),]

interactions[, mindist := abs(start.R-start.D)]
interactions[, distance := as.character(cut(mindist,
                                            breaks=c(0, 1e3, 1e4, 1e5, 1e6, Inf),
                                            labels=c("<1Kb", "1-10Kb", "10-100Kb", "100Kb-1Mb", ">1Mb"),
                                            include.lowest=TRUE))]

interactions[gene_id.R==gene_id.D, distance := "Self-interaction"]
interactions[chr.R!=chr.D, distance := "Interchromosome"]
interactions[, distance := factor(distance,
                                  levels=c("Self-interaction", "<1Kb", "1-10Kb", "10-100Kb", "100Kb-1Mb", ">1Mb", "Interchromosome"))]

saveRDS(interactions, paste0("RData/RNA-DNA_interactions_", gencode_version, "_level_1_2_3_AllOverlapsUnique_ExonsIntrons.rds"))
```

```{r plotExonIntron, eval=FALSE}
if(!exists("interactions"))
   interactions = readRDS(paste0("RData/RNA-DNA_interactions_", gencode_version, "_level_1_2_3_AllOverlapsUnique_ExonsIntrons.rds"))

# # Sum by Replicate
# setkeyv(interactions, c("Condition", "Replicate", "annotation.R", "gene_id.R", "distance"))
# plotdata = interactions[, .N, by=key(interactions)]
# # Mean by Condition
# setkeyv(plotdata, c("Condition", "annotation.R", "gene_id.R", "distance"))
# plotdata = plotdata[, list(N = mean(N)), by=key(plotdata)]
# 
# bins = quantile(log2(plotdata[, N]), probs=seq(0, 1, 0.25))
# plotdata[, quartile := cut(log2(N), breaks=bins, labels=c("I", "II", "III", "IV"), include.lowest=TRUE)]
# 
# plotdata[, quartile := factor(quartile, levels=rev(c("I", "II", "III", "IV")))]
# 
# ggplot(plotdata, aes(x=distance, y=log2(N), fill=annotation.R)) +
#    geom_boxplot(size=0.2) +
#    scale_x_discrete("") + scale_y_continuous("Interactions (log2)") +
#    facet_grid(quartile~Condition, scales="free_y") +
#    theme_bw() + theme(legend.position="right", axis.text.x=element_text(angle = 45, hjust = 1))

setkeyv(interactions, c("Condition", "Replicate", "annotation.R", "gene_id.R", "region.R", "distance"))

plotdata = interactions[, .N, by=key(interactions)]

# Pseudo-count for missing exonic/intronic reads
plotdata = dcast.data.table(plotdata, Condition+Replicate+gene_id.R+annotation.R+distance~region.R, value.var="N",
                            fun.aggregate="sum", fill=1)

plotdata = melt.data.table(plotdata, id.vars=c("Condition", "Replicate", "gene_id.R", "annotation.R", "distance"),
                           measure.vars=c("exon", "intron"), variable.name="gene_feature", value.name="N")

if(!exists("exon_intron_width")){
   exon_intron_width = tx.regions.dt[annotation%in%c("protein_coding", "long_ncRNA") & type%in%c("exon", "intron", "CDS", "five_prime_UTR", "three_prime_UTR"),]
   exon_intron_width = exon_intron_width[, type := ifelse(type%in%"intron", "intron", "exon")][, list(width = sum(width)), by=.(gene_id, type)]
   # Select only genes with at least one intron
   exon_intron_width = exon_intron_width[, list(type, width, N=.N), by=.(gene_id)][N==2]
   setkeyv(exon_intron_width, c("gene_id", "type"))
}

# Add the exonic and intronic total width per RNA gene
setkeyv(plotdata, c("gene_id.R", "gene_feature"))
plotdata = plotdata[exon_intron_width, width := as.numeric(width)]

plotdata[, K := N/width*1e3]

# Assign the facto levels
plotdata[, `:=`(gene_feature = factor(gene_feature, levels=c("exon", "intron")),
                gene_type = factor(annotation.R, levels=c("protein_coding", "long_ncRNA")))]

# Collapse the replicates by condition
plotdata = plotdata[, list(mK=mean(K)), by=c("Condition", "gene_type", "gene_id.R", "gene_feature", "distance")]
 
plotdata2 = copy(plotdata)
plotdata2[distance%in%c("<1Kb"), distance := "1-10Kb"][distance%in%"1-10Kb", distance := "<10Kb"][distance%in%"Self-interaction", distance := "Gene body"]
plotdata2[, distance := factor(distance, levels=c("Gene body", "<10Kb", "10-100Kb", "100Kb-1Mb", ">1Mb", "Interchromosome"))]
# plotdata2[distance%in%c("<1Kb", "1-10Kb"), distance := "Local interactions"][distance%in%"Self-interaction", distance := "Local interactions"]
# plotdata2[, distance := factor(distance, levels=c("Local interactions", "10-100Kb", "100Kb-1Mb", ">1Mb", "Interchromosome"))]
plotdata2 = dcast.data.table(plotdata2, Condition+gene_type+distance+gene_id.R~gene_feature, value.var="mK", fun.aggregate=mean, fill=0)
plotdata2 = plotdata2[exon>0, `:=`(ratioIE = (intron)/(exon), ratioEI = (exon)/(intron))]
setkeyv(plotdata2, c("Condition", "gene_id.R"))
sel = plotdata2[is.finite(ratioEI), length(unique(distance)), by=key(plotdata2)][V1==length(unique(plotdata2[, distance]))]
plotdata2 = plotdata2[sel, nomatch=0]

tab = plotdata2[Condition%in%c("1FA", "4h_ActD"), list(N=.N/length(unique(distance))), by=c("Condition", "gene_type")]
gg = ggviolin(plotdata2[Condition%in%c("1FA", "4h_ActD")], x="distance", y="log2(ratioEI)", fill="distance", facet.by="gene_type~Condition") +
   ggtitle("RNA processing rate (exon vs inton ~ distance)",
           subtitle="Reads are sum and normalised for each genic feature and averaged per interaction and then per gene.\nOnly RNAs with exonic and intronic reads within all distance bins are reported.") +
   stat_compare_means(ref.group="Gene body", hide.ns=TRUE, label = "p.signif") +
   scale_fill_manual("Distance to transcription site", values=rev(brewer.pal(7,"Blues"))) +
   scale_y_continuous("Exonic/Intronic reads (log2)") +
   scale_x_discrete("") +
   geom_text(data=tab, aes(label=paste("N =", N)), x=-Inf, y=-Inf, hjust=-0.2, vjust=-1, colour="black", inherit.aes=FALSE, parse=FALSE, size=4) +
   theme_bw() +
   theme(axis.text.x=element_text(angle=60, hjust=1, vjust=1, size=14),
         axis.text.y=element_text(size=14), legend.text=element_text(size=14),
         axis.title.y=element_text(size=14), legend.title=element_text(size=14),
         strip.text.x=element_text(size=14), strip.text.y=element_text(size=14),
         legend.position="bottom") +
   guides(fill = guide_legend(ncol = 7)) + facet_wrap(Condition~gene_type, nrow=2) + geom_boxplot(outlier.colour=NA, width=0.2)
ggsave(gg, filename="plots/RNA-DNA_processing_rate.png", units="in", height=14, width=10)
ggsave(gg, filename="plots/RNA-DNA_processing_rate.pdf", height=14, width=10)

tab = plotdata2[Condition%in%"1FA", list(N=.N/length(unique(distance))), by=c("Condition", "gene_type")]
gg = ggviolin(plotdata2[Condition%in%"1FA"], x="distance", y="log2(ratioEI)", fill="distance", facet.by="gene_type~Condition") +
   ggtitle("RNA processing rate (exon vs inton ~ distance)",
           subtitle="Reads are sum and normalised for each genic feature and averaged per interaction and then per gene.\nOnly RNAs with exonic and intronic reads within all distance bins are reported.") +
   stat_compare_means(ref.group="Gene body", hide.ns=TRUE, label = "p.signif") +
   scale_fill_manual("Distance to transcription site", values=rev(brewer.pal(7,"Blues"))) +
   scale_y_continuous("Exonic/Intronic reads (log2)") +
   scale_x_discrete("") +
   geom_text(data=tab, aes(label=paste("N =", N)), x=-Inf, y=Inf, hjust=-0.2, vjust=2, colour="black", inherit.aes=FALSE, parse=FALSE, size=4) +
   theme_bw() +
   theme(axis.text.x=element_text(angle=60, hjust=1, vjust=1, size=14),
         axis.text.y=element_text(size=14), legend.text=element_text(size=14),
         axis.title.y=element_text(size=14), legend.title=element_text(size=14),
         strip.text.x=element_text(size=14), strip.text.y=element_text(size=14),
         legend.position="bottom") +
   guides(fill = guide_legend(ncol = 7)) + facet_wrap(~gene_type, nrow=1) + geom_boxplot(outlier.colour=NA, width=0.2)
ggsave(gg, filename="plots/RNA-DNA_processing_rate_1FA.png", units="in", height=8, width=12)
ggsave(gg, filename="plots/RNA-DNA_processing_rate_1FA.pdf", height=8, width=12)

# plotdata2 = copy(plotdata)
# plotdata2[distance%in%c("<1Kb"), distance := "1-10Kb"][distance%in%"1-10Kb", distance := "<10Kb"]
# plotdata2[, distance := factor(distance, levels=c("Self-interaction", "<10Kb", "10-100Kb", "100Kb-1Mb", ">1Mb", "Interchromosome"))]
# plotdata2 = dcast.data.table(plotdata2, Condition+gene_type+distance+gene_id.R~gene_feature, value.var="mK", fun.aggregate=mean, fill=0)
# tab = plotdata2[Condition%in%c("Control1", "Control2", "1FA", "6h_ActD"), list(Num=length(unique(gene_id.R))), by=c("Condition", "gene_type", "distance")]
# 
# gg = ggplot(tab, aes(x=distance, y=Num, fill=gene_type)) +
#    geom_bar(stat="identity", position="dodge", col="black", lwd=0.5) +
#    scale_fill_brewer("Transcript feature", palette="Set1") +
#    scale_y_continuous("Number of interacting transcripts") +
#    scale_x_discrete("Distance to transcription site") +
#    facet_grid(gene_type~Condition, scales="free_y") +
#    theme_bw() + theme(axis.text.x=element_text(angle=60,hjust=1,vjust=1), legend.position="top")
# ggsave(gg, filename="plots/RNA-DNA_processing_rate_transcripts.png", units="in", height=8, width=14)
# 
# interactions[, score := 1]
# plotdata = dcast.data.table(interactions[!is.na(gene_type),], Condition+gene_id.R+gene_type+distance+gene_feature~Replicate, value.var="score", fun.aggregate=sum)
# plotdata = plotdata[rowSums(plotdata[, list(`1`, `2`, `3`)]>1)>=2,][, N := `1`+`2`+`3`]
# plotdata[, `:=`(`1`=NULL, `2`=NULL, `3`=NULL)]
# 
# # Pseudo-count for missing exonic/intronic reads
# plotdata = dcast.data.table(plotdata, Condition+gene_id.R+gene_type+distance~gene_feature, value.var="N",
#                             fun.aggregate="sum", fill=1)
# plotdata[, `:=`(five_prime_UTR = NULL, three_prime_UTR = NULL)]
# plotdata = melt.data.table(plotdata, id.vars=c("Condition", "gene_id.R", "gene_type", "distance"),
#                            measure.vars=c("exon", "intron"), variable.name="gene_feature", value.name="N")
# 
# # Add the exonic and intronic total width per RNA gene
# setkeyv(plotdata, c("gene_id.R", "gene_feature"))
# plotdata[exon_intron_width, width := as.numeric(i.width)]
# 
# plotdata[, K := N/width*1e3]
# 
# # Assign the facto levels
# plotdata[, `:=`(gene_feature = factor(gene_feature, levels=c("exon", "intron")),
#                 gene_type = factor(gene_type, levels=c("protein_coding", "lincRNA")))]
# # Collapse the replicates by condition
# plotdata = plotdata[, list(mK=mean(K)), by=c("Condition", "gene_type", "gene_id.R", "gene_feature", "distance")]
# 
# plotdata2 = copy(plotdata)
# plotdata2[distance%in%c("<1Kb"), distance := "1-10Kb"][distance%in%"1-10Kb", distance := "<10Kb"]
# plotdata2[, distance := factor(distance, levels=c("Self-interaction", "<10Kb", "10-100Kb", "100Kb-1Mb", ">1Mb", "Interchromosome"))]
# plotdata2 = dcast.data.table(plotdata2, Condition+gene_type+distance+gene_id.R~gene_feature, value.var="mK", fun.aggregate=mean, fill=0)
# plotdata2 = plotdata2[exon>0, `:=`(ratioIE = (intron)/(exon), ratioEI = (exon)/(intron))]
# setkeyv(plotdata2, c("Condition", "gene_id.R"))
# sel = plotdata2[is.finite(ratioEI), length(unique(distance)), by=key(plotdata2)][V1==length(unique(plotdata2[, distance]))]
# plotdata2 = plotdata2[sel, nomatch=0]
# 
# tab = plotdata2[Condition%in%c("Control1", "Control2", "1FA", "6h_ActD"), list(N=.N/length(unique(distance))), by=c("Condition", "gene_type")]
# gg = ggplot(plotdata2[Condition%in%c("Control1", "Control2", "1FA", "6h_ActD")], aes(x=distance, y=log2(ratioEI), fill=distance)) +
#    ggtitle("RNA processing rate (exon vs inton ~ distance)",
#            subtitle="Reads are sum and normalised for each genic feature and averaged per interaction and then per gene.\nOnly RNAs with exonic and intronic reads within all distance bins are reported.") +
#    geom_violin(position="dodge") +
#    geom_boxplot(width=0.2, outlier.colour=NA, fill="white") +
#    # scale_fill_brewer("Transcript feature", palette="Set1") +
#    scale_fill_manual("Distance to transcription site", values=rev(brewer.pal(7,"Blues"))) +
#    scale_y_continuous("Exonic/Intronic reads (log2)") +
#    scale_x_discrete("") +
#    facet_grid(gene_type~Condition) +
#    geom_text(data=tab, aes(label=paste("N =", N)), x=Inf, y=Inf, hjust=1.5, vjust=2, colour="black", inherit.aes=FALSE, parse=FALSE, size=4) +
#    theme_bw() +
#    theme(axis.text.x=element_text(angle=60, hjust=1, vjust=1, size=14),
#          axis.text.y=element_text(size=14), legend.text=element_text(size=14),
#          axis.title.y=element_text(size=14), legend.title=element_text(size=14),
#          strip.text.x=element_text(size=14), strip.text.y=element_text(size=14),
#          legend.position="bottom") +
#    guides(fill = guide_legend(ncol = 7))
# ggsave(gg, filename="plots/RNA-DNA_processing_rate_filtered.png", units="in", height=10, width=14)

# ggboxplot(plotdata2[Condition%in%c("Control1", "Control2", "1FA", "6h_ActD")], x="distance", y="log2(ratioEI)", fill="distance", add="box", facet.by="gene_type~Condition", ) +
#    stat_compare_means(ref.group="Self-interaction", hide.ns=TRUE, label = "p.signif") + scale_fill_manual("Distance to transcription site", values=rev(brewer.pal(7,"Blues"))) +
#    scale_y_continuous("Exonic/Intronic reads (log2)") +
#    scale_x_discrete("") + theme_bw() +
#    theme(axis.text.x=element_text(angle=60, hjust=1, vjust=1, size=14),
#          axis.text.y=element_text(size=14), legend.text=element_text(size=14),
#          axis.title.y=element_text(size=14), legend.title=element_text(size=14),
#          strip.text.x=element_text(size=14), strip.text.y=element_text(size=14),
#          legend.position="bottom") +
#    guides(fill = guide_legend(ncol = 7))
```

```{r calculateExonIntronGRID, eval=FALSE}
interactions = fread(paste0("gunzip -c GRIDseq-DNA_biotypes_", gencode_version, "_level_1_2_3_AllOverlapsUnique_.tsv.gz"))

interactions = interactions[annotation.R%in%c("protein_coding", "long_ncRNA") & region.R%in%c("exon", "intron", "CDS", "five_prime_UTR", "three_prime_UTR"),]
interactions[, region.R := ifelse(region.R%in%"intron", "intron", "exon")]
setkeyv(interactions, c("gene_id.R", "region.R"))

exon_intron_width = tx.regions.dt[annotation%in%c("protein_coding", "long_ncRNA") & type%in%c("exon", "intron", "CDS", "five_prime_UTR", "three_prime_UTR"),]
exon_intron_width = exon_intron_width[, type := ifelse(type%in%"intron", "intron", "exon")][, list(width = sum(width)), by=.(gene_id, type)]
# Select only genes with at least one intron
exon_intron_width = exon_intron_width[, list(type, width, N=.N), by=.(gene_id)][N==2]
setkeyv(exon_intron_width, c("gene_id", "type"))

interactions = interactions[exon_intron_width, width := width][!is.na(width),]

interactions[, mindist := abs(start.R-start.D)]
interactions[, distance := as.character(cut(mindist,
                                            breaks=c(0, 1e3, 1e4, 1e5, 1e6, Inf),
                                            labels=c("<1Kb", "1-10Kb", "10-100Kb", "100Kb-1Mb", ">1Mb"),
                                            include.lowest=TRUE))]

interactions[gene_id.R==gene_id.D, distance := "Self-interaction"]
interactions[chr.R!=chr.D, distance := "Interchromosome"]
interactions[, distance := factor(distance,
                                  levels=c("Self-interaction", "<1Kb", "1-10Kb", "10-100Kb", "100Kb-1Mb", ">1Mb", "Interchromosome"))]

saveRDS(interactions, paste0("RData/GRIDseq-DNA_interactions_", gencode_version, "_level_1_2_3_AllOverlapsUnique_ExonsIntrons.rds"))
```

```{r plotExonIntronGRID, eval=FALSE}
if(!exists("interactions"))
   interactions = readRDS(paste0("RData/GRIDseq-DNA_interactions_", gencode_version, "_level_1_2_3_AllOverlapsUnique_ExonsIntrons.rds"))

# # Sum by Replicate
# setkeyv(interactions, c("Condition", "Replicate", "annotation.R", "gene_id.R", "distance"))
# plotdata = interactions[, .N, by=key(interactions)]
# # Mean by Condition
# setkeyv(plotdata, c("Condition", "annotation.R", "gene_id.R", "distance"))
# plotdata = plotdata[, list(N = mean(N)), by=key(plotdata)]
# 
# bins = quantile(log2(plotdata[, N]), probs=seq(0, 1, 0.25))
# plotdata[, quartile := cut(log2(N), breaks=bins, labels=c("I", "II", "III", "IV"), include.lowest=TRUE)]
# 
# plotdata[, quartile := factor(quartile, levels=rev(c("I", "II", "III", "IV")))]
# 
# ggplot(plotdata, aes(x=distance, y=log2(N), fill=annotation.R)) +
#    geom_boxplot(size=0.2) +
#    scale_x_discrete("") + scale_y_continuous("Interactions (log2)") +
#    facet_grid(quartile~Condition, scales="free_y") +
#    theme_bw() + theme(legend.position="right", axis.text.x=element_text(angle = 45, hjust = 1))

setkeyv(interactions, c("Condition", "Replicate", "annotation.R", "gene_id.R", "region.R", "distance"))

plotdata = interactions[, .N, by=key(interactions)]

# Pseudo-count for missing exonic/intronic reads
plotdata = dcast.data.table(plotdata, Condition+Replicate+gene_id.R+annotation.R+distance~region.R, value.var="N",
                            fun.aggregate="sum", fill=1)

plotdata = melt.data.table(plotdata, id.vars=c("Condition", "Replicate", "gene_id.R", "annotation.R", "distance"),
                           measure.vars=c("exon", "intron"), variable.name="gene_feature", value.name="N")

if(!exists("exon_intron_width")){
   exon_intron_width = tx.regions.dt[annotation%in%c("protein_coding", "long_ncRNA") & type%in%c("exon", "intron", "CDS", "five_prime_UTR", "three_prime_UTR"),]
   exon_intron_width = exon_intron_width[, type := ifelse(type%in%"intron", "intron", "exon")][, list(width = sum(width)), by=.(gene_id, type)]
   # Select only genes with at least one intron
   exon_intron_width = exon_intron_width[, list(type, width, N=.N), by=.(gene_id)][N==2]
   setkeyv(exon_intron_width, c("gene_id", "type"))
}

# Add the exonic and intronic total width per RNA gene
setkeyv(plotdata, c("gene_id.R", "gene_feature"))
plotdata = plotdata[exon_intron_width, width := as.numeric(width)]

plotdata[, K := N/width*1e3]

# Assign the facto levels
plotdata[, `:=`(gene_feature = factor(gene_feature, levels=c("exon", "intron")),
                gene_type = factor(annotation.R, levels=c("protein_coding", "long_ncRNA")))]

# Collapse the replicates by condition
plotdata = plotdata[, list(mK=mean(K)), by=c("Condition", "gene_type", "gene_id.R", "gene_feature", "distance")]
 
plotdata2 = copy(plotdata)
plotdata2[distance%in%c("<1Kb"), distance := "1-10Kb"][distance%in%"1-10Kb", distance := "<10Kb"][distance%in%"Self-interaction", distance := "Gene body"]
plotdata2[, distance := factor(distance, levels=c("Gene body", "<10Kb", "10-100Kb", "100Kb-1Mb", ">1Mb", "Interchromosome"))]
# plotdata2[distance%in%c("<1Kb", "1-10Kb"), distance := "Local interactions"][distance%in%"Self-interaction", distance := "Local interactions"]
# plotdata2[, distance := factor(distance, levels=c("Local interactions", "10-100Kb", "100Kb-1Mb", ">1Mb", "Interchromosome"))]
plotdata2 = dcast.data.table(plotdata2, Condition+gene_type+distance+gene_id.R~gene_feature, value.var="mK", fun.aggregate=mean, fill=0)
plotdata2 = plotdata2[exon>0, `:=`(ratioIE = (intron)/(exon), ratioEI = (exon)/(intron))]
setkeyv(plotdata2, c("Condition", "gene_id.R"))
sel = plotdata2[is.finite(ratioEI), length(unique(distance)), by=key(plotdata2)][V1==length(unique(plotdata2[, distance]))]
plotdata2 = plotdata2[sel, nomatch=0]

tab = plotdata2[, list(N=.N/length(unique(distance))), by=c("Condition", "gene_type")]
gg = ggviolin(plotdata2, x="distance", y="log2(ratioEI)", fill="distance", facet.by="gene_type~Condition") +
   ggtitle("GRID-seq RNA processing rate (exon vs inton ~ distance)",
           subtitle="Reads are sum and normalised for each genic feature and averaged per interaction and then per gene.\nOnly RNAs with exonic and intronic reads within all distance bins are reported.") +
   stat_compare_means(ref.group="Gene body", hide.ns=TRUE, label = "p.signif") +
   scale_fill_manual("Distance to transcription site", values=rev(brewer.pal(7,"Blues"))) +
   scale_y_continuous("Exonic/Intronic reads (log2)") +
   scale_x_discrete("") +
   geom_text(data=tab, aes(label=paste("N =", N)), x=-Inf, y=Inf, hjust=-0.2, vjust=2, colour="black", inherit.aes=FALSE, parse=FALSE, size=4) +
   theme_bw() +
   theme(axis.text.x=element_text(angle=60, hjust=1, vjust=1, size=14),
         axis.text.y=element_text(size=14), legend.text=element_text(size=14),
         axis.title.y=element_text(size=14), legend.title=element_text(size=14),
         strip.text.x=element_text(size=14), strip.text.y=element_text(size=14),
         legend.position="bottom") +
   guides(fill = guide_legend(ncol = 7)) + facet_wrap(~gene_type, nrow=1) + geom_boxplot(outlier.colour=NA, width=0.2)
ggsave(gg, filename="plots/RNA-DNA_processing_rate_GRIDseq.png", units="in", height=8, width=12)
ggsave(gg, filename="plots/RNA-DNA_processing_rate_GRIDseq.pdf", height=8, width=12)

# plotdata2 = copy(plotdata)
# plotdata2[distance%in%c("<1Kb"), distance := "1-10Kb"][distance%in%"1-10Kb", distance := "<10Kb"]
# plotdata2[, distance := factor(distance, levels=c("Self-interaction", "<10Kb", "10-100Kb", "100Kb-1Mb", ">1Mb", "Interchromosome"))]
# plotdata2 = dcast.data.table(plotdata2, Condition+gene_type+distance+gene_id.R~gene_feature, value.var="mK", fun.aggregate=mean, fill=0)
# tab = plotdata2[Condition%in%c("Control1", "Control2", "1FA", "6h_ActD"), list(Num=length(unique(gene_id.R))), by=c("Condition", "gene_type", "distance")]
# 
# gg = ggplot(tab, aes(x=distance, y=Num, fill=gene_type)) +
#    geom_bar(stat="identity", position="dodge", col="black", lwd=0.5) +
#    scale_fill_brewer("Transcript feature", palette="Set1") +
#    scale_y_continuous("Number of interacting transcripts") +
#    scale_x_discrete("Distance to transcription site") +
#    facet_grid(gene_type~Condition, scales="free_y") +
#    theme_bw() + theme(axis.text.x=element_text(angle=60,hjust=1,vjust=1), legend.position="top")
# ggsave(gg, filename="plots/RNA-DNA_processing_rate_transcripts.png", units="in", height=8, width=14)
# 
# interactions[, score := 1]
# plotdata = dcast.data.table(interactions[!is.na(gene_type),], Condition+gene_id.R+gene_type+distance+gene_feature~Replicate, value.var="score", fun.aggregate=sum)
# plotdata = plotdata[rowSums(plotdata[, list(`1`, `2`, `3`)]>1)>=2,][, N := `1`+`2`+`3`]
# plotdata[, `:=`(`1`=NULL, `2`=NULL, `3`=NULL)]
# 
# # Pseudo-count for missing exonic/intronic reads
# plotdata = dcast.data.table(plotdata, Condition+gene_id.R+gene_type+distance~gene_feature, value.var="N",
#                             fun.aggregate="sum", fill=1)
# plotdata[, `:=`(five_prime_UTR = NULL, three_prime_UTR = NULL)]
# plotdata = melt.data.table(plotdata, id.vars=c("Condition", "gene_id.R", "gene_type", "distance"),
#                            measure.vars=c("exon", "intron"), variable.name="gene_feature", value.name="N")
# 
# # Add the exonic and intronic total width per RNA gene
# setkeyv(plotdata, c("gene_id.R", "gene_feature"))
# plotdata[exon_intron_width, width := as.numeric(i.width)]
# 
# plotdata[, K := N/width*1e3]
# 
# # Assign the facto levels
# plotdata[, `:=`(gene_feature = factor(gene_feature, levels=c("exon", "intron")),
#                 gene_type = factor(gene_type, levels=c("protein_coding", "lincRNA")))]
# # Collapse the replicates by condition
# plotdata = plotdata[, list(mK=mean(K)), by=c("Condition", "gene_type", "gene_id.R", "gene_feature", "distance")]
# 
# plotdata2 = copy(plotdata)
# plotdata2[distance%in%c("<1Kb"), distance := "1-10Kb"][distance%in%"1-10Kb", distance := "<10Kb"]
# plotdata2[, distance := factor(distance, levels=c("Self-interaction", "<10Kb", "10-100Kb", "100Kb-1Mb", ">1Mb", "Interchromosome"))]
# plotdata2 = dcast.data.table(plotdata2, Condition+gene_type+distance+gene_id.R~gene_feature, value.var="mK", fun.aggregate=mean, fill=0)
# plotdata2 = plotdata2[exon>0, `:=`(ratioIE = (intron)/(exon), ratioEI = (exon)/(intron))]
# setkeyv(plotdata2, c("Condition", "gene_id.R"))
# sel = plotdata2[is.finite(ratioEI), length(unique(distance)), by=key(plotdata2)][V1==length(unique(plotdata2[, distance]))]
# plotdata2 = plotdata2[sel, nomatch=0]
# 
# tab = plotdata2[Condition%in%c("Control1", "Control2", "1FA", "6h_ActD"), list(N=.N/length(unique(distance))), by=c("Condition", "gene_type")]
# gg = ggplot(plotdata2[Condition%in%c("Control1", "Control2", "1FA", "6h_ActD")], aes(x=distance, y=log2(ratioEI), fill=distance)) +
#    ggtitle("RNA processing rate (exon vs inton ~ distance)",
#            subtitle="Reads are sum and normalised for each genic feature and averaged per interaction and then per gene.\nOnly RNAs with exonic and intronic reads within all distance bins are reported.") +
#    geom_violin(position="dodge") +
#    geom_boxplot(width=0.2, outlier.colour=NA, fill="white") +
#    # scale_fill_brewer("Transcript feature", palette="Set1") +
#    scale_fill_manual("Distance to transcription site", values=rev(brewer.pal(7,"Blues"))) +
#    scale_y_continuous("Exonic/Intronic reads (log2)") +
#    scale_x_discrete("") +
#    facet_grid(gene_type~Condition) +
#    geom_text(data=tab, aes(label=paste("N =", N)), x=Inf, y=Inf, hjust=1.5, vjust=2, colour="black", inherit.aes=FALSE, parse=FALSE, size=4) +
#    theme_bw() +
#    theme(axis.text.x=element_text(angle=60, hjust=1, vjust=1, size=14),
#          axis.text.y=element_text(size=14), legend.text=element_text(size=14),
#          axis.title.y=element_text(size=14), legend.title=element_text(size=14),
#          strip.text.x=element_text(size=14), strip.text.y=element_text(size=14),
#          legend.position="bottom") +
#    guides(fill = guide_legend(ncol = 7))
# ggsave(gg, filename="plots/RNA-DNA_processing_rate_filtered.png", units="in", height=10, width=14)

# ggboxplot(plotdata2[Condition%in%c("Control1", "Control2", "1FA", "6h_ActD")], x="distance", y="log2(ratioEI)", fill="distance", add="box", facet.by="gene_type~Condition", ) +
#    stat_compare_means(ref.group="Self-interaction", hide.ns=TRUE, label = "p.signif") + scale_fill_manual("Distance to transcription site", values=rev(brewer.pal(7,"Blues"))) +
#    scale_y_continuous("Exonic/Intronic reads (log2)") +
#    scale_x_discrete("") + theme_bw() +
#    theme(axis.text.x=element_text(angle=60, hjust=1, vjust=1, size=14),
#          axis.text.y=element_text(size=14), legend.text=element_text(size=14),
#          axis.title.y=element_text(size=14), legend.title=element_text(size=14),
#          strip.text.x=element_text(size=14), strip.text.y=element_text(size=14),
#          legend.position="bottom") +
#    guides(fill = guide_legend(ncol = 7))
```

<!-- ![](plots/RNA-DNA_processing_rate.png) -->

### KCNQ1OT1 GSEA

```{r gsea, eval=FALSE}
require("fgsea")

known.interactions = c("Ascl2", "Tssc4", "Kcnq1", "Cdkn1c", "Slc22a18", "Phlda2", "Osbpl5")
known.interactions = list(`Literature derived` = genes.dt[gene_name%in%known.interactions, gene_id])

interactions = fread(paste0("gunzip -c RNA-DNA_counts_", gencode_version, "_level_1_2_3_AllOverlapsUnique_significant", fdr.th, "_extended_tsc_noDistance.tsv.gz"))

Kcnq1ot1.gene_id = genes.dt[gene_name%in%"Kcnq1ot1", gene_id]

subset = interactions[gene_id.R%in%Kcnq1ot1.gene_id & Condition%in%"1FA", .N, by="gene_id.D"][order(-N)]

subset = setNames(subset[, N], subset[, gene_id.D])

fgseaRes <- fgsea(pathways = known.interactions, 
                  stats = subset,
                  minSize=5,
                  maxSize=500,
                  nperm=10000)

pdf(file="plots/GSEA_Kcnq1ot1_1FA.pdf", width=10, height=6)
plotEnrichment(known.interactions[["Literature derived"]], subset) +
   ggtitle("1FA RADICL-seq Kcnq1ot1 known genomic targets") +
   scale_x_continuous("RADICL-seq interactions (ranked)") +
   scale_y_continuous("Enrichment Score") +
   geom_text(data=fgseaRes, aes(label=paste("p =", scales::scientific(fgseaRes$pval))), x=Inf, y=Inf, hjust=1.5, vjust=4, colour="black", inherit.aes=FALSE, parse=FALSE, size=4)
dev.off()

subset = interactions[gene_id.R%in%Kcnq1ot1.gene_id & Condition%in%"2FA", .N, by="gene_id.D"][order(-N)]

subset = setNames(subset[, N], subset[, gene_id.D])

fgseaRes <- fgsea(pathways = known.interactions, 
                  stats = subset,
                  minSize=5,
                  maxSize=500,
                  nperm=10000)

pdf(file="plots/GSEA_Kcnq1ot1_2FA.pdf", width=10, height=6)
plotEnrichment(known.interactions[["Literature derived"]], subset) +
   ggtitle("2FA RADICL-seq Kcnq1ot1 known genomic targets") +
   scale_x_continuous("RADICL-seq interactions (ranked)") +
   scale_y_continuous("Enrichment Score") +
   geom_text(data=fgseaRes, aes(label=paste("p =", scales::scientific(fgseaRes$pval))), x=Inf, y=Inf, hjust=1.5, vjust=4, colour="black", inherit.aes=FALSE, parse=FALSE, size=4)
dev.off()
```

<!-- ### TAD -->

<!-- ```{r TAD, eval=FALSE} -->
<!-- tad = fread("gunzip -c mESC-J1.mm10.chr10.25000.simpleNorm.tsv.gz", col.names=c("chr.R", "start.R", "end.R", "chr.D", "start.D", "end.D", "freq")) -->
<!-- tad[, `:=`(bin.R = paste(chr.R, start.R/25e3+1, sep="_"), bin.D = paste(chr.D, start.D/25e3+1, sep="_"))] -->
<!-- tad = tad[freq>0,] -->
<!-- tad.gr = with(tad, GRanges(chr.R, IRanges(start.R, end.R), strand="*", bin.R)) -->

<!-- interactions = readRDS(paste0("RData/RNA-DNA_overlaps_", gencode_version, "_level_1_2_3_AllOverlapsUnique.rds")) -->

<!-- interactions = interactions[gene_id.R%in%c(coding.sel, unlist(noncoding.sel[c("lincRNA", "antisense")])) & gene_id.D%in%c(coding.sel, unlist(noncoding.sel), c("Multigenic", "Intergenic")),] -->
<!-- interactions = interactions[!name%in%paste0("Hiseq190717_lane1_", c("ACAGTG", "ATCACG", "CAGATC", "GCCAAT")),] -->

<!-- setkey(interactions, name) -->
<!-- interactions = interactions[tab0[, list(Label, Condition, BiolReplicate)], nomatch=0] -->

<!-- interactions[, name := NULL] -->

<!-- interactions[Condition%in%"Control" & BiolReplicate%in%c(1,2), Condition := "Control1"] -->
<!-- interactions[Condition%in%"Control" & BiolReplicate%in%c(3,4), Condition := "Control2"] -->
<!-- interactions[Condition%in%"Control2", BiolReplicate := as.numeric(ifelse(BiolReplicate==3, 1, 2))] -->

<!-- setnames(interactions, "BiolReplicate", "Replicate") -->

<!-- interactions = interactions[Condition%in%"1FA",] -->

<!-- interactions.gr = with(interactions, GRanges(chr.R, IRanges(start.R, end.R), strand.R)) -->

<!-- save(tad, interactions, file="tmp.RData") -->

<!-- rm(tad, interactions) -->
<!-- gc() -->

<!-- overlaps = data.table(as.data.frame(findOverlaps(tad.gr, interactions.gr)), key="queryHits") -->

<!-- ``` -->

### Reproducibility

```{r reproducibility, eval=FALSE}
interactions = fread(paste0("gunzip -c RNA-DNA_biotypes_", gencode_version, "_level_1_2_3_AllOverlapsUnique_noIntergenic.tsv.gz"))

setkeyv(interactions, c("Condition", "Replicate", "gene_id.R", "gene_id.D"))
interactions = interactions[, .N, by=key(interactions)]
setnames(interactions, "N", "Count")

interactions[, `:=`(log2Count = log2(Count), Label = paste0(Condition, "_rep", Replicate))]

countTable = dcast.data.table(interactions, gene_id.R+gene_id.D~Label, value.var="Count", fill=0)

countTable = as.matrix(countTable[, grepl("FA", colnames(countTable)), with=FALSE])

countTable = countTable[rowSums(countTable >= 1) > 1,]

pdf("plots/RNA-DNA_plotCorrelation.pdf", width=16, height=16)
   pl = plotCorrelation2(countTable)
dev.off()

ountTable = dcast.data.table(interactions, gene_id.R+gene_id.D~Label, value.var="Count", fill=0)

countTable = as.matrix(countTable[, colnames(countTable)%in%paste0(c("1FA", "4h_ActD", "Control"), rep(c("_rep1", "_rep2", "_rep3"), each=3)), with=FALSE])

countTable = countTable[rowSums(countTable >= 1) > 1,]

pdf("plots/RNA-DNA_plotCorrelation_Suppl.pdf", width=16, height=16)
   pl = plotCorrelation2(countTable)
dev.off()

interactions = dcast(interactions, Condition+gene_id.R+gene_id.D~Replicate, value.var="Count", fill=NA)

count = interactions[Condition%in%"Control",]
c.max = ceiling(log2(max(count[, 4:ncol(count), with=FALSE], na.rm=TRUE)))

corVal = data.frame(value=cor(count[, list(`1`, `2`)], use="complete")[1,2])
gg12 = ggplot(count[!(is.na(`1`)|is.na(`2`))], aes(x=log2(`1`), y=log2(`2`))) +
   geom_hex(bins=100) +
   scale_x_continuous("Replicate 1 log2(count)", limits=c(-0.1,c.max)) +
   scale_y_continuous("Replicate 2 log2(count)", limits=c(-0.1,c.max)) +
   geom_text(data=corVal, aes(label=paste("R =", round(value, 3))), x=0, y=c.max, hjust=0, vjust=1, colour="black", inherit.aes=FALSE, parse=FALSE, size=4) +
   theme_bw() + coord_fixed()
png("plots/RNA-DNA_Control_replicates_reproducibility.png", units="in", height=5, width=6, res=300)
   ggarrange(gg12, ncol=1, nrow=1)
dev.off()

count = interactions[Condition%in%"1FA",]
c.max = ceiling(log2(max(count[, 4:ncol(count), with=FALSE], na.rm=TRUE)))

corVal = data.frame(value=cor(count[, list(`1`, `2`)], use="complete")[1,2])
gg12 = ggplot(count[!(is.na(`1`)|is.na(`2`))], aes(x=log2(`1`), y=log2(`2`))) +
   geom_hex(bins=100) +
   scale_x_continuous("Replicate 1 log2(count)", limits=c(-0.1,c.max)) +
   scale_y_continuous("Replicate 2 log2(count)", limits=c(-0.1,c.max)) +
   geom_text(data=corVal, aes(label=paste("R =", round(value, 3))), x=0, y=c.max, hjust=0, vjust=1, colour="black", inherit.aes=FALSE, parse=FALSE, size=4) +
   theme_bw() + coord_fixed()
corVal = data.frame(value=cor(count[, list(`1`, `3`)], use="complete")[1,2])
gg13 = ggplot(count[!(is.na(`1`)|is.na(`3`))], aes(x=log2(`1`), y=log2(`3`))) +
   geom_hex(bins=100) +
   scale_x_continuous("Replicate 1 log2(count)", limits=c(-0.1,c.max)) +
   scale_y_continuous("Replicate 3 log2(count)", limits=c(-0.1,c.max)) +
   geom_text(data=corVal, aes(label=paste("R =", round(value, 3))), x=0, y=c.max, hjust=0, vjust=1, colour="black", inherit.aes=FALSE, parse=FALSE, size=4) +
   theme_bw() + coord_fixed()
corVal = data.frame(value=cor(count[, list(`2`, `3`)], use="complete")[1,2])
gg23 = ggplot(count[!(is.na(`2`)|is.na(`3`))], aes(x=log2(`2`), y=log2(`3`))) +
   geom_hex(bins=100) +
   scale_x_continuous("Replicate 2 log2(count)", limits=c(-0.1,c.max)) +
   scale_y_continuous("Replicate 3 log2(count)", limits=c(-0.1,c.max)) +
   geom_text(data=corVal, aes(label=paste("R =", round(value, 3))), x=0, y=c.max, hjust=0, vjust=1, colour="black", inherit.aes=FALSE, parse=FALSE, size=4) +
   theme_bw() + coord_fixed()

png("plots/RNA-DNA_1FA_replicates_reproducibility.png", units="in", height=5, width=16, res=300)
   ggarrange(gg12, gg13, gg23, ncol=3, nrow=1)
dev.off()

count = interactions[Condition%in%"2FA",]
c.max = ceiling(log2(max(count[, 4:ncol(count), with=FALSE], na.rm=TRUE)))

corVal = data.frame(value=cor(count[, list(`1`, `2`)], use="complete")[1,2])
gg12 = ggplot(count[!(is.na(`1`)|is.na(`2`))], aes(x=log2(`1`), y=log2(`2`))) +
   geom_hex(bins=100) +
   scale_x_continuous("Replicate 1 log2(count)", limits=c(-0.1,c.max)) +
   scale_y_continuous("Replicate 2 log2(count)", limits=c(-0.1,c.max)) +
   geom_text(data=corVal, aes(label=paste("R =", round(value, 3))), x=0, y=c.max, hjust=0, vjust=1, colour="black", inherit.aes=FALSE, parse=FALSE, size=4) +
   theme_bw() + coord_fixed()
corVal = data.frame(value=cor(count[, list(`1`, `3`)], use="complete")[1,2])
gg13 = ggplot(count[!(is.na(`1`)|is.na(`3`))], aes(x=log2(`1`), y=log2(`3`))) +
   geom_hex(bins=100) +
   scale_x_continuous("Replicate 1 log2(count)", limits=c(-0.1,c.max)) +
   scale_y_continuous("Replicate 3 log2(count)", limits=c(-0.1,c.max)) +
   geom_text(data=corVal, aes(label=paste("R =", round(value, 3))), x=0, y=c.max, hjust=0, vjust=1, colour="black", inherit.aes=FALSE, parse=FALSE, size=4) +
   theme_bw() + coord_fixed()
corVal = data.frame(value=cor(count[, list(`2`, `3`)], use="complete")[1,2])
gg23 = ggplot(count[!(is.na(`2`)|is.na(`3`))], aes(x=log2(`2`), y=log2(`3`))) +
   geom_hex(bins=100) +
   scale_x_continuous("Replicate 2 log2(count)", limits=c(-0.1,c.max)) +
   scale_y_continuous("Replicate 3 log2(count)", limits=c(-0.1,c.max)) +
   geom_text(data=corVal, aes(label=paste("R =", round(value, 3))), x=0, y=c.max, hjust=0, vjust=1, colour="black", inherit.aes=FALSE, parse=FALSE, size=4) +
   theme_bw() + coord_fixed()

png("plots/RNA-DNA_2FA_replicates_reproducibility.png", units="in", height=5, width=16, res=300)
   ggarrange(gg12, gg13, gg23, ncol=3, nrow=1)
dev.off()

count = interactions[Condition%in%"4h_ActD",]
count[, `3` := NULL]
c.max = ceiling(log2(max(count[, 4:ncol(count), with=FALSE], na.rm=TRUE)))

corVal = data.frame(value=cor(count[, list(`1`, `2`)], use="complete")[1,2])
gg1 = ggplot(count[!(is.na(`1`)|is.na(`2`))], aes(x=log2(`1`), y=log2(`2`))) +
   ggtitle("ActD 4h") +
   geom_hex(bins=100) +
   scale_x_continuous("Replicate 1 log2(count)", limits=c(-0.1,c.max)) +
   scale_y_continuous("Replicate 2 log2(count)", limits=c(-0.1,c.max)) +
   geom_text(data=corVal, aes(label=paste("R =", round(value, 3))), x=0, y=c.max, hjust=0, vjust=1, colour="black", inherit.aes=FALSE, parse=FALSE, size=4) +
   theme_bw() + coord_fixed()

png("plots/RNA-DNA_ActD_replicates_reproducibility.png", units="in", height=5, width=6, res=300)
   ggarrange(gg1, ncol=1, nrow=1)
dev.off()

# count = melt.data.table(interactions, id.vars=c("Condition", "gene_id.R", "gene_id.D"), variable.name="Replicate", value.name="Count")
# count = dcast(count[Condition%in%c("4h_ActD", "6h_ActD") & Replicate%in%c(1, 2),], gene_id.R+gene_id.D~Condition+Replicate, value.var="Count", fill=NA)
# 
# c.max = ceiling(log2(max(count[, 4:ncol(count), with=FALSE], na.rm=TRUE)))
# 
# corVal = data.frame(value=cor(count[, list(`4h_ActD_1`, `6h_ActD_1`)], use="complete")[1,2])
# gg41.61 = ggplot(count[!(is.na(`4h_ActD_1`)|is.na(`6h_ActD_1`))], aes(x=log2(`4h_ActD_1`), y=log2(`6h_ActD_1`))) +
#    geom_hex(bins=100) +
#    scale_x_continuous("ActD 4h replicate 1 log2(count)", limits=c(-0.1,c.max)) +
#    scale_y_continuous("ActD 6h replicate 1 log2(count)", limits=c(-0.1,c.max)) +
#    geom_text(data=corVal, aes(label=paste("R =", round(value, 3))), x=0, y=c.max, hjust=0, vjust=1, colour="black", inherit.aes=FALSE, parse=FALSE, size=4) +
#    theme_bw() + coord_fixed()
# 
# corVal = data.frame(value=cor(count[, list(`4h_ActD_1`, `6h_ActD_2`)], use="complete")[1,2])
# gg41.62 = ggplot(count[!(is.na(`4h_ActD_1`)|is.na(`6h_ActD_2`))], aes(x=log2(`4h_ActD_1`), y=log2(`6h_ActD_2`))) +
#    geom_hex(bins=100) +
#    scale_x_continuous("ActD 4h replicate 1 log2(count)", limits=c(-0.1,c.max)) +
#    scale_y_continuous("ActD 6h replicate 2 log2(count)", limits=c(-0.1,c.max)) +
#    geom_text(data=corVal, aes(label=paste("R =", round(value, 3))), x=0, y=c.max, hjust=0, vjust=1, colour="black", inherit.aes=FALSE, parse=FALSE, size=4) +
#    theme_bw() + coord_fixed()
# 
# corVal = data.frame(value=cor(count[, list(`4h_ActD_2`, `6h_ActD_1`)], use="complete")[1,2])
# gg42.61 = ggplot(count[!(is.na(`4h_ActD_2`)|is.na(`6h_ActD_1`))], aes(x=log2(`4h_ActD_2`), y=log2(`6h_ActD_1`))) +
#    geom_hex(bins=100) +
#    scale_x_continuous("ActD 4h replicate 2 log2(count)", limits=c(-0.1,c.max)) +
#    scale_y_continuous("ActD 6h replicate 1 log2(count)", limits=c(-0.1,c.max)) +
#    geom_text(data=corVal, aes(label=paste("R =", round(value, 3))), x=0, y=c.max, hjust=0, vjust=1, colour="black", inherit.aes=FALSE, parse=FALSE, size=4) +
#    theme_bw() + coord_fixed()
# 
# corVal = data.frame(value=cor(count[, list(`4h_ActD_2`, `6h_ActD_2`)], use="complete")[1,2])
# gg42.62 = ggplot(count[!(is.na(`4h_ActD_2`)|is.na(`6h_ActD_2`))], aes(x=log2(`4h_ActD_2`), y=log2(`6h_ActD_2`))) +
#    geom_hex(bins=100) +
#    scale_x_continuous("ActD 4h replicate 2 log2(count)", limits=c(-0.1,c.max)) +
#    scale_y_continuous("ActD 6h replicate 2 log2(count)", limits=c(-0.1,c.max)) +
#    geom_text(data=corVal, aes(label=paste("R =", round(value, 3))), x=0, y=c.max, hjust=0, vjust=1, colour="black", inherit.aes=FALSE, parse=FALSE, size=4) +
#    theme_bw() + coord_fixed()
# 
# png("plots/RNA-DNA_ActD_cross-replicates_reproducibility.png", units="in", height=10, width=12, res=300)
#    ggarrange(gg41.61, gg41.62, gg42.61, gg42.62, ncol=2, nrow=2)
# dev.off()

interactions = interactions[, list(Condition, gene_id.R, gene_id.D, Count=rowSums(.SD, na.rm = TRUE)), .SDcols=c("1", "2", "3")]
count = dcast(interactions, gene_id.R+gene_id.D~Condition, value.var="Count", fill=NA)

s.max = ceiling(log2(max(count[, 3:ncol(count), with=FALSE], na.rm=TRUE)))

corVal = data.frame(value=cor(count[, list(`Control`, `1FA`)], use="complete")[1,2])
gg12 = ggplot(count[!(is.na(`Control`)|is.na(`1FA`))], aes(x=log2(`Control`), y=log2(`1FA`))) +
   geom_hex(bins=100) +
   scale_x_continuous("Control log2(count)", limits=c(-0.1,s.max)) +
   scale_y_continuous("1FA log2(count)",     limits=c(-0.1,s.max)) +
   geom_text(data=corVal, aes(label=paste("R =", round(value, 3))), x=0, y=s.max, hjust=0, vjust=1, colour="black", inherit.aes=FALSE, parse=FALSE, size=4) +
   theme_bw() + coord_fixed()
corVal = data.frame(value=cor(count[, list(`Control`, `2FA`)], use="complete")[1,2])
gg13 = ggplot(count[!(is.na(`Control`)|is.na(`2FA`))], aes(x=log2(`Control`), y=log2(`2FA`))) +
   geom_hex(bins=100) +
   scale_x_continuous("Control log2(count)", limits=c(-0.1,s.max)) +
   scale_y_continuous("2FA log2(count)",     limits=c(-0.1,s.max)) +
   geom_text(data=corVal, aes(label=paste("R =", round(value, 3))), x=0, y=s.max, hjust=0, vjust=1, colour="black", inherit.aes=FALSE, parse=FALSE, size=4) +
   theme_bw() + coord_fixed()
corVal = data.frame(value=cor(count[, list(`1FA`, `2FA`)], use="complete")[1,2])
gg23 = ggplot(count[!(is.na(`1FA`)|is.na(`2FA`))], aes(x=log2(`1FA`), y=log2(`2FA`))) +
   geom_hex(bins=100) +
   scale_x_continuous("1FA log2(count)", limits=c(-0.1,s.max)) +
   scale_y_continuous("2FA log2(count)", limits=c(-0.1,s.max)) +
   geom_text(data=corVal, aes(label=paste("R =", round(value, 3))), x=0, y=s.max, hjust=0, vjust=1, colour="black", inherit.aes=FALSE, parse=FALSE, size=4) +
   theme_bw() + coord_fixed()

png("plots/RNA-DNA_Reproducibility.png", units="in", height=5, width=16, res=300)
   ggarrange(gg12, gg13, gg23, ncol=3, nrow=1)
dev.off()

png("plots/RNA-DNA_Reproducibility_1FA-2FA.png", units="in", height=5, width=6, res=300)
   ggarrange(gg23, ncol=1, nrow=1)
dev.off()

count = data.table(t(t(count[, 3:ncol(count), with=FALSE]) / colSums(count[, 3:ncol(count), with=FALSE], na.rm=TRUE))*1e6)

s.min = floor(log2(min(count, na.rm=TRUE)))
s.max = ceiling(log2(max(count, na.rm=TRUE)))

corVal = data.frame(value=cor(count[, list(`Control`, `1FA`)], use="complete")[1,2])
gg12 = ggplot(count[!(is.na(`Control`)|is.na(`1FA`))], aes(x=log2(`Control`), y=log2(`1FA`))) +
   geom_hex(bins=100) +
   scale_x_continuous("Control log2(RPM)", limits=c(s.min,s.max)) +
   scale_y_continuous("1FA log2(RPM)",     limits=c(s.min,s.max)) +
   geom_text(data=corVal, aes(label=paste("R =", round(value, 3))), x=s.min, y=s.max, hjust=0, vjust=1, colour="black", inherit.aes=FALSE, parse=FALSE, size=4) +
   theme_bw() + coord_fixed()
corVal = data.frame(value=cor(count[, list(`Control`, `2FA`)], use="complete")[1,2])
gg13 = ggplot(count[!(is.na(`Control`)|is.na(`2FA`))], aes(x=log2(`Control`), y=log2(`2FA`))) +
   geom_hex(bins=100) +
   scale_x_continuous("Control log2(RPM)", limits=c(s.min,s.max)) +
   scale_y_continuous("2FA log2(RPM)",     limits=c(s.min,s.max)) +
   geom_text(data=corVal, aes(label=paste("R =", round(value, 3))), x=s.min, y=s.max, hjust=0, vjust=1, colour="black", inherit.aes=FALSE, parse=FALSE, size=4) +
   theme_bw() + coord_fixed()
corVal = data.frame(value=cor(count[, list(`1FA`, `2FA`)], use="complete")[1,2])
gg23 = ggplot(count[!(is.na(`1FA`)|is.na(`2FA`))], aes(x=log2(`1FA`), y=log2(`2FA`))) +
   geom_hex(bins=100) +
   scale_x_continuous("1FA log2(RPM)", limits=c(s.min,s.max)) +
   scale_y_continuous("2FA log2(RPM)", limits=c(s.min,s.max)) +
   geom_text(data=corVal, aes(label=paste("R =", round(value, 3))), x=s.min, y=s.max, hjust=0, vjust=1, colour="black", inherit.aes=FALSE, parse=FALSE, size=4) +
   theme_bw() + coord_fixed()

png("plots/RNA-DNA_Reproducibility_RPM.png", units="in", height=5, width=16, res=300)
   ggarrange(gg12, gg13, gg23, ncol=3, nrow=1)
dev.off()

png("plots/RNA-DNA_Reproducibility_1FA-2FA_RPM.png", units="in", height=5, width=6, res=300)
   ggarrange(gg23, ncol=1, nrow=1)
dev.off()
```

<!-- ### Comparison of filtering methods -->

<!-- ```{r examples, eval=FALSE} -->
<!-- if(!exists("fdr.th")) fdr.th = 0.05 -->

<!-- examples = data.table(name=c("Rn7sk", "Kcnq1ot1", "Firre", "Tfcp2l1")) -->
<!-- examples[, id := genes.dt[match(name, gene_name), gene_id]] -->
<!-- setkey(examples, id) -->

<!-- filtered = rbindlist(list(readRDS(paste0("RNA-DNA_counts_", gencode_version, "_level_1_2_3_AllOverlapsUniqueNoActD_Control_filtered.rds")), -->
<!--                           readRDS(paste0("RNA-DNA_counts_", gencode_version, "_level_1_2_3_AllOverlapsUniqueNoActD_1FA_filtered.rds")), -->
<!--                           readRDS(paste0("RNA-DNA_counts_", gencode_version, "_level_1_2_3_AllOverlapsUniqueNoActD_2FA_filtered.rds")))) -->
<!-- filtered = unique(filtered[Condition%in%c("1FA", "2FA"), list(Condition, gene_id.R, gene_id.D)]) -->
<!-- filtered = filtered[gene_id.R%in%examples[, id],][, Type := "Filtered"] -->
<!-- setkeyv(filtered, c("gene_id.R", "gene_id.D", "Condition")) -->

<!-- subset = readRDS(paste0("RData/RNA-DNA_noMultiOverlapsUniqueNoActD_counts_", gencode_version, "_level_1_2_3_pvalue_FDR.rds")) -->
<!-- subset = subset[Condition%in%c("1FA", "2FA"), list(Condition, gene_id.R, gene_id.D, Count, pvald, fdrd.by, Rjhd)] -->
<!-- subset = subset[gene_id.R%in%examples[, id],] -->
<!-- setkeyv(subset, c("gene_id.R", "gene_id.D", "Condition")) -->
<!-- subset = merge(subset, filtered, all.x=TRUE) -->
<!-- subset[is.na(Type), Type := "Raw"] -->

<!-- setkey(subset, gene_id.R) -->
<!-- subset = subset[examples,] -->

<!-- y.max = max(subset[, log2(Count)]) -->
<!-- data = dcast.data.table(subset, name+gene_id.D~Condition, value.var="Count") -->
<!-- tab = data[, list(cor(log2(`1FA`), log2(`2FA`), use="pairwise.complete.obs")), by="name"] -->
<!-- gg1 = ggplot(data, aes(x=log2(`1FA`), y=log2(`2FA`))) + -->
<!--    ggtitle("Raw") + -->
<!--    geom_point(alpha=0.5, size=1) + -->
<!--    scale_colour_brewer("", palette="Set1") + -->
<!--    scale_x_continuous("1FA - Average count (log2)", limits=c(0, y.max+0.5)) + -->
<!--    scale_y_continuous("2FA - Average count (log2)", limits=c(0, y.max+0.5)) + -->
<!--    geom_text(data=tab, aes(label=paste("R =", round(V1, 2))), x=y.max, y=0, hjust=1, vjust=0, colour="black", inherit.aes=FALSE, parse=FALSE, size=4) + -->
<!--    facet_wrap(~name, ncol=4) + coord_fixed() + theme_bw() -->

<!-- data = dcast.data.table(subset[Type%in%"Filtered"], name+gene_id.D~Condition, value.var="Count") -->
<!-- tab = data[, list(cor(log2(`1FA`), log2(`2FA`), use="pairwise.complete.obs")), by="name"] -->
<!-- gg2 = ggplot(data, aes(x=log2(`1FA`), y=log2(`2FA`))) + -->
<!--    ggtitle("Filtered") + -->
<!--    geom_point(alpha=0.5, size=1) + -->
<!--    scale_colour_brewer("", palette="Set1") + -->
<!--    scale_x_continuous("1FA - Average count (log2)", limits=c(0, y.max+0.5)) + -->
<!--    scale_y_continuous("2FA - Average count (log2)", limits=c(0, y.max+0.5)) + -->
<!--    geom_text(data=tab, aes(label=paste("R =", round(V1, 2))), x=y.max, y=0, hjust=1, vjust=0, colour="black", inherit.aes=FALSE, parse=FALSE, size=4) + -->
<!--    facet_wrap(~name, ncol=4) + coord_fixed() + theme_bw() -->

<!-- data = dcast.data.table(subset[fdrd.by<fdr.th], name+gene_id.D~Condition, value.var="Count") -->
<!-- tab = data[, list(cor(log2(`1FA`), log2(`2FA`), use="pairwise.complete.obs")), by="name"] -->
<!-- gg3 = ggplot(data, aes(x=log2(`1FA`), y=log2(`2FA`))) + -->
<!--    ggtitle("Significant (FDR<0.05)") + -->
<!--    geom_point(alpha=0.5, size=1) + -->
<!--    scale_colour_brewer("", palette="Set1") + -->
<!--    scale_x_continuous("1FA - Average count (log2)", limits=c(0, y.max+0.5)) + -->
<!--    scale_y_continuous("2FA - Average count (log2)", limits=c(0, y.max+0.5)) + -->
<!--    geom_text(data=tab, aes(label=paste("R =", round(V1, 2))), x=y.max, y=0, hjust=1, vjust=0, colour="black", inherit.aes=FALSE, parse=FALSE, size=4) + -->
<!--    facet_wrap(~name, ncol=4) + coord_fixed() + theme_bw() -->

<!-- ggarrange(gg1, gg2, gg3, ncol=1, nrow=3) -->



<!-- data = dcast.data.table(subset, gene_id.R+gene_id.D~Condition, value.var="Rjhd") -->
<!-- gg1 = ggplot(data, aes(x=`1FA`, y=`2FA`)) + -->
<!--    ggtitle("Raw") + -->
<!--    geom_point(alpha=0.5, size=1) + -->
<!--    scale_colour_brewer("", palette="Set1") + -->
<!--    scale_x_continuous("1FA log2(obs/exp)") + -->
<!--    scale_y_continuous("2FA log2(obs/exp)") + -->
<!--    facet_wrap(~gene_id.R, ncol=4) + coord_fixed() + theme_bw() -->

<!-- data = dcast.data.table(subset[Type%in%"Filtered"], gene_id.R+gene_id.D~Condition, value.var="Rjhd") -->
<!-- gg2 = ggplot(data, aes(x=`1FA`, y=`2FA`)) + -->
<!--    ggtitle("Filtered") + -->
<!--    geom_point(alpha=0.5, size=1) + -->
<!--    scale_colour_brewer("", palette="Set1") + -->
<!--    scale_x_continuous("1FA log2(obs/exp)") + -->
<!--    scale_y_continuous("2FA log2(obs/exp)") + -->
<!--    facet_wrap(~gene_id.R, ncol=4) + coord_fixed() + theme_bw() -->

<!-- ggarrange(gg1, gg2, ncol=1, nrow=2) -->

<!-- subset = readRDS(paste0("RData/RNA-DNA_noMultiOverlapsUniqueNoActD_counts_", gencode_version, "_level_1_2_3_pvalue_FDR.rds")) -->
<!-- subset = subset[Condition%in%c("1FA", "2FA")] -->
<!-- setkeyv(subset, c("gene_id.R", "gene_id.D", "Condition")) -->
<!-- subset = subset[gene_id.R%in%examples[, id], list(Count = mean(Count), Rjh = mean(Rjh), fdr = any(fdr<0.01)), by=key(subset)] -->

<!-- tab = data.table(unique(subset[, list(gene_id.R, gene_id.D)]), type="Filtered") -->
<!-- setkeyv(tab, c("gene_id.R", "gene_id.D")) -->
<!-- data = dcast(interactions[Condition%in%c("1FA", "2FA")], gene_id.R+gene_id.D~Condition, value.var="N", fun.aggregate=mean, fill=0) -->
<!-- setkeyv(data, c("gene_id.R", "gene_id.D")) -->
<!-- data = merge(data, tab, all.x=TRUE) -->
<!-- data[gene_id.R==gene_id.D, type := "Filtered"][is.na(type), type := "Raw"] -->
<!-- setkey(data, gene_id.R) -->
<!-- data = merge(data, examples, by.x=key(data), by.y=key(examples), all.x=TRUE) -->

<!-- ggplot(data, aes(x=log2(`1FA`), y=log2(`2FA`), col=type)) + -->
<!--    geom_point(alpha=0.5, size=1) + -->
<!--    scale_colour_brewer("", palette="Set1") + -->
<!--    scale_x_continuous("1FA - Average count (log2)") + -->
<!--    scale_y_continuous("2FA - Average count (log2)") + -->
<!--    facet_wrap(~gene_id.R, ncol=2) + coord_fixed() + theme_bw() -->

<!-- data = dcast(subset, gene_id.R+gene_id.D~Condition, value.var=c("Count", "Rjh", "fdr"), fill=0) -->
<!-- tab = data.table(unique(subset[, list(gene_id.R, gene_id.D)]), type="Filtered") -->


<!-- interactions[, group := factor("Raw", levels=c("Raw", "Filtered", "Significant"))][!is.na(fdr), group := "Filtered"][!is.na(fdr) & fdr<0.01, group := "Significant"] -->

<!-- setkeyv(interactions, c("gene_id.R", "gene_id.D", "Condition")) -->
<!-- interactions = interactions[, list(N=mean(N), group=tail(.SD[, group], 1)), by=key(interactions)] -->

<!-- interactions = dcast.data.table(interactions, gene_id.R+gene_id.D~Condition, value.var=c("N", "group")) -->

<!-- ggplot(interactions, aes(x=log2(N_1FA), y=log2(N_2FA), col=group_1FA)) + geom_point() + facet_wrap(~gene_id.R, nrow=2) -->
<!-- ``` -->

### Circos representation

```{r circosBiotypes, eval=FALSE}
interactions = fread(paste0("gunzip -c RNA-DNA_biotypes_", gencode_version, "_level_1_2_3_AllOverlapsUnique_noIntergenic.tsv.gz"))

setkeyv(interactions, c("Condition", "annotation.R", "annotation.D"))

# subset = interactions[gene_id.R!=gene_id.D, .N, by=key(interactions)]
subset = interactions[, .N, by=key(interactions)]

subset = subset[Condition%in%"1FA" & N>=1000,][order(N)]

subset[, label := factor(paste(annotation.R, annotation.D, sep=" - "), levels=c(paste(annotation.R, annotation.D, sep=" - ")))]

gg = ggbarplot(subset, x="label", y="N", fill="Condition", palette="npg", color="black", sort.val="asc") +
   scale_x_discrete("") + scale_y_log10("Interaction count by RNA and DNA types", breaks=10^c(0:8), labels=scales::comma, position = "right") +
   coord_flip() + guides(fill=FALSE) +
   theme(panel.grid.major.y=element_line(size=0.1, color="grey"),
         panel.grid.major.x=element_line(size=0.25, color="grey"),
         panel.grid.minor.x=element_line(size=0.1, color="grey"),
         axis.text.x=element_text(angle=60, hjust=0, vjust=0.5))
ggsave(gg, filename="plots/RNA-DNA_interactions_by_allBiotypes_1FA.pdf", width=8, height=12)

subset = interactions[gene_id.R!=gene_id.D, .N, by=key(interactions)]

subset = subset[Condition%in%"1FA" & N>=1000,][order(N)]

subset[, label := factor(paste(annotation.R, annotation.D, sep=" - "), levels=c(paste(annotation.R, annotation.D, sep=" - ")))]

gg = ggbarplot(subset, x="label", y="N", fill="Condition", palette="npg", color="black", sort.val="asc") +
   scale_x_discrete("") + scale_y_log10("Interaction count by RNA and DNA types", breaks=10^c(0:8), labels=scales::comma, position = "right") +
   coord_flip() + guides(fill=FALSE) +
   theme(panel.grid.major.y=element_line(size=0.1, color="grey"),
         panel.grid.major.x=element_line(size=0.25, color="grey"),
         panel.grid.minor.x=element_line(size=0.1, color="grey"),
         axis.text.x=element_text(angle=60, hjust=0, vjust=0.5))
ggsave(gg, filename="plots/RNA-DNA_interactions_by_allBiotypes_1FA_noSelf.pdf", width=8, height=12)
```

```{r circosBiotypesFDR, eval=FALSE}
# noncoding.names = noncoding.names[c(2, 3, 1, 4, 5, 6, 7, 8, 9)]

interactions = fread(paste0("gunzip -c RNA-DNA_counts_", gencode_version, "_level_1_2_3_AllOverlapsUnique_significant", fdr.th, "_extended_tsc_noDistance.tsv.gz"))

# interactions = readRDS(paste0("RData/RNA-DNA_noMultiOverlapsUniqueNoActD_counts_", gencode_version, "_level_1_2_3_pvalue_FDR.rds"))
# subset = interactions[fdr.by<fdr.th,]

# # Remove self-interactions
# subset = subset[gene_id.R!=gene_id.D,]
# 
# setkey(genes.dt, gene_id)
# 
# setkey(subset, gene_id.R)
# subset[genes.dt, biotypeRNA := gene_type]
# setkey(subset, gene_id.D)
# subset[genes.dt, biotypeDNA := gene_type]
setkeyv(interactions, c("annotation.R", "annotation.D"))

for( cond in c("1FA", "2FA") ){
   myset = interactions[Condition%in%cond, .N, by=key(interactions)]
   myset = myset[N>100]
   
   lab.order = c("protein_coding", "long_ncRNA", "ncRNA", "other")
   myset[, `:=`(biotype.R = factor(paste0("R_", annotation.R), levels=paste0("R_", lab.order)),
                biotype.D = factor(paste0("D_", annotation.D), levels=paste0("D_", lab.order)))]
   # myset[annotation.R%in%c("protein_coding", noncoding.names),
   #       annotation.R := factor(paste0("R_", annotation.R), levels=paste0("R_", c("protein_coding", noncoding.names)))]
   # myset[annotation.D%in%c("protein_coding", noncoding.names),
   #       annotation.D := factor(paste0("D_", annotation.D), levels=paste0("D_", c("protein_coding", noncoding.names)))]
   setkeyv(myset, c("biotype.R", "biotype.D"))
   
   myset = as.matrix(data.frame(dcast.data.table(myset, biotype.R~biotype.D, value.var="N"), row.names=1))
   myset = apply(myset, 2, rev)
   # myset = t(apply(myset, 1, rev))
   # colnames(myset) = gsub("\\.", "", colnames(myset))
   
   # grid.col = setNames(rainbow(length(unlist(dimnames(myset)))), union(rownames(myset), colnames(myset)))
   # matrix.col = matrix(rep(grid.col[match(rownames(myset), names(grid.col))], each=6), byrow=TRUE, 8, 6)
   palette = get_palette(palette="npg", 6)
   all = c(setNames(palette[1:length(unique(gsub("[DR]_", "",
                                                 c(rownames(myset), colnames(myset)))))],
                    unique(gsub("[DR]_", "R_", c(rownames(myset), colnames(myset))))),
           setNames(palette[1:length(unique(gsub("[DR]_", "",
                                                 c(rownames(myset), colnames(myset)))))],
                    unique(gsub("[DR]_", "D_", c(rownames(myset), colnames(myset))))))

   png(paste0("plots/Circos_Biotype2_above500_", cond, "_FDR.png"), units="in", type="cairo", res=300, width=16, height=16)
      circos.clear()
      circos.par(gap.after = c(rep(1, nrow(myset)-1), 8, rep(1, ncol(myset)-1), 8),
                 track.height = 0.05, track.margin = c(0.05, 0), start.degree = 175)
      chordDiagram(myset, annotationTrack = c("grid"),
                   preAllocateTracks = 4, grid.col = all, link.lwd = 0.5, link.lty = 1, link.border = "grey", transparency = 0.2)
      circos.track(track.index = 4, panel.fun = function(x, y) {
         sector.index = get.cell.meta.data("sector.index")
         xlim = get.cell.meta.data("xlim")
         ylim = get.cell.meta.data("ylim")
         circos.text(mean(xlim), ylim[1] + 1.2, gsub("[DR]_", "", sector.index),
                     facing = "clockwise", niceFacing = TRUE, adj = c(0.2, 0.5), cex = 1)
      }, bg.border = NA)
      highlight.sector(colnames(myset), track.index = 1, col = "dodgerblue3", text="DNA", text.col="white", niceFacing = TRUE)
      highlight.sector(rownames(myset), track.index = 1, col = "firebrick3",  text="RNA", text.col="white", niceFacing = TRUE)
      title(cond)
   dev.off()
   
   myset = interactions[Condition%in%cond, .N, by=key(interactions)]
   myset = myset[N>10]
   
   # myset = myset[!biotypeRNA%in%"protein_coding"]
   
   myset[, `:=`(biotype.R = factor(paste0("R_", annotation.R), levels=paste0("R_", unique(annotation.R))),
                biotype.D = factor(paste0("D_", annotation.D), levels=paste0("D_", unique(annotation.D))))]
   # myset[annotation.R%in%c("protein_coding", noncoding.names),
   #       annotation.R := factor(paste0("R_", annotation.R), levels=paste0("R_", c("protein_coding", noncoding.names)))]
   # myset[annotation.D%in%c("protein_coding", noncoding.names),
   #       annotation.D := factor(paste0("D_", annotation.D), levels=paste0("D_", c("protein_coding", noncoding.names)))]
   setkeyv(myset, c("biotype.R", "biotype.D"))
   
   myset = as.matrix(data.frame(dcast.data.table(myset, biotype.R~biotype.D, value.var="N"), row.names=1))
   myset = apply(myset, 2, rev)
   # myset = t(apply(myset, 1, rev))
   # colnames(myset) = gsub("\\.", "", colnames(myset))
   
   # grid.col = setNames(rainbow(length(unlist(dimnames(myset)))), union(rownames(myset), colnames(myset)))
   # matrix.col = matrix(rep(grid.col[match(rownames(myset), names(grid.col))], each=6), byrow=TRUE, 8, 6)
   palette = get_palette(palette="npg", 6)
   all = c(setNames(palette[1:length(unique(gsub("[DR]_", "",
                                                 c(rownames(myset), colnames(myset)))))],
                    unique(gsub("[DR]_", "R_", c(rownames(myset), colnames(myset))))),
           setNames(palette[1:length(unique(gsub("[DR]_", "",
                                                 c(rownames(myset), colnames(myset)))))],
                    unique(gsub("[DR]_", "D_", c(rownames(myset), colnames(myset))))))

   png(paste0("plots/Circos_Biotype2_above10_", cond, "_FDR.png"), units="in", type="cairo", res=300, width=16, height=16)
      circos.clear()
      circos.par(gap.after = c(rep(1, nrow(myset)-1), 8, rep(1, ncol(myset)-1), 8),
                 track.height = 0.05, track.margin = c(0.05, 0), start.degree = 175)
      chordDiagram(myset, annotationTrack = c("grid"),
                   preAllocateTracks = 4, grid.col = all, link.lwd = 0.5, link.lty = 1, link.border = "grey", transparency = 0.2)
      circos.track(track.index = 4, panel.fun = function(x, y) {
         sector.index = get.cell.meta.data("sector.index")
         xlim = get.cell.meta.data("xlim")
         ylim = get.cell.meta.data("ylim")
         circos.text(mean(xlim), ylim[1] + 1.2, gsub("[DR]_", "", sector.index),
                     facing = "clockwise", niceFacing = TRUE, adj = c(0.2, 0.5), cex = 1)
      }, bg.border = NA)
      highlight.sector(colnames(myset), track.index = 1, col = "dodgerblue3", text="DNA", text.col="white", niceFacing = TRUE)
      highlight.sector(rownames(myset), track.index = 1, col = "firebrick3",  text="RNA", text.col="white", niceFacing = TRUE)
      title(cond)
   dev.off()
}
```

### Comparison with RAP data

```{r rap, eval=FALSE}
# bed = fread("gunzip -c ~/Downloads/wgEncodeLicrHistoneEse14H3k04me3ME0129olaStdPk.broadPeak.gz")
# bed = fread("gunzip -c ~/Downloads/wgEncodeSydhHistEse14H3k04me3StdPk.narrowPeak.gz")
# bed.gr = with(bed, GRanges(V1, IRanges(V2, V3), strand="*"))

# interactions = readRDS(paste0("RData/RNA-DNA_overlaps_", gencode_version, "_level_1_2_3_AllOverlapsUnique.rds"))
# 
# interactions = interactions[gene_id.R%in%coding.sel & gene_id.D%in%c(coding.sel, unlist(noncoding.sel), c("Multigenic", "Intergenic")),]
# interactions = interactions[!name%in%paste0("Hiseq190717_lane1_", c("ACAGTG", "ATCACG", "CAGATC", "GCCAAT")),]
# 
# setkey(interactions, name)
# interactions = interactions[tab0[, list(Label, Condition, BiolReplicate)], nomatch=0]
# 
# interactions = interactions[Condition%in%"1FA",]
# interactions[, name := NULL]
# 
# interactions = with(interactions, GRanges(chr.R, IRanges(start.R, end.R), strand.R, gene_id.R, gene_id.D))
# 

interactions = fread(paste0("gunzip -c RNA-DNA_biotypes_", gencode_version, "_level_1_2_3_AllOverlapsUnique.tsv.gz"))

genes.pc = unique(tx.regions.dt[annotation%in%"protein_coding", gene_id])
genes.pc = genes[genes$gene_id%in%genes.pc,]
# genes.pc = genes[genes$gene_id%in%coding.sel,]
# # genes.pc = genes
genes.pc.counts = countOverlaps(genes.pc, with(interactions, GRanges(chr.R, IRanges(start.R, end.R), strand.R)))
genes.pc.tpm = countToTPM(genes.pc.counts, genes.pc)
# # genes.pc = genes.pc[countToTPM(genes.pc.counts, genes.pc)>=20,]
# genes.pc = genes.pc[countOverlaps(flank(genes.pc, 5e3, start=TRUE, both=TRUE), bed.gr)>=1]


# malat1 = interactions[Condition%in%c("1FA", "2FA") & gene_id.R%in%"ENSMUSG00000092341.2",]
malat1 = interactions[Condition%in%c("1FA", "2FA") & gene_id.R%in%"ENSMUSG00000092341.2",]

# random = interactions[Condition%in%"1FA" & !gene_id.R%in%c("ENSMUSG00000092341.2", "Intergenic"),]
# random = random[sample(1:nrow(random), nrow(malat1)),]
# 
# coding = interactions[Condition%in%"1FA" & annotation.R%in%"protein_coding",]
# coding = coding[sample(1:nrow(coding), nrow(malat1)),]

# malat1 = malat1[, N := 1]
# random = random[, N := 1]
# coding = coding[, N := 1]

# malat1 = malat1[, .N, by=c("chr.D", "start.D", "end.D")]
# random = random[, .N, by=c("chr.D", "start.D", "end.D")]
# coding = coding[, .N, by=c("chr.D", "start.D", "end.D")]

malat1 = sort.GenomicRanges(with(malat1[!annotation.D%in%c("Multigenic", "Intergenic")],
                                 GRanges(chr.D, IRanges(start.D, end.D), strand="*")), ignore.strand=TRUE)
# malat1 = sort.GenomicRanges(with(malat1,
#                                  GRanges(chr.D, IRanges(start.D, end.D), strand="*")), ignore.strand=TRUE)
# random = sort.GenomicRanges(with(random, GRanges(chr.D, IRanges(start.D, end.D), strand="*", score=N)), ignore.strand=TRUE)
# coding = sort.GenomicRanges(with(coding, GRanges(chr.D, IRanges(start.D, end.D), strand="*", score=N)), ignore.strand=TRUE)

# export.bedGraph(resize(malat1, width=round(rnorm(length(malat1), mean=27.31745, sd=0.8995841)), fix="center"), "MACS/malat1_1FA.bed")
export.bedGraph(resize(malat1, width=27, fix="center"), "MACS/malat1_1FA.bed")
# export.bed(random, "MACS/random_1FA.bed")
# export.bed(coding, "MACS/coding_1FA.bed")

# system("bash runMe.sh")

macs.cols = c("chr", "start", "end", "name", "score", ".", "fc", "pval", "qval")

rap.fdr = 0.2
rap.tpm = 10

rap = fread("MACS/malat1_RAP_peaks.broadPeak", col.names=macs.cols)
# rap = rap[V9>=(-log10(0.05)),][V1%in%paste0("chr", c(1:19, "X", "Y"))]
rap = rap[qval>(-log10(rap.fdr)),][chr%in%paste0("chr", c(1:19, "X", "Y"))]

rap.gr = with(rap, GRanges(chr, IRanges(start, end), strand="*"))

rap.genes = genes.pc[overlapsAny(genes.pc[genes.pc.tpm>=rap.tpm,], rap.gr),]

rad = fread("MACS/malat1_1FA_peaks.broadPeak", col.names=macs.cols)
# rad = rad[V9>=(-log10(0.05)),][V1%in%paste0("chr", c(1:19, "X", "Y"))]
rad = rad[qval>(-log10(rap.fdr)),][chr%in%paste0("chr", c(1:19, "X", "Y"))]

rad.gr = with(rad, GRanges(chr, IRanges(start, end), strand="*"))

rad.genes = genes.pc[overlapsAny(genes.pc[genes.pc.tpm>=rap.tpm,], rad.gr),]

table(overlapsAny(rad.genes, rap.genes))
table(overlapsAny(rap.genes, rad.genes))

vennlist = list(RAP=rap.genes$gene_id, RADICL=rad.genes$gene_id)
venn.plot.P = venn.diagram(vennlist, filename=NULL, fill=c("dodgerblue3", "firebrick3")[1:length(vennlist)],
                               resolution=300, imagetype="png", units="in", height=10, width=10,
                               print.mode = c("raw", "percent"), fontfamily = "Helvetica", main.fontfamily = "Helvetica", cat.fontfamily = "Helvetica")
pdf(file=paste0("plots/RNA-DNA_RAP_Malat1_1FA_2FA_FDR", rap.fdr, "_TPM", rap.tpm, ".pdf"), width=6, height=6)
   grid.arrange(gTree(children=venn.plot.P), ncol=1)
dev.off()

rap = fread("MACS/malat1_RAP_peaks.broadPeak")
rap = rap[V9>=(-log10(0.01)),][V1%in%paste0("chr", c(1:19, "X", "Y"))]

rap.gr = with(rap, GRanges(V1, IRanges(V2, V3), strand="*"))

rap.genes = genes.pc[overlapsAny(genes.pc, rap.gr),]

rad = fread("MACS/malat1_1FA_peaks.broadPeak")
rad = rad[V9>=(-log10(0.01)),][V1%in%paste0("chr", c(1:19, "X", "Y"))]

rad.gr = with(rad, GRanges(V1, IRanges(V2, V3), strand="*"))

rad.genes = genes.pc[overlapsAny(genes.pc, rad.gr),]

table(overlapsAny(rad.genes, rap.genes))
table(overlapsAny(rap.genes, rad.genes))

vennlist = list(RAP=rap.genes$gene_id, RADICL=rad.genes$gene_id)
venn.plot.P = venn.diagram(vennlist, filename=NULL, fill=c("dodgerblue3", "firebrick3")[1:length(vennlist)],
                               resolution=300, imagetype="png", units="in", height=10, width=10,
                               print.mode = c("raw", "percent"), fontfamily = "Helvetica", main.fontfamily = "Helvetica", cat.fontfamily = "Helvetica")
pdf(filename=paste0("plots/RNA-DNA_RAP_Malat1_1FA_overlaps_at_FDR_0.01.pdf"), width=6, height=6)
   grid.arrange(gTree(children=venn.plot.P), ncol=1)
dev.off()

##### Test

rap.fdr = 0.01
rad.fdr = 0.05
rap.tpm = 5

test = as.data.table(cbind(c(0.01, 0.05, 0.1), rep(c(1,5,10,20), each=length(c(0.01, 0.05, 0.1)))))

for( i in 1:nrow(test) ){
   rap.fdr = 0.01
   rad.fdr = 0.05
   rap.tpm = test[i, V2]
   
   rap = fread("MACS/malat1_RAP_peaks.broadPeak", col.names=macs.cols)
   # rap = rap[V9>=(-log10(0.05)),][V1%in%paste0("chr", c(1:19, "X", "Y"))]
   rap = rap[qval>(-log10(rap.fdr)),][chr%in%paste0("chr", c(1:19, "X", "Y"))]
   
   rap.gr = with(rap, GRanges(chr, IRanges(start, end), strand="*"))
   
   rap.genes = genes.pc[genes.pc.tpm>=rap.tpm,][overlapsAny(genes.pc[genes.pc.tpm>=rap.tpm,], rap.gr),]
   
   rad = fread("MACS/malat1_1FA_peaks.broadPeak", col.names=macs.cols)
   # rad = rad[V9>=(-log10(0.05)),][V1%in%paste0("chr", c(1:19, "X", "Y"))]
   rad = rad[qval>(-log10(rad.fdr)),][chr%in%paste0("chr", c(1:19, "X", "Y"))]
   
   rad.gr = with(rad, GRanges(chr, IRanges(start, end), strand="*"))
   
   rad.genes = genes.pc[genes.pc.tpm>=rap.tpm,][overlapsAny(genes.pc[genes.pc.tpm>=rap.tpm,], rad.gr),]
   
   tab = table(overlapsAny(rap.genes, rad.genes))
   print(paste(rap.fdr, rad.fdr, rap.tpm, length(rap.gr), length(rad.gr), (tab/sum(tab))[2]))
}


for( rap.tpm in c(1, 2, 5, 10, 20) ){
   rap = fread("MACS/test_rap_peaks.narrowPeak", select=1:3)
   rap.gr = with(rap, GRanges(V1, IRanges(V2, V3), strand="*"))
   rap.genes = genes.pc[genes.pc.tpm>=rap.tpm,][overlapsAny(genes.pc[genes.pc.tpm>=rap.tpm,], rap.gr),]
   
   rad = fread("MACS/test_peaks.narrowPeak", select=1:3)
   rad.gr = with(rad, GRanges(V1, IRanges(V2, V3), strand="*"))
   rad.genes = genes.pc[genes.pc.tpm>=rap.tpm,][overlapsAny(genes.pc[genes.pc.tpm>=rap.tpm,], rad.gr),]

      tab = table(overlapsAny(rap.genes, rad.genes))
   print(paste(rap.tpm, length(rap.gr), length(rad.gr), length(genes.pc[genes.pc.tpm>=rap.tpm,]), length(rap.genes), length(rad.genes), (tab/sum(tab))[2]))
}
```

```{r rapCirclize, eval=FALSE}
macs.cols = c("chr", "start", "end", "name", "score", ".", "fc", "pval", "qval")
rap = fread("MACS/malat1_RAP_peaks.broadPeak", col.names=macs.cols)
# rap.gr = with(rap, GRanges(chr, IRanges(start, end), strand="*", fc, pval, qval))
rap = rap[, list(chr, start, end, value=log10(score))]

rad = fread("MACS/malat1_1FA_peaks.broadPeak", col.names=macs.cols)
# rad.gr = with(rad, GRanges(chr, IRanges(start, end), strand="*", fc, pval, qval))
rad = rad[, list(chr, start, end, value=log10(score))]

cytoband = read.cytoband(species="mm10")
cytoband_df = cytoband$df
chromosome = cytoband$chromosome

circos.clear()
circos.par(start.degree = 90, gap.degree=c(rep(1, length(chromosome)-1), 5))
circos.initializeWithIdeogram(plotType = c("axis", "labels"), species = "mm10")
circos.genomicTrack(rap, panel.fun = function(region, value, ...) {
                       circos.genomicLines(region, value, type = "h", col="blue")
                    })
circos.genomicTrack(rad, track.index = 2, panel.fun = function(region, value, ...) {
                       circos.genomicLines(region, value, type = "h", col="red")
                    })

circos.clear()
circos.par(start.degree = 90)
circos.initializeWithIdeogram(plotType = c("labels"), species = "mm10", sort.chr = TRUE)
circos.track(ylim = c(0, 1), track.height = 0.4, bg.border=NA)
circos.genomicIdeogram(species = "mm10")
circos.genomicDensity(rap, col="#1874cd80", border="black", lwd=0.5, track.index=2, bg.border=NA)
circos.genomicDensity(rad, col="#cd262660", border="black", lwd=0.5, track.index=2, bg.border=NA)
```

### Intracromosomal and Interchromosomal

```{r ratioIntraInter, eval=FALSE}
interactions = fread(paste0("gunzip -c RNA-DNA_biotypes_", gencode_version, "_level_1_2_3_AllOverlapsUnique.tsv.gz"))

interactions[, interactionType := ifelse(chr.R==chr.D, "Intrachromosomal", "Interchromosomal")]

setkeyv(interactions, c("gene_id.R", "interactionType", "annotation.R", "Condition", "Replicate"))
subset = interactions[gene_id.R!=gene_id.D & !gene_id.R%in%c("Multigenic", "Intergenic"), .N, by=key(interactions)]

subset = dcast.data.table(subset, Condition+Replicate+annotation.R+gene_id.R~interactionType, value.var="N", fill=1)
subset[, `:=`(Sum = log2(Interchromosomal+Intrachromosomal), Ratio = log2(Interchromosomal/Intrachromosomal))]

# gg1 = ggplot(subset[Interchromosomal+Intrachromosomal>=10], aes(x=Biotype, y=Ratio, fill=as.factor(Replicate))) +
#    ggtitle("Interchromosomal vs intrachromosomal reads ratio", subtitle="A pseudo-count of 1 has been added to those interactions with 0 count in either intrachromosomal or interchromosomal fraction") +
#    geom_boxplot(outlier.size=0.5) + 
#    facet_grid(~Condition) +
#    scale_x_discrete("") +
#    scale_y_continuous("log2(Inter/Intra)") +
#    scale_fill_brewer("Replicate", palette="Set1") +
#    theme_bw() +
#    theme(legend.position="right", axis.text.x=element_text(angle = 45, hjust = 1))
# ggsave(gg1, filename="plots/RNA-DNA_Inter_vs_Intrachromosomal_interactions.png", units="in", width=12, height=4, points=10)
# 
# gg2 = ggplot(subset[Interchromosomal+Intrachromosomal>=10], aes(x=Biotype2, y=Ratio, fill=as.factor(Replicate))) +
#       ggtitle("Interchromosomal vs intrachromosomal reads ratio", subtitle="A pseudo-count of 1 has been added to those interactions with 0 count in either intrachromosomal or interchromosomal fraction\nlincRNA, processed_transcript, antisense, processed_pseudogene, unprocessed_pseudogene and misc_RNA were grouped under 'long ncRNA'") +
#    geom_boxplot(outlier.size=0.5) + 
#    facet_grid(~Condition) +
#    scale_x_discrete("") +
#    scale_y_continuous("log2(Inter/Intra)") +
#    scale_fill_brewer("Replicate", palette="Set1") +
#    theme_bw() +
#    theme(legend.position="right", axis.text.x=element_text(angle = 45, hjust = 1))
# ggsave(gg2, filename="plots/RNA-DNA_Inter_vs_Intrachromosomal_interactions_groupedBiotype.png", units="in", width=12, height=4, points=10)
# 
# png(paste0("plots/RNA-DNA_Inter_vs_Intrachromosomal_interactions_Biotypes.png"), units="in", height=10, width=16, res=300)
#    ggarrange(gg1, gg2, ncol=1, nrow=2)
# dev.off()

counts = subset[is.finite(Ratio), list(Interchromosomal=sum(Interchromosomal), Intrachromosomal=sum(Intrachromosomal), Sum=mean(Sum), Ratio=mean(Ratio)), by=c("Condition", "gene_id.R", "annotation.R")]
counts = counts[Interchromosomal+Intrachromosomal>=100,]
counts[, annotation.R := factor(annotation.R, levels=c("protein_coding", "long_ncRNA", "ncRNA", "other"))]

gg1 = ggviolin(counts, x="annotation.R", y="Ratio", facet.by="Condition") +
   ggtitle("Interchromosomal vs intrachromosomal reads ratio (Transcript collapsed - No Self)",
           subtitle="A pseudo-count of 1 has been added to those interactions with 0 count in either intrachromosomal or interchromosomal fraction") +
   stat_compare_means(ref.group="protein_coding", method="wilcox.test", hide.ns=TRUE, label = "p.signif") +
   scale_x_discrete("") + scale_y_continuous("log2(Inter/Intra)") +
   theme_bw() +
   theme(axis.text.x=element_text(angle=60, hjust=1, vjust=1, size=14),
         axis.text.y=element_text(size=14), legend.text=element_text(size=14),
         axis.title.y=element_text(size=14), legend.title=element_text(size=14),
         strip.text.x=element_text(size=14), strip.text.y=element_text(size=14),
         legend.position="bottom") +
   facet_wrap(~Condition, nrow=2) + geom_boxplot(outlier.colour=NA, width=0.1)

gg3 = ggplot(counts[Condition%in%"1FA"], aes(x=Ratio, y=Sum)) +
   ggtitle("Sum and ration of Interchromosomal and intrachromosomal raw reads (Transcript collapsed - No Self)",
           subtitle="A pseudo-count of 1 has been added to those interactions with 0 count in either intrachromosomal or interchromosomal fraction\nlincRNA, processed_transcript, antisense, processed_pseudogene and unprocessed_pseudogene were grouped under 'lncRNA'") +
   geom_point(size=0.5, alpha=0.5) +
   # scale_colour_manual("Biotype", values=c('#8dd3c7','#ffffb3','#bebada','#fb8072','#80b1d3','#fdb462','#b3de69','#fccde5','#d9d9d9','#bc80bd')) +
   # scale_colour_brewer("Biotype", palette="Set1") +
   scale_x_continuous("Ratio\nlog2(interchromosomal/intrachromosomal)") +
   scale_y_continuous("Sum\nlog2(interchromosomal+intrachromosomal)") +
   facet_wrap(~annotation.R, ncol=5, drop=FALSE) + theme_bw() + theme(legend.position="bottom")

# counts = subset[is.finite(Ratio), list(Interchromosomal=sum(Interchromosomal), Intrachromosomal=sum(Intrachromosomal), Sum=mean(Sum), Ratio=mean(Ratio)), by=c("Condition", "gene_id.R", "Biotype2")]
# counts = counts[Interchromosomal+Intrachromosomal>=10]
# 
# gg2 = ggviolin(counts, x="Biotype2", y="Ratio", facet.by="Condition") +
#    ggtitle("Interchromosomal vs intrachromosomal reads ratio (Transcript collapsed - No Self)",
#            subtitle="A pseudo-count of 1 has been added to those interactions with 0 count in either intrachromosomal or interchromosomal fraction\nlincRNA, processed_transcript, antisense, processed_pseudogene and unprocessed_pseudogene were grouped under 'lncRNA'") +
#    stat_compare_means(ref.group="protein_coding", method="wilcox.test", hide.ns=TRUE, label = "p.signif") +
#    scale_x_discrete("") + scale_y_continuous("log2(Inter/Intra)") +
#    theme_bw() +
#    theme(axis.text.x=element_text(angle=60, hjust=1, vjust=1, size=14),
#          axis.text.y=element_text(size=14), legend.text=element_text(size=14),
#          axis.title.y=element_text(size=14), legend.title=element_text(size=14),
#          strip.text.x=element_text(size=14), strip.text.y=element_text(size=14),
#          legend.position="bottom") +
#    facet_wrap(~Condition, nrow=2) + geom_boxplot(outlier.colour=NA, width=0.1)

pdf(paste0("plots/RNA-DNA_Inter_vs_Intrachromosomal_interactions_Biotypes_perTranscript.pdf"), height=8, width=20)
   ggarrange(gg1, gg3, ncol=2, nrow=1)
dev.off()

gg2.1 = ggviolin(counts[Condition%in%"1FA"], x="annotation.R", y="Ratio", facet.by="Condition") +
   ggtitle("Interchromosomal vs intrachromosomal reads ratio (Transcript collapsed - No Self)",
           subtitle="A pseudo-count of 1 has been added to those interactions with 0 count in either intrachromosomal or interchromosomal fraction\nlincRNA, processed_transcript, antisense, processed_pseudogene and unprocessed_pseudogene were grouped under 'lncRNA'") +
   stat_compare_means(ref.group="protein_coding", method="wilcox.test", hide.ns=TRUE, label = "p.signif") +
   scale_x_discrete("") + scale_y_continuous("log2(Inter/Intra)") +
   theme_bw() +
   theme(axis.text.x=element_text(angle=60, hjust=1, vjust=1, size=14),
         axis.text.y=element_text(size=14), legend.text=element_text(size=14),
         axis.title.y=element_text(size=14), legend.title=element_text(size=14),
         strip.text.x=element_text(size=14), strip.text.y=element_text(size=14),
         legend.position="bottom") +
   facet_wrap(~Condition, nrow=2) + geom_boxplot(outlier.colour=NA, width=0.1)
ggsave(gg2.1, filename=paste0("plots/RNA-DNA_Inter_vs_Intrachromosomal_interactions_Biotypes_perTranscript_1FA.png"), height=8, width=10)
ggsave(gg2.1, filename=paste0("plots/RNA-DNA_Inter_vs_Intrachromosomal_interactions_Biotypes_perTranscript_1FA.pdf"), height=8, width=10)

# Significant
interactions = fread(paste0("gunzip -c RNA-DNA_counts_", gencode_version, "_level_1_2_3_AllOverlapsUnique_significant", fdr.th, "_extended_tsc_noDistance.tsv.gz"))

interactions[, interactionType := ifelse(chr.R==chr.D, "Intrachromosomal", "Interchromosomal")]

setkeyv(interactions, c("gene_id.R", "interactionType", "annotation.R", "Condition", "Replicate"))
subset = interactions[gene_id.R!=gene_id.D & !gene_id.R%in%c("Multigenic", "Intergenic"), .N, by=key(interactions)]

subset = dcast.data.table(subset, Condition+Replicate+annotation.R+gene_id.R~interactionType, value.var="N", fill=1)
subset[, `:=`(Sum = log2(Interchromosomal+Intrachromosomal), Ratio = log2(Interchromosomal/Intrachromosomal))]

# gg1 = ggplot(subset[Interchromosomal+Intrachromosomal>=10], aes(x=Biotype, y=Ratio, fill=as.factor(Replicate))) +
#    ggtitle("Interchromosomal vs intrachromosomal reads ratio", subtitle="A pseudo-count of 1 has been added to those interactions with 0 count in either intrachromosomal or interchromosomal fraction") +
#    geom_boxplot(outlier.size=0.5) + 
#    facet_grid(~Condition) +
#    scale_x_discrete("") +
#    scale_y_continuous("log2(Inter/Intra)") +
#    scale_fill_brewer("Replicate", palette="Set1") +
#    theme_bw() +
#    theme(legend.position="right", axis.text.x=element_text(angle = 45, hjust = 1))
# 
# gg2 = ggplot(subset[Interchromosomal+Intrachromosomal>=10], aes(x=Biotype2, y=Ratio, fill=as.factor(Replicate))) +
#       ggtitle("Interchromosomal vs intrachromosomal reads ratio", subtitle="A pseudo-count of 1 has been added to those interactions with 0 count in either intrachromosomal or interchromosomal fraction\nlincRNA, processed_transcript, antisense, processed_pseudogene, unprocessed_pseudogene and misc_RNA were grouped under 'long ncRNA'") +
#    geom_boxplot(outlier.size=0.5) + 
#    facet_grid(~Condition) +
#    scale_x_discrete("") +
#    scale_y_continuous("log2(Inter/Intra)") +
#    scale_fill_brewer("Replicate", palette="Set1") +
#    theme_bw() +
#    theme(legend.position="right", axis.text.x=element_text(angle = 45, hjust = 1))
# 
# png(paste0("plots/RNA-DNA_Inter_vs_Intrachromosomal_interactions_Biotypes_filtered.png"), units="in", height=10, width=16, res=300)
#    ggarrange(gg1, gg2, ncol=1, nrow=2)
# dev.off()

counts = subset[is.finite(Ratio), list(Interchromosomal=sum(Interchromosomal), Intrachromosomal=sum(Intrachromosomal), Sum=mean(Sum), Ratio=mean(Ratio)), by=c("Condition", "gene_id.R", "annotation.R")]
counts = counts[Interchromosomal+Intrachromosomal>=10,]
counts[, annotation.R := factor(annotation.R, levels=c("protein_coding", "long_ncRNA", "ncRNA", "other"))]

gg1 = ggviolin(counts, x="annotation.R", y="Ratio", facet.by="Condition") +
   ggtitle("Interchromosomal vs intrachromosomal reads ratio (Transcript collapsed - No Self)",
           subtitle="A pseudo-count of 1 has been added to those interactions with 0 count in either intrachromosomal or interchromosomal fraction") +
   stat_compare_means(ref.group="protein_coding", method="wilcox.test", hide.ns=TRUE, label = "p.signif") +
   scale_x_discrete("") + scale_y_continuous("log2(Inter/Intra)") +
   theme_bw() +
   theme(axis.text.x=element_text(angle=60, hjust=1, vjust=1, size=14),
         axis.text.y=element_text(size=14), legend.text=element_text(size=14),
         axis.title.y=element_text(size=14), legend.title=element_text(size=14),
         strip.text.x=element_text(size=14), strip.text.y=element_text(size=14),
         legend.position="bottom") +
   facet_wrap(~Condition, nrow=2) + geom_boxplot(outlier.colour=NA, width=0.1)

gg4 = ggplot(counts[Condition%in%"1FA"], aes(x=Ratio, y=Sum)) +
   ggtitle("Sum and ration of Interchromosomal and intrachromosomal filtered reads (Transcript collapsed - No Self)",
           subtitle="A pseudo-count of 1 has been added to those interactions with 0 count in either intrachromosomal or interchromosomal fraction\nlincRNA, processed_transcript, antisense, processed_pseudogene and unprocessed_pseudogene were grouped under 'lncRNA'") +
   geom_point(size=0.5, alpha=0.5) +
   # scale_colour_manual("Biotype", values=c('#8dd3c7','#ffffb3','#bebada','#fb8072','#80b1d3','#fdb462','#b3de69','#fccde5','#d9d9d9','#bc80bd')) +
   # scale_colour_brewer("Biotype", palette="Set1") +
   scale_x_continuous("Ratio\nlog2(interchromosomal/intrachromosomal)") +
   scale_y_continuous("Sum\nlog2(interchromosomal+intrachromosomal)") +
   facet_wrap(~annotation.R, ncol=5, drop=FALSE) + theme_bw() + theme(legend.position="bottom")

# counts = subset[is.finite(Ratio), list(Interchromosomal=sum(Interchromosomal), Intrachromosomal=sum(Intrachromosomal), Sum=mean(Sum), Ratio=mean(Ratio)), by=c("Condition", "gene_id.R", "Biotype2")]
# counts = counts[Interchromosomal+Intrachromosomal>=10]
# 
# gg2 = ggviolin(counts, x="Biotype2", y="Ratio", facet.by="Condition") + 
#       ggtitle("Interchromosomal vs intrachromosomal reads ratio (Transcript collapsed - No Self)",
#            subtitle="A pseudo-count of 1 has been added to those interactions with 0 count in either intrachromosomal or interchromosomal fraction\nlincRNA, processed_transcript, antisense, processed_pseudogene and unprocessed_pseudogene were grouped under 'lncRNA'") +
#    stat_compare_means(ref.group="protein_coding", method="wilcox.test", hide.ns=TRUE, label = "p.signif") +
#    scale_x_discrete("") + scale_y_continuous("log2(Inter/Intra)") +
#    theme_bw() +
#    theme(axis.text.x=element_text(angle=60, hjust=1, vjust=1, size=14),
#          axis.text.y=element_text(size=14), legend.text=element_text(size=14),
#          axis.title.y=element_text(size=14), legend.title=element_text(size=14),
#          strip.text.x=element_text(size=14), strip.text.y=element_text(size=14),
#          legend.position="bottom") +
#    facet_wrap(~Condition, nrow=2) + geom_boxplot(outlier.colour=NA, width=0.1)

pdf(paste0("plots/RNA-DNA_Inter_vs_Intrachromosomal_interactions_Biotypes_perTranscript_Significant.pdf"), height=8, width=20)
   ggarrange(gg1, gg3, ncol=2, nrow=1)
dev.off()

gg4.1 = ggviolin(counts[Condition%in%"1FA"], x="annotation.R", y="Ratio", facet.by="Condition") +
   ggtitle("Interchromosomal vs intrachromosomal reads ratio (Transcript collapsed - No Self)",
           subtitle="A pseudo-count of 1 has been added to those interactions with 0 count in either intrachromosomal or interchromosomal fraction\nlincRNA, processed_transcript, antisense, processed_pseudogene and unprocessed_pseudogene were grouped under 'lncRNA'") +
   stat_compare_means(ref.group="protein_coding", method="wilcox.test", hide.ns=TRUE, label = "p.signif") +
   scale_x_discrete("") + scale_y_continuous("log2(Inter/Intra)") +
   theme_bw() +
   theme(axis.text.x=element_text(angle=60, hjust=1, vjust=1, size=14),
         axis.text.y=element_text(size=14), legend.text=element_text(size=14),
         axis.title.y=element_text(size=14), legend.title=element_text(size=14),
         strip.text.x=element_text(size=14), strip.text.y=element_text(size=14),
         legend.position="bottom") +
   facet_wrap(~Condition, nrow=2) + geom_boxplot(outlier.colour=NA, width=0.1)
ggsave(gg4.1, filename=paste0("plots/RNA-DNA_Inter_vs_Intrachromosomal_interactions_Biotypes_perTranscript_Significant_1FA.png"), height=8, width=10)
ggsave(gg4.1, filename=paste0("plots/RNA-DNA_Inter_vs_Intrachromosomal_interactions_Biotypes_perTranscript_Significant_1FA.pdf"), height=8, width=10)
```

### Circos

```{r extendFunction}
extend_chromosomes = function(bed, chromosome, prefix = "zoom_") {
    zoom_bed = bed[bed[[1]] %in% chromosome, , drop = FALSE]
    zoom_bed[[1]] = paste0(prefix, zoom_bed[[1]])
    rbind(bed, zoom_bed)
}
```

```{r circosRNAsFiltered, eval=FALSE}
interactions = fread(paste0("gunzip -c RNA-DNA_counts_", gencode_version, "_level_1_2_3_AllOverlapsUnique_significant", fdr.th, "_extended_tsc_noDistance.tsv.gz"))

subset = interactions[, list(Count=.N), by=c("Condition", "gene_id.R", "gene_id.D")]

dna_ids = copy(genes)
dna_ids = keepSeqlevels(dna_ids, paste0("chr", c(1:19, "X", "Y")), pruning.mode="coarse")
dna_ids = data.table(as.data.frame(dna_ids), key="gene_id")

for( gene in c("ENSMUSG00000092274.2", "ENSMUSG00000065037.1", "ENSMUSG00000101609.1", "ENSMUSG00000085396.7", "ENSMUSG00000077323.1", "ENSMUSG00000065176.1", "ENSMUSG00000080542.1", "ENSMUSG00000080358.1", "ENSMUSG00000004637.15", "ENSMUSG00000069769.13", "ENSMUSG00000038518.15", "ENSMUSG00000026380.10", "ENSMUSG00000012396.12") ){
   gene_name = genes.dt[gene_id%in%gene, gene_name]
   gene_chr = as.character(genes.dt[gene_id%in%gene, seqnames])

   # message(gene_name)
   
   myset = subset[gene_id.R%in%gene & gene_id.D%in%dna_ids[, gene_id],]
   
   if(nrow(myset)>1){
      breaks = seq(log2(min(myset[, Count])), log2(max(myset[, Count])), length.out=9)
      col_fun = colorRamp2(breaks, c('#ffffd9','#edf8b1','#c7e9b4','#7fcdbb','#41b6c4','#1d91c0','#225ea8','#253494','#081d58'))
      lgd_links = Legend(at = seq(min(round(range(breaks))), max(round(range(breaks))), length.out=5),
                         col_fun = col_fun, type="grid", title_position = "topcenter", title = "log2(Obs/Exp)", direction = "horizontal")
      lgd_list_vertical = packLegend(lgd_links)
      
      myset = split(myset, myset$Condition)
      
      bed1 = lapply(myset,
                    function(x)
                       cbind(dna_ids[match(x[, gene_id.R], dna_ids[, gene_id]),], rou1=0.8))
      for( i in seq_along(bed1) )
         setkey(bed1[[i]], gene_type)
      
      bed2 = lapply(myset,
                    function(x)
                       cbind(dna_ids[match(x[, gene_id.D], dna_ids[, gene_id]),], rou2=0.9))
      for( i in seq_along(bed2) )
         setkey(bed2[[i]], gene_type)
      
      colors = data.table(label=c("protein_coding", "lincRNA"),
                          value=c('#1f78b4', '#e31a1c'),
                          key="label")
      
      bed2 = lapply(seq_along(bed2),
                    function(x)
                       bed2[[as.numeric(as.character(x))]][, value := col_fun(log2(myset[[as.numeric(as.character(x))]][, Count]))])


      cytoband = read.cytoband(species="mm10")
      cytoband_df = cytoband$df
      chromosome = cytoband$chromosome

      # xrange = c(cytoband$chr.len, cytoband$chr.len[gene_chr])
      # normal_chr_index = 1:length(chromosome)
      # zoomed_chr_index = length(chromosome)+1
      # 
      # # normalize in normal chromsomes and zoomed chromosomes separately
      # sector.width = c(xrange[normal_chr_index] / sum(xrange[normal_chr_index]), 
      #                  xrange[zoomed_chr_index] / sum(xrange[zoomed_chr_index]))
      
      png(paste0("plots/circos_", gene_name, "_filtered.png"), units="in", type="cairo", res=300, width=18, height=8)
      layout(matrix(1:length(bed1), 2, length(bed1)/2))
      for( i in seq_along(bed1) ){
         circos.par(start.degree = 90, gap.degree=c(rep(1, length(chromosome)-1), 5))
         circos.initializeWithIdeogram(plotType = c("axis", "labels"), species = "mm10")
         # circos.initializeWithIdeogram(plotType = c("axis", "labels"),  extend_chromosomes(cytoband_df, gene_chr),
         #                               sector.width = sector.width)
         circos.genomicLabels(bed1[[i]][1,], side = "outside", labels=gene_name)
         circos.genomicLink(bed1[[i]], bed2[[i]], col=bed2[[i]]$value, rou1=bed1[[i]]$rou1, rou2=bed2[[i]]$rou2)
         text(0, 1.2, names(bed1)[i], cex=2, font=2)
         title(names(bed1)[i])
         circos.clear()
      }
      pushViewport(viewport(x = unit(2, "mm"), y = unit(4, "mm"), 
                            width = grobWidth(lgd_list_vertical)*2, 
                            height = grobHeight(lgd_list_vertical), 
                            just = c("left", "bottom")))
      grid.draw(lgd_list_vertical)
      dev.off()
   }
}
```

```{r circosRNAs, eval=FALSE}
interactions = readRDS(paste0("RData/RNA-DNA_noMultiOverlapsUniqueNoActD_counts_", gencode_version, "_level_1_2_3_pvalue_FDR.rds"))

# subset = interactions[, .N, by=c("Condition", "gene_id.R", "gene_id.D")]
subset = interactions[fdr.bh<0.01,]

dna_ids = copy(genes)
dna_ids = keepSeqlevels(dna_ids, paste0("chr", c(1:19, "X", "Y")), pruning.mode="coarse")
dna_ids = data.table(as.data.frame(dna_ids), key="gene_id")

for( gene in c("ENSMUSG00000092274.2", "ENSMUSG00000065037.1", "ENSMUSG00000101609.1", "ENSMUSG00000085396.7", "ENSMUSG00000077323.1", "ENSMUSG00000065176.1", "ENSMUSG00000080542.1", "ENSMUSG00000080358.1", "ENSMUSG00000004637.15", "ENSMUSG00000069769.13", "ENSMUSG00000038518.15") ){
   gene_name = genes.dt[gene_id%in%gene, gene_name]
   gene_chr = as.character(genes.dt[gene_id%in%gene, seqnames])

   message(gene_name)
   
   myset = subset[gene_id.R%in%gene & gene_id.D%in%dna_ids[, gene_id],]
   
   if(nrow(myset)>1){
      breaks = seq(log2(min(myset[, Count])), log2(max(myset[, Count])), length.out=9)
      col_fun = colorRamp2(breaks, c('#ffffd9','#edf8b1','#c7e9b4','#7fcdbb','#41b6c4','#1d91c0','#225ea8','#253494','#081d58'))
      lgd_links = Legend(at = seq(min(round(range(breaks))), max(round(range(breaks))), length.out=5),
                         col_fun = col_fun, type="grid", title_position = "topcenter", title = "log2(Obs/Exp)", direction = "horizontal")
      lgd_list_vertical = packLegend(lgd_links)
      
      myset = split(myset, myset$Condition)
      
      bed1 = lapply(myset,
                    function(x)
                       cbind(dna_ids[match(x[, gene_id.R], dna_ids[, gene_id]),], rou1=0.8))
      for( i in seq_along(bed1) )
         setkey(bed1[[i]], gene_type)
      
      bed2 = lapply(myset,
                    function(x)
                       cbind(dna_ids[match(x[, gene_id.D], dna_ids[, gene_id]),], rou2=0.9))
      for( i in seq_along(bed2) )
         setkey(bed2[[i]], gene_type)
      
      colors = data.table(label=c("protein_coding", "lincRNA"),
                          value=c('#1f78b4', '#e31a1c'),
                          key="label")
      
      bed2 = lapply(seq_along(bed2),
                    function(x)
                       bed2[[as.numeric(as.character(x))]][, value := col_fun(log2(myset[[as.numeric(as.character(x))]][, Count]))])


      cytoband = read.cytoband(species="mm10")
      cytoband_df = cytoband$df
      chromosome = cytoband$chromosome

      # xrange = c(cytoband$chr.len, cytoband$chr.len[gene_chr])
      # normal_chr_index = 1:length(chromosome)
      # zoomed_chr_index = length(chromosome)+1
      # 
      # # normalize in normal chromsomes and zoomed chromosomes separately
      # sector.width = c(xrange[normal_chr_index] / sum(xrange[normal_chr_index]), 
      #                  xrange[zoomed_chr_index] / sum(xrange[zoomed_chr_index]))
      
      png(paste0("plots/circos_", gene_name, ".png"), units="in", type="cairo", res=300, width=18, height=8)
      layout(matrix(1:length(bed1), 2, length(bed1)/2))
      for( i in seq_along(bed1) ){
         circos.par(start.degree = 90, gap.degree=c(rep(1, length(chromosome)-1), 5))
         circos.initializeWithIdeogram(plotType = c("axis", "labels"), species = "mm10")
         # circos.initializeWithIdeogram(plotType = c("axis", "labels"),  extend_chromosomes(cytoband_df, gene_chr),
         #                               sector.width = sector.width)
         circos.genomicLabels(bed1[[i]][1,], side = "outside", labels=gene_name)
         circos.genomicLink(bed1[[i]], bed2[[i]], col=bed2[[i]]$value, rou1=bed1[[i]]$rou1, rou2=bed2[[i]]$rou2)
         text(0, 1.2, names(bed1)[i], cex=2, font=2)
         title(names(bed1)[i])
         circos.clear()
      }
      pushViewport(viewport(x = unit(2, "mm"), y = unit(4, "mm"), 
                            width = grobWidth(lgd_list_vertical)*2, 
                            height = grobHeight(lgd_list_vertical), 
                            just = c("left", "bottom")))
      grid.draw(lgd_list_vertical)
      dev.off()
   }
}
```

### Intronic Rate (old)

```{r, eval=FALSE}
examples = unique(gff[gene_type%in%"protein_coding", list(name=gene_name, id=gene_id)])
setkey(examples, id)

interactions = rbindlist(list(readRDS(paste0("RNA-DNA_overlaps_", gencode_version, "_level_1_2_3_AllOverlapsUniqueNoActD_1FA.rds"))))
interactions = interactions[gene_id.R%in%examples[, id]]

gff.sel = gff[gene_id%in%examples[, id]]

interactions.gr = with(interactions, GRanges(chr.R, IRanges(start.R, end.R), strand.R))
gff.sel.gr = with(gff.sel, GRanges(seqnames, IRanges(start, end), strand))

overlaps = as.data.table(findOverlaps(interactions.gr, gff.sel.gr, type="within", minoverlap=16))
sel_dup = unique(overlaps[duplicated(queryHits), queryHits])
overlaps = overlaps[!queryHits%in%sel_dup,]

exon_intron_width = gff[gene_feature%in%c("exon", "intron"),][, list(width = sum(end-start)), by=c("gene_id", "gene_feature")]
setkeyv(exon_intron_width, c("gene_id", "gene_feature"))

interactions[, `:=`(gene_type = as.character(NA), gene_feature = as.character(NA), width = as.numeric(NA)),]
interactions[overlaps[, queryHits], `:=`(gene_type = gff.sel[overlaps[, subjectHits], gene_type],
                                         gene_feature = gff.sel[overlaps[, subjectHits], gene_feature],
                                         width = as.numeric(gff.sel[overlaps[, subjectHits], end-start]))]

mindist = rowMin(as.matrix(interactions[, list(one=abs(start.R-end.D), two=abs(start.D-end.R))]))
interactions[, distance := as.character(cut(mindist,
                                            breaks=c(0, 1e3, 1e4, 1e5, 1e6, Inf),
                                            labels=c("<1Kb", "1-10Kb", "10-100Kb", "100Kb-1Mb", ">1Mb"),
                                            include.lowest=TRUE))]
interactions[gene_id.R==gene_id.D, distance := "Self-interaction"]
interactions[chr.R!=chr.D, distance := "Interchromosome"]
interactions[, distance := factor(distance,
                                  levels=c("Self-interaction", "<1Kb", "1-10Kb", "10-100Kb", "100Kb-1Mb", ">1Mb", "Interchromosome"))]

setkeyv(interactions, c("Condition", "gene_id.R", "gene_id.D", "gene_feature", "distance"))

plotdata = interactions[!is.na(gene_type), .N, by=key(interactions)]

setkeyv(plotdata, c("gene_id.R", "gene_feature"))
plotdata[exon_intron_width, width := as.numeric(i.width)]

plotdata = plotdata[gene_feature%in%c("exon", "intron")][, K := N/width*1e3]

sel = plotdata[gene_feature%in%"exon" & !distance%like%"Self"][N>20, .N, by=c("Condition", "gene_id.R")][N>7, gene_id.R]
plotdata = plotdata[gene_id.R%in%sel,]

plotdata = dcast.data.table(plotdata, Condition+gene_id.R+gene_id.D+distance~gene_feature, value.var=c("N", "K"))
plotdata = plotdata[!(is.na(K_exon)|is.na(K_intron))]

setkey(plotdata, gene_id.R)
plotdata[examples, name := name]

ggplot(plotdata, aes(x=distance, y=log2(K_intron/K_exon), fill=name)) + geom_boxplot(outlier.size=0.5) + theme_bw()
plotdata[, Ratio := log2(K_intron/K_exon)]


Tet1 = fread("HiC_Tet1_all.txt")
Tet1[, V5 := gsub("\"|;", " ", V5)]
Tet1[, gene_id := sub(".*ID=(\\S+).*", "\\1", V5)]
Tet1[, gene_name := sub(".*gene_name=(\\S+).*", "\\1", V5)]
Tet1[, V5 := NULL]

plotdata[gene_id.R%in%"ENSMUSG00000047146.16"][N_exon+N_intron>10][order(-log2(K_intron/K_exon))][gene_id.D%in%Tet1[, gene_id]]
table(plotdata[gene_id.R%in%"ENSMUSG00000047146.16"][N_exon+N_intron>10][order(-log2(K_intron/K_exon))][!gene_id.D%in%Tet1[, gene_id], Ratio>-0.18809097])
table(plotdata[gene_id.R%in%"ENSMUSG00000047146.16"][N_exon+N_intron>10][order(-log2(K_intron/K_exon))][gene_id.D%in%Tet1[, gene_id], Ratio>-0.18809097])

Rpl12 = fread("HiC_Rpl12.txt")
Rpl12[, V5 := gsub("\"|;", " ", V5)]
Rpl12[, gene_id := sub(".*ID=(\\S+).*", "\\1", V5)]
Rpl12[, gene_name := sub(".*gene_name=(\\S+).*", "\\1", V5)]
Rpl12[, V5 := NULL]

plotdata[gene_id.R%in%"ENSMUSG00000038900.17"][N_exon+N_intron>10][order(-log2(K_intron/K_exon))][gene_id.D%in%Rpl12[, gene_id]]

Dido1 = fread("HiC_Dido1.txt")
Dido1[, V5 := gsub("\"|;", " ", V5)]
Dido1[, gene_id := sub(".*ID=(\\S+).*", "\\1", V5)]
Dido1[, gene_name := sub(".*gene_name=(\\S+).*", "\\1", V5)]
Dido1[, V5 := NULL]

plotdata[gene_id.R%in%"ENSMUSG00000038914.15"][N_exon+N_intron>10][order(-log2(K_intron/K_exon))][gene_id.D%in%Dido1[, gene_id]]
table(plotdata[gene_id.R%in%"ENSMUSG00000038914.15"][N_exon+N_intron>10][order(-log2(K_intron/K_exon))][!gene_id.D%in%Dido1[, gene_id], Ratio>0.09773587])
table(plotdata[gene_id.R%in%"ENSMUSG00000038914.15"][N_exon+N_intron>10][order(-log2(K_intron/K_exon))][gene_id.D%in%Dido1[, gene_id], Ratio>0.09773587])





setkeyv(plotdata, c("Condition", "gene_id.R", "gene_feature", "distance"))



setkeyv(plotdata, c("Condition", "BiolReplicate", "gene_type", "gene_feature", "distance"))
plotdata[, C := length(unique(gene_id.R)), by=key(plotdata)]

```

```{r calculateExonIntronJunction, eval=FALSE}
interactions = readRDS(paste0("RData/RNA-DNA_Junctions_overlaps_", gencode_version, "_level_1_2_3_AllOverlapsUnique.rds"))

interactions[, c("Condition", "BiolReplicate") := tstrsplit(name, "_")][, name := NULL]

interactions.gr = with(interactions, GRanges(chr.R, IRanges(start.R, end.R), strand.R))

overlaps.rna = as.data.table(findOverlaps(interactions.gr, genes, ignore.strand=FALSE, minoverlap=16))
setkey(overlaps.rna, subjectHits)
overlaps.rna = overlaps.rna[genes.dt[, list(index, gene_id.R=gene_id)], nomatch=0][, subjectHits := NULL]
setkey(overlaps.rna, queryHits)

interactions = interactions[gene_id.R%in%c(coding.sel, noncoding.sel[["lincRNA"]]) & gene_id.D%in%c(coding.sel, unlist(noncoding.sel), c("Multigenic", "Intergenic")),]

gff.sel = gff[gene_type%in%c("protein_coding", "lincRNA")]

interactions.gr = with(interactions, GRanges(chr.R, IRanges(start.R, end.R), strand.R))
gff.sel.gr = with(gff.sel, GRanges(seqnames, IRanges(start, end), strand))

overlaps = as.data.table(findOverlaps(interactions.gr, gff.sel.gr, minoverlap=16))
sel_dup = unique(overlaps[duplicated(queryHits), queryHits])
overlaps = overlaps[!queryHits%in%sel_dup,]

rm(interactions.gr)
exon_intron_width = gff[gene_feature%in%c("exon", "intron"),][, list(width = sum(end-start)), by=c("gene_id", "gene_feature")]
setkeyv(exon_intron_width, c("gene_id", "gene_feature"))

interactions[, `:=`(gene_type = as.character(NA), gene_feature = as.character(NA), width = as.numeric(NA)),]
interactions[overlaps[, queryHits], `:=`(gene_type = gff.sel[overlaps[, subjectHits], gene_type],
                                         gene_feature = gff.sel[overlaps[, subjectHits], gene_feature],
                                         width = as.numeric(gff.sel[overlaps[, subjectHits], end-start]))]
rm(overlaps)

mindist = rowMin(as.matrix(interactions[, list(one=abs(start.R-end.D), two=abs(start.D-end.R))]))
interactions[, distance := as.character(cut(mindist,
                                            breaks=c(0, 1e3, 1e4, 1e5, 1e6, Inf),
                                            labels=c("<1Kb", "1-10Kb", "10-100Kb", "100Kb-1Mb", ">1Mb"),
                                            include.lowest=TRUE))]
rm(mindist)

interactions[gene_id.R==gene_id.D, distance := "Self-interaction"]
interactions[chr.R!=chr.D, distance := "Interchromosome"]
interactions[, distance := factor(distance,
                                  levels=c("Self-interaction", "<1Kb", "1-10Kb", "10-100Kb", "100Kb-1Mb", ">1Mb", "Interchromosome"))]

setkeyv(interactions, c("Condition", "BiolReplicate", "gene_type", "gene_id.R", "gene_feature", "width", "distance"))

saveRDS(interactions, paste0("RData/RNA-DNA_overlaps_", gencode_version, "_level_1_2_3_AllOverlapsUnique_ExonsIntrons.rds"))
```

```{r plotExonIntronJunction, eval=FALSE}
# if(!exists("interactions")) interactions = readRDS(paste0("RData/RNA-DNA_overlaps_", gencode_version, "_level_1_2_3_AllOverlapsUnique_ExonsIntrons_filtered.rds"))
if(!exists("interactions")) interactions = readRDS(paste0("RData/RNA-DNA_overlaps_", gencode_version, "_level_1_2_3_AllOverlapsUnique_ExonsIntrons.rds"))

plotdata = interactions[!is.na(gene_type), .N, by=key(interactions)]

setkeyv(plotdata, c("gene_id.R", "gene_feature"))
plotdata[exon_intron_width, width := as.numeric(i.width)]

setkeyv(plotdata, c("Condition", "BiolReplicate", "gene_type", "gene_id.R", "gene_feature", "distance"))
plotdata = plotdata[, list(N=sum(N), K=sum(N/width*1e3), mN=mean(N), mK=mean(N/width*1e3)), by=key(plotdata)]
# setkeyv(plotdata, c("Condition", "BiolReplicate", "gene_type", "gene_feature", "distance"))
# plotdata[, C := length(unique(gene_id.R)), by=key(plotdata)]

plotdata[, `:=`(gene_feature = factor(gene_feature, levels=c("exon", "intron", "five_prime_UTR", "three_prime_UTR")),
                gene_type = factor(gene_type, levels=c("protein_coding", "lincRNA")))]
plotdata = plotdata[, list(mK=mean(mK)), by=c("Condition", "gene_type", "gene_id.R", "gene_feature", "distance")]


sel = unique(plotdata[gene_feature%in%c("exon", "intron"),][, .N, by=c("Condition", "gene_id.R", "distance")][N>=2, gene_id.R])
plotdata2 = plotdata[gene_id.R%in%sel,]
plotdata2[distance%in%c("<1Kb"), distance := "1-10Kb"][distance%in%"1-10Kb", distance := "<10Kb"]
plotdata2[, distance := factor(distance, levels=c("Self-interaction", "<10Kb", "10-100Kb", "100Kb-1Mb", ">1Mb", "Interchromosome"))]
plotdata2 = plotdata2[, list(N=sum(N)), by=c("Condition", "gene_type", "gene_feature", "distance")]
plotdata2[, P := N/sum(N), by=c("Condition", "gene_type", "distance")]

ggplot(plotdata2, aes(x=distance, y=P, fill=gene_feature)) +
   geom_bar(stat="identity", position="stack", col="black", lwd=0.5) +
   scale_fill_brewer("Transcript feature", palette="Set1") +
   scale_y_continuous("Reads fraction", labels=scales::percent) +
   scale_x_discrete("Distance to transcription site") +
   facet_grid(gene_type~Condition) +
   theme_bw() + theme(axis.text.x=element_text(angle=34,hjust=1,vjust=0.5), legend.position="top")


sel = unique(plotdata[gene_feature%in%c("exon", "intron"),][, .N, by=c("Condition", "gene_id.R", "distance")][N>=2, gene_id.R])
# plotdata2 = plotdata[gene_feature%in%c("exon", "intron"), list(N=sum(N), K=sum(K)), by=c("Condition", "BiolReplicate", "gene_type", "gene_feature", "distance", "gene_id.R")]
plotdata2 = plotdata[gene_feature%in%c("exon", "intron") & gene_id.R%in%sel,]
# plotdata2[distance%in%c("<1Kb", "1-10Kb"), distance := "10-100Kb"][distance%in%"10-100Kb", distance := "<100Kb"]
# plotdata2[, distance := factor(distance, levels=c("Self-interaction", "<100Kb", "100Kb-1Mb", ">1Mb", "Interchromosome"))]
plotdata2[distance%in%c("<1Kb"), distance := "1-10Kb"][distance%in%"1-10Kb", distance := "<10Kb"]
plotdata2[, distance := factor(distance, levels=c("Self-interaction", "<10Kb", "10-100Kb", "100Kb-1Mb", ">1Mb", "Interchromosome"))]
plotdata2 = dcast.data.table(plotdata2, Condition+gene_type+distance+gene_id.R~gene_feature, value.var="mK", fun.aggregate=mean, fill=0)
plotdata2 = plotdata2[, `:=`(ratioIE = (intron)/(exon), ratioEI = (exon)/(intron)) ]
setkeyv(plotdata2, c("Condition", "gene_id.R"))
sel = plotdata2[is.finite(ratioEI) & ratioEI>0, length(unique(distance)), by=key(plotdata2)][V1==length(unique(plotdata2[, distance]))]
plotdata2 = plotdata2[sel, nomatch=0]

ggplot(plotdata2, aes(x=distance, y=log2(ratioEI), fill=distance)) +
   geom_violin(position="dodge") +
   geom_boxplot(width=0.2, outlier.colour=NA, fill="white") +
   # scale_fill_brewer("Transcript feature", palette="Set1") +
   scale_fill_manual("Distance to transcription site", values=rev(brewer.pal(7,"Blues"))) +
   scale_y_continuous("Exonic/Intronic reads (log2)") +
   scale_x_discrete("") +
   facet_grid(gene_type~Condition) +
   theme_bw() +
   theme(axis.text.x=element_text(angle=90, hjust=1, vjust=0.5, size=14),
         axis.text.y=element_text(size=14), legend.text=element_text(size=14),
         axis.title.y=element_text(size=14), legend.title=element_text(size=14),
         strip.text.x=element_text(size=14), strip.text.y=element_text(size=14),
         legend.position="top") +
   guides(fill = guide_legend(ncol = 7))


tab = plotdata[, list(Num=length(unique(gene_id.R))), by=c("Condition", "gene_type", "distance", "BiolReplicate")]

ggplot(tab, aes(x=distance, y=Num, fill=as.factor(BiolReplicate))) +
   geom_bar(stat="identity", position="dodge", col="black", lwd=0.5) +
   scale_fill_brewer("Transcript feature", palette="Set1") +
   scale_y_continuous("Number of interacting transcripts") +
   scale_x_discrete("Distance to transcription site") +
   facet_grid(gene_type~Condition, scales="free_y") +
   theme_bw() + theme(axis.text.x=element_text(angle=90,hjust=1,vjust=0.5), legend.position="top")
```

### Figures

```{r figures, eval=FALSE}
# Figure 1B
tab = melt.data.table(fread("Figure1b.tsv", col.names=c("Fraction", "1FA", "2FA")), id.vars="Fraction")
tab[, Fraction := sub("(\\w)(\\w*)", "\\U\\1\\L\\2", Fraction, perl=TRUE)][, Fraction := factor(Fraction, levels=unique(Fraction))]
tab[, Label := paste("Step ", c("I", "II", "III", "IV", "V"))][, Label := factor(Label, levels=paste("Step ", c("I", "II", "III", "IV", "V")))]
gg = ggplot(tab, aes(x=Label, y=value)) +
   geom_bar(stat="identity", position="dodge", col="black") +
   scale_x_discrete("") + scale_y_continuous("Reads Fraction") +
   facet_grid(~variable, scales="free_y") +
   theme_cowplot() + theme(axis.text.x = element_text(angle = 45, hjust=1, size=18),
                           axis.text=element_text(size=16), axis.title=element_text(size=18),
                           strip.background=element_blank(), strip.text=element_text(size=20))
save_plot(gg, filename="plots/Figur1B.pdf", base_height=8, base_aspect_ratio=1.4)
save_plot(gg, filename="plots/Figur1B.eps", base_height=8, base_aspect_ratio=1.4)

tab = melt.data.table(fread("Figure1b_2.tsv", header=TRUE), id.vars="Fraction")
tab[, v.next := shift(value, type="lead"), by="variable"]
tab[!is.na(v.next), value := value-v.next]
tab[, Fraction := sub("(\\w)(\\w*)", "\\U\\1\\L\\2", Fraction, perl=TRUE)][, Fraction := factor(Fraction, levels=unique(Fraction))]
tab[, Label := c("Erroneous ligations", "PCR duplicates", "Ribosomal RNAs", "Multi-mappers", "Uniquely mapped")][, Label := factor(Label, levels=unique(Label))]
tab[variable%in%"1FA", variable := "RADICL"][variable%in%"1FA Trimmed", variable := "RADICL\n(Trimmed)"][, variable := factor(variable, levels=c("RADICL", "RADICL\n(Trimmed)", "GRID"))]

gg = ggplot(tab, aes(x=variable, y=value, fill=Label)) +
   geom_bar(stat="identity", position="stack", col="black", width=0.8) +
   scale_x_discrete("") + scale_y_continuous("Dataset percentage", labels=scales::percent) +
   scale_fill_economist() +
   theme_few() + theme(axis.text.x = element_text(angle = 30, hjust=1, size=18),
                       axis.text=element_text(size=16), axis.title=element_text(size=18),
                       strip.background=element_blank(), strip.text=element_text(size=20),
                       legend.text=element_text(size=16), legend.title=element_blank())
save_plot(gg, filename="plots/Figur1B_2S.pdf", base_height=7, base_aspect_ratio=1.2)
save_plot(gg, filename="plots/Figur1B_2S.eps", base_height=7, base_aspect_ratio=1.2)

tab = melt.data.table(fread("Figure1b_2.tsv", header=TRUE), id.vars="Fraction")
tab[, v.next := shift(value, type="lead"), by="variable"]
tab[!is.na(v.next), value := value-v.next]
tab[, Fraction := sub("(\\w)(\\w*)", "\\U\\1\\L\\2", Fraction, perl=TRUE)][, Fraction := factor(Fraction, levels=unique(Fraction))]
tab[, Label := c("Erroneous ligations", "PCR duplicates", "Ribosomal RNAs", "Multi-mappers", "Uniquely mapped")][, Label := factor(Label, levels=unique(Label))]
tab[variable%in%"1FA", variable := "RADICL"][variable%in%"1FA Trimmed", variable := "RADICL\n(Trimmed)"][, variable := factor(variable, levels=c("RADICL", "RADICL\n(Trimmed)", "GRID"))]
tab = tab[Label%in%c("Multi-mappers", "Uniquely mapped"),][, value := value/sum(value), by="variable"]

gg = ggplot(tab, aes(x=variable, y=value, fill=Label)) +
   geom_bar(stat="identity", position="stack", col="black", width=0.8) +
   scale_x_discrete("") + scale_y_continuous("Dataset percentage", labels=scales::percent) +
   scale_fill_economist() +
   theme_few() + theme(axis.text.x = element_text(angle = 30, hjust=1, size=18),
                       axis.text=element_text(size=16), axis.title=element_text(size=18),
                       strip.background=element_blank(), strip.text=element_text(size=20),
                       legend.text=element_text(size=16), legend.title=element_blank())
save_plot(gg, filename="plots/Figur1B_2.pdf", base_height=7, base_aspect_ratio=1.2)
save_plot(gg, filename="plots/Figur1B_2.eps", base_height=7, base_aspect_ratio=1.2)

# Figure 2A-B
tab = melt.data.table(fread("Figure2.tsv"), id.vars=c("Tag", "Region", "Group"))

tab = tab[, Region := factor(Region, levels=c("intergenic", "genic", "coding_exon", "intronic", "lincRNA_exon", "snRNA", "antisense", "other"))]
tab = tab[order(variable, Group, Region)][, list(Tag, Region, value, percentage=round(value*100, 1), position = cumsum(value) - 0.5*value), by=c("Tag", "Group", "variable")]
tab[, Tag := factor(Tag, levels=c("RNA", "DNA"))]

gg = ggplot(tab[variable%in%"1FA"], aes(x=as.factor(Group), y=value, fill=Region)) +
   geom_bar(position="stack", stat="identity", col="black", width=0.7) +
   scale_fill_tableau(palette="tableau10medium") +
   coord_polar(theta="y") +
   facet_wrap(~Tag) + theme_minimal() +
   geom_text(aes(x = Group+0.5, y = 1-position, label = paste0(percentage, "%")), col="black", size = 4) +
   theme(axis.text = element_blank(), axis.ticks = element_blank(), panel.grid  = element_blank(),
         axis.text.y = element_blank(), axis.text.x = element_blank(), axis.title=element_blank(),
         panel.background = element_blank(), strip.background = element_blank(), strip.text.x = element_text(size = 14),
         strip.text.y = element_text(size = 14), legend.title=element_text(size = 14), legend.text=element_text(size = 14))
ggsave(gg, filename="plots/Figure2A-B_1FA.pdf", width=16, height=6)

gg = ggplot(tab[variable%in%"2FA"], aes(x=as.factor(Group), y=value, fill=Region)) +
   geom_bar(position="stack", stat="identity", col="black", width=0.7) +
   scale_fill_tableau(palette="tableau10medium") +
   coord_polar(theta="y") +
   facet_wrap(~Tag) + theme_minimal() +
   geom_text(aes(x = Group+0.5, y = 1-position, label = paste0(percentage, "%")), col="black", size = 4) +
   theme(axis.text = element_blank(), axis.ticks = element_blank(), panel.grid  = element_blank(),
         axis.text.y = element_blank(), axis.text.x = element_blank(), axis.title=element_blank(),
         panel.background = element_blank(), strip.background = element_blank(), strip.text.x = element_text(size = 14),
         strip.text.y = element_text(size = 14), legend.title=element_text(size = 14), legend.text=element_text(size = 14))
ggsave(gg, filename="plots/Figure2A-B_2FA.pdf", width=16, height=6)

gg = ggplot(tab[variable%in%"GRID"], aes(x=as.factor(Group), y=value, fill=Region)) +
   geom_bar(position="stack", stat="identity", col="black", width=0.7) +
   scale_fill_tableau(palette="tableau10medium") +
   coord_polar(theta="y") +
   facet_wrap(~Tag) + theme_minimal() +
   geom_text(aes(x = Group+0.5, y = 1-position, label = paste0(percentage, "%")), col="black", size = 4) +
   theme(axis.text = element_blank(), axis.ticks = element_blank(), panel.grid  = element_blank(),
         axis.text.y = element_blank(), axis.text.x = element_blank(), axis.title=element_blank(),
         panel.background = element_blank(), strip.background = element_blank(), strip.text.x = element_text(size = 14),
         strip.text.y = element_text(size = 14), legend.title=element_text(size = 14), legend.text=element_text(size = 14))
ggsave(gg, filename="plots/Figure2A-B_GRID.pdf", width=16, height=6)

# Figure 2E
fig = fread("Figure2e.tab", skip=1)
fig[, V3 := factor(V3, levels=c(unique(V3)))]
fig[, V5 := factor(ifelse(V5=="", 1, 2), levels=c(1, 2))]

gg = ggplot(fig, aes(x=V2, y=V4, colour=V3, linetype=V5)) +
   geom_line(lwd=1) +
   scale_x_continuous("Distance from peaks (bp)") +
   scale_y_continuous("Normalized count") +
   scale_colour_manual(values=c('#1f78b4', '#33a02c', '#e31a1c', '#a6cee3', '#b2df8a', '#fb9a99')) +
   scale_linetype(guide=FALSE) +
   theme_bw() + theme(legend.position=c(0.8, 0.8), legend.title=element_blank(), legend.box.background = element_rect(),
                      legend.text=element_text(size=12), axis.text=element_text(size=12), axis.title=element_text(size=14))
ggsave(gg, filename="plots/Figure2F_new.pdf", width=8, height=6)
ggsave(gg, filename="plots/Figure2F_new.eps", width=8, height=6)
gg = ggplot(fig, aes(x=V2, y=V4, colour=V3, linetype=V5)) +
   geom_smooth(method="loess", span=0.2, n=1000, lwd=1, se=FALSE) +
   scale_x_continuous("Distance from peaks (bp)") +
   scale_y_continuous("Normalized count") +
   scale_colour_manual("Legend", values=c('#1f78b4', '#33a02c', '#e31a1c', '#a6cee3', '#b2df8a', '#fb9a99')) +
   scale_linetype(guide=FALSE) +
   theme_bw() + theme(legend.position=c(0.8, 0.8), legend.title=element_blank(), legend.box.background = element_rect(),
                      legend.text=element_text(size=12), axis.text=element_text(size=12), axis.title=element_text(size=14))
ggsave(gg, filename="plots/Figure2F_new_smooth.pdf", width=8, height=6)
ggsave(gg, filename="plots/Figure2F_new_smooth.eps", width=8, height=6)

# Figure 4B
fig = fread("Figure4b.tsv", header=TRUE)
fig[, `Experiment type` := factor(`Experiment type`, levels=c(unique(`Experiment type`)))]
fig[, Interaction := ifelse(Interaction%in%"interchrom", "Trans", "Cis")]
fig = fig[`Experiment type`%in%c("1FA", "4h-ActD", "Control"),]

gg = ggplot(fig, aes(x=`Experiment type`, y=`Triplexes per DNA per RNA`, fill=Interaction)) +
   geom_bar(stat="identity", position="dodge", colour="black") +
   scale_x_discrete("", labels=c("Protein mediated", "ActD", "Non-protein mediated")) +
   scale_y_continuous("Frequency") +
   scale_fill_jco() +
   theme_bw() + theme(legend.position=c(0, 1), legend.box.background = element_rect(), legend.justification=c(-0.1, 1.2),
                      legend.title=element_text(size=14), legend.text=element_text(size=12),
                      axis.text=element_text(size=12), axis.title=element_text(size=14)) +
   guides(fill=guide_legend(title="Interaction type"))
ggsave(gg, filename="plots/Figure4B.pdf", width=8, height=6)
ggsave(gg, filename="plots/Figure4B.eps", width=8, height=6)

# Figure 4C
load("TFfamily_30nt.Robj")

gg = ggplot(table30nt, aes(x=Interaction, y=perc/100, fill=TF_family)) +
   geom_bar(stat="identity", col="black") +
   scale_fill_ptol("TF family") +
   scale_y_continuous("Percentage", labels=scales::percent) +
   scale_x_discrete("Interaction type", labels=c("Cis over 1 Mb", "Trans")) +
   theme_bw()
ggsave(gg, filename="plots/Figure4C.pdf", width=6, height=6)
ggsave(gg, filename="plots/Figure4C.eps", width=6, height=6)

# Figure 5E-F
figE = fread("chia_overlap_stats.25kb.significant.20180208.csv")
figE[, intType := capwords(intType)]
figE[, c("Condition", "Replicate"):=tstrsplit(exprId, "\\.")]
tab1 = summarySE(figE[Condition%in%c("1FA", "4h_ActD", "Control"),], measurevar="or", groupvars=c("Condition", "intType"))

gg1 = ggplot(tab1, aes(x=Condition, y=or, fill=intType)) + 
   geom_bar(position="dodge", stat="identity", col="black") +
   geom_errorbar(aes(ymin=or-se, ymax=or+se),
                 width=.2,                    # Width of the error bars
                 position=position_dodge(.9)) +
   scale_fill_jco() +
   scale_x_discrete("", labels=c("Protein mediated", "ActD", "Non-protein mediated")) +
   scale_y_continuous("OddsRatio") +
   theme_bw() + theme(legend.position=c(0, 1), legend.box.background = element_rect(), legend.justification=c(-0.1, 1.2),
                      legend.title=element_text(size=14), legend.text=element_text(size=12),
                      axis.text=element_text(size=12), axis.title=element_text(size=14),) +
   guides(fill=guide_legend(title="Interaction type"))

figF = fread("chia_overlap_stats.20180208.tads.csv")
figF[, intType := factor(ifelse(intType%in%"inter", "Trans", "Cis"), levels=c("Cis", "Trans"))]
figF[, c("Condition", "Replicate"):=tstrsplit(exprId, "\\.")]
tab2 = summarySE(figF[Condition%in%c("1FA", "4h_ActD", "Control"),], measurevar="or", groupvars=c("Condition", "intType"))

# Error bars represent standard error of the mean
gg2 = ggplot(tab2, aes(x=Condition, y=or, fill=intType)) + 
    geom_bar(position="dodge", stat="identity", col="black") +
    geom_errorbar(aes(ymin=or-se, ymax=or+se),
                  width=.2,                    # Width of the error bars
                  position=position_dodge(.9)) +
   scale_fill_jco() +
   scale_x_discrete("", labels=c("Protein mediated", "ActD", "Non-protein mediated")) +
   scale_y_continuous("OddsRatio") +
   theme_bw() + theme(legend.position=c(0, 1), legend.box.background = element_rect(), legend.justification=c(-0.1, 1.2),
                      legend.title=element_text(size=14), legend.text=element_text(size=12),
                      axis.text=element_text(size=12), axis.title=element_text(size=14)) +
   guides(fill=guide_legend(title="Interaction type"))

pdf("plots/Figure5EF.pdf", width=12, height=6)
   ggarrange(gg1, gg2, ncol=2, nrow=1)
dev.off()
postscript("plots/Figure5EF.eps", width=12, height=6, paper="special")
   ggarrange(gg1, gg2, ncol=2, nrow=1)
dev.off()
ggsave(gg1, filename="plots/Figure5E.pdf", width=6, height=6)
ggsave(gg1, filename="plots/Figure5E.eps", width=6, height=6)
ggsave(gg2, filename="plots/Figure5F.pdf", width=6, height=6)
ggsave(gg2, filename="plots/Figure5F.eps", width=6, height=6)

gg1 = ggplot(tab1, aes(x=Condition, y=or, fill=intType)) + 
   geom_bar(position="dodge", stat="identity", col="black") +
   scale_fill_jco() +
   scale_x_discrete("", labels=c("Protein mediated", "ActD", "Non-protein mediated")) +
   scale_y_continuous("OddsRatio") +
   theme_bw() + theme(legend.position=c(0, 1), legend.box.background = element_rect(), legend.justification=c(-0.1, 1.2),
                      legend.title=element_text(size=14), legend.text=element_text(size=12),
                      axis.text=element_text(size=12), axis.title=element_text(size=14),) +
   guides(fill=guide_legend(title="Interaction type"))

gg2 = ggplot(tab2, aes(x=Condition, y=or, fill=intType)) + 
   geom_bar(position="dodge", stat="identity", col="black") +
   scale_fill_jco() +
   scale_x_discrete("", labels=c("Protein mediated", "ActD", "Non-protein mediated")) +
   scale_y_continuous("OddsRatio") +
   theme_bw() + theme(legend.position=c(0, 1), legend.box.background = element_rect(), legend.justification=c(-0.1, 1.2),
                      legend.title=element_text(size=14), legend.text=element_text(size=12),
                      axis.text=element_text(size=12), axis.title=element_text(size=14)) +
   guides(fill=guide_legend(title="Interaction type"))

pdf("plots/Figure5EF_noSEM.pdf", width=12, height=6)
   ggarrange(gg1, gg2, ncol=2, nrow=1)
dev.off()
postscript("plots/Figure5EF_noSEM.eps", width=12, height=6, paper="special")
   ggarrange(gg1, gg2, ncol=2, nrow=1)
dev.off()

# Figure 6E
fig = fread("Figure6/figure6c.tab", skip=1, col.names=c("index", "position", "variable", "value"))
fig[, variable := factor(variable, levels=c(unique(variable)))]

gg1 = ggplot(fig, aes(x=position, y=value, colour=variable)) +
   ggtitle("", subtitle="H3K36me3") +
   geom_line(lwd=1) +
   scale_x_continuous("Distance from peaks (bp)") +
   scale_y_continuous("Normalized count") +
   scale_color_ptol("cyl") +
   theme_bw() + theme(legend.position=c(0.8, 0.8), legend.title=element_blank(), legend.box.background = element_rect())
gg1.1 = ggplot(fig, aes(x=position, y=value, colour=variable)) +
   ggtitle("", subtitle="H3K36me3") +
   geom_smooth(method="loess", formula=y~x, n=1000, span=0.2, lwd=1, se=FALSE) +
   scale_x_continuous("Distance from peaks (bp)") +
   scale_y_continuous("Normalized count") +
   scale_color_ptol("cyl") +
   theme_bw() + theme(legend.position=c(0.8, 0.8), legend.title=element_blank(), legend.box.background = element_rect())
ggsave(gg1.1, file="plots/Figure6E.pdf", height=5, width=6)

# # Figure 6F
# fig = fread("Figure6/rna.H3K27ac_12.10000.1.norm.smooth.tab", skip=1, col.names=c("index", "position", "variable", "value"))
# fig[, variable := factor(variable, levels=c(unique(variable)))]
# 
# gg2 = ggplot(fig, aes(x=position, y=value, colour=variable)) +
#    ggtitle("", subtitle="H3K27ac") +
#    geom_line(lwd=1) +
#    scale_x_continuous("Distance from peaks (bp)") +
#    scale_y_continuous("Normalized count") +
#    scale_color_ptol("cyl") +
#    theme_bw() + theme(legend.position=c(0.8, 0.8), legend.title=element_blank(), legend.box.background = element_rect())
# gg2.1 = ggplot(fig, aes(x=position, y=value, colour=variable)) +
#    ggtitle("", subtitle="H3K27ac") +
#    geom_smooth(method="loess", formula=y~x, n=1000, span=0.2, lwd=1, se=FALSE) +
#    scale_x_continuous("Distance from peaks (bp)") +
#    scale_y_continuous("Normalized count") +
#    scale_color_ptol("cyl") +
#    theme_bw() + theme(legend.position=c(0.8, 0.8), legend.title=element_blank(), legend.box.background = element_rect())

# Figure 6F
fig = fread("Figure6/rna.H3K4me3_12.10000.1.norm.smooth.tab", skip=1, col.names=c("index", "position", "variable", "value"))
fig[, variable := factor(variable, levels=c(unique(variable)))]

gg3 = ggplot(fig, aes(x=position, y=value, colour=variable)) +
   ggtitle("", subtitle="H3K4me3") +
   geom_line(lwd=1) +
   scale_x_continuous("Distance from peaks (bp)") +
   scale_y_continuous("Normalized count") +
   scale_color_ptol("cyl") +
   theme_bw() + theme(legend.position=c(0.8, 0.8), legend.title=element_blank(), legend.box.background = element_rect())
gg3.1 = ggplot(fig, aes(x=position, y=value, colour=variable)) +
   ggtitle("", subtitle="H3K4me3") +
   geom_smooth(method="loess", formula=y~x, n=1000, span=0.2, lwd=1, se=FALSE) +
   scale_x_continuous("Distance from peaks (bp)") +
   scale_y_continuous("Normalized count") +
   scale_color_ptol("cyl") +
   theme_bw() + theme(legend.position=c(0.8, 0.8), legend.title=element_blank(), legend.box.background = element_rect())
ggsave(gg3.1, file="plots/Figure6F.pdf", height=5, width=6)

# Figure 6G
fig = fread("Figure6/rna.H3K9ac_12.10000.1.norm.smooth.tab", skip=1, col.names=c("index", "position", "variable", "value"))
fig[, variable := factor(variable, levels=c(unique(variable)))]

gg4 = ggplot(fig, aes(x=position, y=value, colour=variable)) +
   ggtitle("", subtitle="H3K9ac") +
   geom_line(lwd=1) +
   scale_x_continuous("Distance from peaks (bp)") +
   scale_y_continuous("Normalized count") +
   scale_color_ptol("cyl") +
   theme_bw() + theme(legend.position=c(0.8, 0.8), legend.title=element_blank(), legend.box.background = element_rect())
gg4.1 = ggplot(fig, aes(x=position, y=value, colour=variable)) +
   ggtitle("", subtitle="H3K9ac") +
   geom_smooth(method="loess", formula=y~x, n=1000, span=0.2, lwd=1, se=FALSE) +
   scale_x_continuous("Distance from peaks (bp)") +
   scale_y_continuous("Normalized count") +
   scale_color_ptol("cyl") +
   theme_bw() + theme(legend.position=c(0.8, 0.8), legend.title=element_blank(), legend.box.background = element_rect())
ggsave(gg4.1, file="plots/Figure6G.pdf", height=5, width=6)

pdf("plots/Figures6E-H.pdf", width=14, height=4)
   ggarrange(gg1, gg3, gg4, ncol=3, nrow=1, labels=c("E", "F", "G"))
dev.off()

pdf("plots/Figures6E-H_smooth.pdf", width=14, height=4)
   ggarrange(gg1.1, gg3.1, gg4.1, ncol=3, nrow=1, labels=c("E", "F", "G"))
dev.off()
```
